// melange.js/js_exn.js
// Generated by Melange
'use strict';

const Caml_exceptions = require("melange.js/caml_exceptions.js");
const Caml_js_exceptions = require("./caml_js_exceptions.js");
const Caml_option = require("melange.js/caml_option.js");

const $$Error = /* @__PURE__ */ Caml_exceptions.create("Js__Js_exn.Error");

function asJsExn(exn) {
  if (exn.MEL_EXN_ID === $$Error) {
    return Caml_option.some(exn._1);
  }
  
}

const anyToExnInternal = Caml_js_exceptions.internalToOCamlException;

function raiseError(str) {
  throw new Error(str);
}

function raiseEvalError(str) {
  throw new EvalError(str);
}

function raiseRangeError(str) {
  throw new RangeError(str);
}

function raiseReferenceError(str) {
  throw new ReferenceError(str);
}

function raiseSyntaxError(str) {
  throw new SyntaxError(str);
}

function raiseTypeError(str) {
  throw new TypeError(str);
}

function raiseUriError(str) {
  throw new URIError(str);
}

module.exports = {
  $$Error,
  asJsExn,
  anyToExnInternal,
  raiseError,
  raiseEvalError,
  raiseRangeError,
  raiseReferenceError,
  raiseSyntaxError,
  raiseTypeError,
  raiseUriError,
}
/* Caml_js_exceptions Not a pure module */
