// melange.belt/belt_HashSet.js
// Generated by Melange
'use strict';

const Belt__Belt_internalBucketsType = require("./belt_internalBucketsType.js");
const Belt__Belt_internalSetBuckets = require("./belt_internalSetBuckets.js");

function copyBucket(hash, h_buckets, ndata_tail, _old_bucket) {
  while (true) {
    const old_bucket = _old_bucket;
    if (old_bucket === undefined) {
      return;
    }
    const nidx = hash(old_bucket.key) & (h_buckets.length - 1 | 0);
    const tail = ndata_tail[nidx];
    if (tail !== undefined) {
      tail.next = old_bucket;
    } else {
      h_buckets[nidx] = old_bucket;
    }
    ndata_tail[nidx] = old_bucket;
    _old_bucket = old_bucket.next;
    continue;
  };
}

function remove(h, key) {
  const eq = h.eq;
  const h_buckets = h.buckets;
  const i = h.hash(key) & (h_buckets.length - 1 | 0);
  const l = h_buckets[i];
  if (l === undefined) {
    return;
  }
  const next_cell = l.next;
  if (eq(l.key, key)) {
    h.size = h.size - 1 | 0;
    h_buckets[i] = next_cell;
    return;
  } else if (next_cell !== undefined) {
    let _prec = l;
    let _cell = next_cell;
    while (true) {
      const cell = _cell;
      const prec = _prec;
      const cell_next = cell.next;
      if (eq(cell.key, key)) {
        prec.next = cell_next;
        h.size = h.size - 1 | 0;
        return;
      }
      if (cell_next === undefined) {
        return;
      }
      _cell = cell_next;
      _prec = cell;
      continue;
    };
  } else {
    return;
  }
}

function addBucket(h, key, _cell, eq) {
  while (true) {
    const cell = _cell;
    if (eq(cell.key, key)) {
      return;
    }
    const n = cell.next;
    if (n !== undefined) {
      _cell = n;
      continue;
    }
    h.size = h.size + 1 | 0;
    cell.next = {
      key: key,
      next: undefined
    };
    return;
  };
}

function add0(h, key, hash, eq) {
  const h_buckets = h.buckets;
  const buckets_len = h_buckets.length;
  const i = hash(key) & (buckets_len - 1 | 0);
  const l = h_buckets[i];
  if (l !== undefined) {
    addBucket(h, key, l, eq);
  } else {
    h.size = h.size + 1 | 0;
    h_buckets[i] = {
      key: key,
      next: undefined
    };
  }
  if (h.size > (buckets_len << 1)) {
    const odata = h.buckets;
    const osize = odata.length;
    const nsize = (osize << 1);
    if (nsize < osize) {
      return;
    }
    const h_buckets$1 = new Array(nsize);
    const ndata_tail = new Array(nsize);
    h.buckets = h_buckets$1;
    for (let i$1 = 0; i$1 < osize; ++i$1) {
      copyBucket(hash, h_buckets$1, ndata_tail, odata[i$1]);
    }
    for (let i$2 = 0; i$2 < nsize; ++i$2) {
      const tail = ndata_tail[i$2];
      if (tail !== undefined) {
        tail.next = undefined;
      }
      
    }
    return;
  }
  
}

function add(h, key) {
  add0(h, key, h.hash, h.eq);
}

function has(h, key) {
  const eq = h.eq;
  const h_buckets = h.buckets;
  const nid = h.hash(key) & (h_buckets.length - 1 | 0);
  const bucket = h_buckets[nid];
  if (bucket !== undefined) {
    let _cell = bucket;
    while (true) {
      const cell = _cell;
      if (eq(cell.key, key)) {
        return true;
      }
      const nextCell = cell.next;
      if (nextCell === undefined) {
        return false;
      }
      _cell = nextCell;
      continue;
    };
  } else {
    return false;
  }
}

function make(hintSize, id) {
  return Belt__Belt_internalBucketsType.make(id.hash, id.eq, hintSize);
}

function size(h) {
  return h.size;
}

function fromArray(arr, id) {
  const eq = id.eq;
  const hash = id.hash;
  const len = arr.length;
  const v = Belt__Belt_internalBucketsType.make(hash, eq, len);
  for (let i = 0; i < len; ++i) {
    add0(v, arr[i], hash, eq);
  }
  return v;
}

function mergeMany(h, arr) {
  const eq = h.eq;
  const hash = h.hash;
  const len = arr.length;
  for (let i = 0; i < len; ++i) {
    add0(h, arr[i], hash, eq);
  }
}

const clear = Belt__Belt_internalBucketsType.clear;

const isEmpty = Belt__Belt_internalBucketsType.isEmpty;

const copy = Belt__Belt_internalSetBuckets.copy;

const forEachU = Belt__Belt_internalSetBuckets.forEachU;

const forEach = Belt__Belt_internalSetBuckets.forEach;

const reduceU = Belt__Belt_internalSetBuckets.reduceU;

const reduce = Belt__Belt_internalSetBuckets.reduce;

const logStats = Belt__Belt_internalSetBuckets.logStats;

const toArray = Belt__Belt_internalSetBuckets.toArray;

const getBucketHistogram = Belt__Belt_internalSetBuckets.getBucketHistogram;

module.exports = {
  make,
  clear,
  isEmpty,
  add,
  copy,
  has,
  remove,
  forEachU,
  forEach,
  reduceU,
  reduce,
  size,
  logStats,
  toArray,
  fromArray,
  mergeMany,
  getBucketHistogram,
}
/* No side effect */
