// melange.belt/belt_MapString.js
// Generated by Melange
'use strict';

const Belt__Belt_internalAVLtree = require("./belt_internalAVLtree.js");
const Belt__Belt_internalMapString = require("./belt_internalMapString.js");
const Caml_option = require("melange.js/caml_option.js");
const Curry = require("melange.js/curry.js");

function set(t, newK, newD) {
  if (t === undefined) {
    return Belt__Belt_internalAVLtree.singleton(newK, newD);
  }
  const n = Caml_option.valFromOption(t);
  const k = n.k;
  if (newK === k) {
    return Belt__Belt_internalAVLtree.updateValue(n, newD);
  }
  const v = n.v;
  if (newK < k) {
    return Belt__Belt_internalAVLtree.bal(set(n.l, newK, newD), k, v, n.r);
  } else {
    return Belt__Belt_internalAVLtree.bal(n.l, k, v, set(n.r, newK, newD));
  }
}

function updateU(t, x, f) {
  if (t !== undefined) {
    const n = Caml_option.valFromOption(t);
    const k = n.k;
    if (x === k) {
      const data = f(Caml_option.some(n.v));
      if (data !== undefined) {
        return Belt__Belt_internalAVLtree.updateValue(n, Caml_option.valFromOption(data));
      }
      const l = n.l;
      const r = n.r;
      if (l === undefined) {
        return r;
      }
      if (r === undefined) {
        return l;
      }
      const rn = Caml_option.valFromOption(r);
      const kr = {
        contents: rn.k
      };
      const vr = {
        contents: rn.v
      };
      const r$1 = Belt__Belt_internalAVLtree.removeMinAuxWithRef(rn, kr, vr);
      return Belt__Belt_internalAVLtree.bal(l, kr.contents, vr.contents, r$1);
    }
    const v = n.v;
    const l$1 = n.l;
    const r$2 = n.r;
    if (x < k) {
      const ll = updateU(l$1, x, f);
      if (l$1 === ll) {
        return t;
      } else {
        return Belt__Belt_internalAVLtree.bal(ll, k, v, r$2);
      }
    }
    const rr = updateU(r$2, x, f);
    if (r$2 === rr) {
      return t;
    } else {
      return Belt__Belt_internalAVLtree.bal(l$1, k, v, rr);
    }
  }
  const data$1 = f(undefined);
  if (data$1 !== undefined) {
    return Belt__Belt_internalAVLtree.singleton(x, Caml_option.valFromOption(data$1));
  } else {
    return t;
  }
}

function update(t, x, f) {
  return updateU(t, x, Curry.__1(f));
}

function removeAux(n, x) {
  const v = n.k;
  const l = n.l;
  const r = n.r;
  if (x === v) {
    if (l === undefined) {
      return r;
    }
    if (r === undefined) {
      return l;
    }
    const rn = Caml_option.valFromOption(r);
    const kr = {
      contents: rn.k
    };
    const vr = {
      contents: rn.v
    };
    const r$1 = Belt__Belt_internalAVLtree.removeMinAuxWithRef(rn, kr, vr);
    return Belt__Belt_internalAVLtree.bal(l, kr.contents, vr.contents, r$1);
  }
  if (x < v) {
    if (l === undefined) {
      return n;
    }
    const ll = removeAux(Caml_option.valFromOption(l), x);
    if (ll === l) {
      return n;
    } else {
      return Belt__Belt_internalAVLtree.bal(ll, v, n.v, r);
    }
  }
  if (r === undefined) {
    return n;
  }
  const rr = removeAux(Caml_option.valFromOption(r), x);
  return Belt__Belt_internalAVLtree.bal(l, v, n.v, rr);
}

function remove(n, x) {
  if (n !== undefined) {
    return removeAux(n, x);
  }
  
}

function removeMany(t, keys) {
  const len = keys.length;
  if (t !== undefined) {
    let _t = t;
    let _i = 0;
    while (true) {
      const i = _i;
      const t$1 = _t;
      if (i >= len) {
        return t$1;
      }
      const ele = keys[i];
      const u = removeAux(t$1, ele);
      if (u === undefined) {
        return u;
      }
      _i = i + 1 | 0;
      _t = Caml_option.valFromOption(u);
      continue;
    };
  }
  
}

function mergeMany(h, arr) {
  const len = arr.length;
  let v = h;
  for (let i = 0; i < len; ++i) {
    const match = arr[i];
    v = set(v, match[0], match[1]);
  }
  return v;
}

let empty;

const isEmpty = Belt__Belt_internalAVLtree.isEmpty;

const has = Belt__Belt_internalMapString.has;

const cmpU = Belt__Belt_internalMapString.cmpU;

const cmp = Belt__Belt_internalMapString.cmp;

const eqU = Belt__Belt_internalMapString.eqU;

const eq = Belt__Belt_internalMapString.eq;

const findFirstByU = Belt__Belt_internalAVLtree.findFirstByU;

const findFirstBy = Belt__Belt_internalAVLtree.findFirstBy;

const forEachU = Belt__Belt_internalAVLtree.forEachU;

const forEach = Belt__Belt_internalAVLtree.forEach;

const reduceU = Belt__Belt_internalAVLtree.reduceU;

const reduce = Belt__Belt_internalAVLtree.reduce;

const everyU = Belt__Belt_internalAVLtree.everyU;

const every = Belt__Belt_internalAVLtree.every;

const someU = Belt__Belt_internalAVLtree.someU;

const some = Belt__Belt_internalAVLtree.some;

const size = Belt__Belt_internalAVLtree.size;

const toList = Belt__Belt_internalAVLtree.toList;

const toArray = Belt__Belt_internalAVLtree.toArray;

const fromArray = Belt__Belt_internalMapString.fromArray;

const keysToArray = Belt__Belt_internalAVLtree.keysToArray;

const valuesToArray = Belt__Belt_internalAVLtree.valuesToArray;

const minKey = Belt__Belt_internalAVLtree.minKey;

const minKeyUndefined = Belt__Belt_internalAVLtree.minKeyUndefined;

const maxKey = Belt__Belt_internalAVLtree.maxKey;

const maxKeyUndefined = Belt__Belt_internalAVLtree.maxKeyUndefined;

const minimum = Belt__Belt_internalAVLtree.minimum;

const minUndefined = Belt__Belt_internalAVLtree.minUndefined;

const maximum = Belt__Belt_internalAVLtree.maximum;

const maxUndefined = Belt__Belt_internalAVLtree.maxUndefined;

const get = Belt__Belt_internalMapString.get;

const getUndefined = Belt__Belt_internalMapString.getUndefined;

const getWithDefault = Belt__Belt_internalMapString.getWithDefault;

const getExn = Belt__Belt_internalMapString.getExn;

const checkInvariantInternal = Belt__Belt_internalAVLtree.checkInvariantInternal;

const mergeU = Belt__Belt_internalMapString.mergeU;

const merge = Belt__Belt_internalMapString.merge;

const keepU = Belt__Belt_internalAVLtree.keepSharedU;

const keep = Belt__Belt_internalAVLtree.keepShared;

const partitionU = Belt__Belt_internalAVLtree.partitionSharedU;

const partition = Belt__Belt_internalAVLtree.partitionShared;

const split = Belt__Belt_internalMapString.split;

const mapU = Belt__Belt_internalAVLtree.mapU;

const map = Belt__Belt_internalAVLtree.map;

const mapWithKeyU = Belt__Belt_internalAVLtree.mapWithKeyU;

const mapWithKey = Belt__Belt_internalAVLtree.mapWithKey;

module.exports = {
  empty,
  isEmpty,
  has,
  cmpU,
  cmp,
  eqU,
  eq,
  findFirstByU,
  findFirstBy,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  size,
  toList,
  toArray,
  fromArray,
  keysToArray,
  valuesToArray,
  minKey,
  minKeyUndefined,
  maxKey,
  maxKeyUndefined,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getWithDefault,
  getExn,
  checkInvariantInternal,
  remove,
  removeMany,
  set,
  updateU,
  update,
  mergeU,
  merge,
  mergeMany,
  keepU,
  keep,
  partitionU,
  partition,
  split,
  mapU,
  map,
  mapWithKeyU,
  mapWithKey,
}
/* No side effect */
