// melange.js/caml_hash.mjs
// Generated by Melange

import * as Caml_hash_primitive from "./caml_hash_primitive.mjs";

function push_back(q, v) {
  const cell = {
    content: v,
    next: undefined
  };
  const last = q.last;
  if (last !== undefined) {
    q.length = q.length + 1 | 0;
    last.next = cell;
    q.last = cell;
  } else {
    q.length = 1;
    q.first = cell;
    q.last = cell;
  }
}

function unsafe_pop(q) {
  const cell = q.first;
  const next = cell.next;
  if (next !== undefined) {
    q.length = q.length - 1 | 0;
    q.first = next;
  } else {
    q.length = 0;
    q.first = undefined;
    q.last = undefined;
  }
  return cell.content;
}

function caml_hash(count, _limit, seed, obj) {
  let hash = seed;
  const match = typeof obj;
  switch (match) {
    case "number" : {
      const u = obj | 0;
      hash = Caml_hash_primitive.caml_hash_mix_int(hash, (u + u | 0) + 1 | 0);
      return Caml_hash_primitive.caml_hash_final_mix(hash);
      }
    case "string" : {
      hash = Caml_hash_primitive.caml_hash_mix_string(hash, obj);
      return Caml_hash_primitive.caml_hash_final_mix(hash);
      }
    default:
      const queue = {
        length: 0,
        first: undefined,
        last: undefined
      };
      let num = count;
      push_back(queue, obj);
      num = num - 1 | 0;
      while (queue.length !== 0 && num > 0) {
        const obj$1 = unsafe_pop(queue);
        const match$1 = typeof obj$1;
        switch (match$1) {
          case "boolean" : {
            const u$1 = obj$1 ? 1 : 0;
            hash = Caml_hash_primitive.caml_hash_mix_int(hash, (u$1 + u$1 | 0) + 1 | 0);
            num = num - 1 | 0;
            break;
            }
          case "number" : {
            const u$2 = obj$1 | 0;
            hash = Caml_hash_primitive.caml_hash_mix_int(hash, (u$2 + u$2 | 0) + 1 | 0);
            num = num - 1 | 0;
            break;
            }
          case "string" : {
            hash = Caml_hash_primitive.caml_hash_mix_string(hash, obj$1);
            num = num - 1 | 0;
            break;
            }
          case "function" : {
            }
          case "symbol" : {
            }
          case "undefined" : {
            break;
            }
          default:
            const size = obj$1.length | 0;
            if (size !== 0) {
              const obj_tag = obj$1.TAG;
              const tag = (size << 10) | obj_tag;
              if (obj_tag === 248) {
                hash = Caml_hash_primitive.caml_hash_mix_int(hash, obj$1[1]);
              } else {
                hash = Caml_hash_primitive.caml_hash_mix_int(hash, tag);
                const v = size - 1 | 0;
                const block = v < num ? v : num;
                for (let i = 0; i <= block; ++i) {
                  push_back(queue, obj$1[i]);
                }
              }
            } else {
              const size$1 = (function(obj,cb){
            var size = 0
            for(var k in obj){
              cb(obj[k])
              ++ size
            }
            return size
          }
              )(obj$1, (function (v) {
                push_back(queue, v);
              }));
              hash = Caml_hash_primitive.caml_hash_mix_int(hash, (size$1 << 10) | 0);
            }
        }
      };
      return Caml_hash_primitive.caml_hash_final_mix(hash);
  }
}

function caml_string_hash(seed, s) {
  const h = Caml_hash_primitive.caml_hash_mix_string(seed, s);
  const h$1 = Caml_hash_primitive.caml_hash_final_mix(h);
  return h$1 & 1073741823;
}

export {
  caml_hash,
  caml_string_hash,
}
/* No side effect */
