// melange.js/caml_obj.mjs
// Generated by Melange

import * as Caml from "melange.js/caml.mjs";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";

const for_in = (function(o,foo){
        for (var x in o) { foo(x) }}
);

const caml_obj_dup = (function(x){
  if(Array.isArray(x)){
    var len = x.length
    var v = new Array(len)
    for(var i = 0 ; i < len ; ++i){
      v[i] = x[i]
    }
    if(x.TAG !== undefined){
      v.TAG = x.TAG // TODO this can be removed eventually
    }
    return v
  }
  return Object.assign({},x)
}
);

const update_dummy = (function(x,y){
  var k
  if(Array.isArray(y)){
    for(k = 0; k < y.length ; ++k){
      x[k] = y[k]
    }
    if(y.TAG !== undefined){
      x.TAG = y.TAG
    }
  } else {
    for (var k in y){
      x[k] = y[k]
    }
  }
}
);

function caml_compare(a, b) {
  if (a === b) {
    return 0;
  }
  const a_type = typeof a;
  const b_type = typeof b;
  switch (a_type) {
    case "bigint" :
      if (b_type === "bigint") {
        return Caml.caml_float_compare(a, b);
      }
      break;
    case "boolean" :
      if (b_type === "boolean") {
        return Caml.caml_int_compare(a, b);
      }
      break;
    case "function" :
      if (b_type === "function") {
        throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
            MEL_EXN_ID: "Invalid_argument",
            _1: "compare: functional value"
          });
      }
      break;
    case "number" :
      if (b_type === "number") {
        return Caml.caml_float_compare(a, b);
      }
      break;
    case "string" :
      if (b_type === "string") {
        return Caml.caml_string_compare(a, b);
      } else {
        return 1;
      }
    case "undefined" :
      return -1;
  }
  switch (b_type) {
    case "string" :
      return -1;
    case "undefined" :
      return 1;
    default:
      if (a_type === "boolean") {
        return 1;
      }
      if (b_type === "boolean") {
        return -1;
      }
      if (a_type === "function") {
        return 1;
      }
      if (b_type === "function") {
        return -1;
      }
      if (a_type === "number") {
        if (b === null || b.MEL_PRIVATE_NESTED_SOME_NONE !== undefined) {
          return 1;
        } else {
          return -1;
        }
      }
      if (b_type === "number") {
        if (a === null || a.MEL_PRIVATE_NESTED_SOME_NONE !== undefined) {
          return -1;
        } else {
          return 1;
        }
      }
      if (a === null) {
        if (b.MEL_PRIVATE_NESTED_SOME_NONE !== undefined) {
          return 1;
        } else {
          return -1;
        }
      }
      if (b === null) {
        if (a.MEL_PRIVATE_NESTED_SOME_NONE !== undefined) {
          return -1;
        } else {
          return 1;
        }
      }
      if (a.MEL_PRIVATE_NESTED_SOME_NONE !== undefined) {
        if (b.MEL_PRIVATE_NESTED_SOME_NONE !== undefined) {
          return aux_obj_compare(a, b);
        } else {
          return -1;
        }
      }
      const tag_a = a.TAG;
      const tag_b = b.TAG;
      if (tag_a === 248) {
        return Caml.caml_int_compare(a[1], b[1]);
      }
      if (tag_a === 251) {
        throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
            MEL_EXN_ID: "Invalid_argument",
            _1: "equal: abstract value"
          });
      }
      if (tag_a !== tag_b) {
        if (tag_a < tag_b) {
          return -1;
        } else {
          return 1;
        }
      }
      const len_a = a.length | 0;
      const len_b = b.length | 0;
      if (len_a === len_b) {
        if (Array.isArray(a)) {
          let _i = 0;
          while (true) {
            const i = _i;
            if (i === len_a) {
              return 0;
            }
            const res = caml_compare(a[i], b[i]);
            if (res !== 0) {
              return res;
            }
            _i = i + 1 | 0;
            continue;
          };
        } else if ((a instanceof Date && b instanceof Date
          )) {
          return (a - b
          );
        } else {
          return aux_obj_compare(a, b);
        }
      } else if (len_a < len_b) {
        let _i$1 = 0;
        while (true) {
          const i$1 = _i$1;
          if (i$1 === len_a) {
            return -1;
          }
          const res$1 = caml_compare(a[i$1], b[i$1]);
          if (res$1 !== 0) {
            return res$1;
          }
          _i$1 = i$1 + 1 | 0;
          continue;
        };
      } else {
        let _i$2 = 0;
        while (true) {
          const i$2 = _i$2;
          if (i$2 === len_b) {
            return 1;
          }
          const res$2 = caml_compare(a[i$2], b[i$2]);
          if (res$2 !== 0) {
            return res$2;
          }
          _i$2 = i$2 + 1 | 0;
          continue;
        };
      }
  }
}

function aux_obj_compare(a, b) {
  const min_key_lhs = {
    contents: undefined
  };
  const min_key_rhs = {
    contents: undefined
  };
  const do_key = function (param, key) {
    const min_key = param[2];
    const b = param[1];
    if (!(!Object.prototype.hasOwnProperty.call(b, key) || caml_compare(param[0][key], b[key]) > 0)) {
      return;
    }
    const mk = min_key.contents;
    if (mk !== undefined && key >= mk) {
      return;
    } else {
      min_key.contents = key;
      return;
    }
  };
  const partial_arg = [
    a,
    b,
    min_key_rhs
  ];
  const do_key_a = function (param) {
    return do_key(partial_arg, param);
  };
  const partial_arg$1 = [
    b,
    a,
    min_key_lhs
  ];
  const do_key_b = function (param) {
    return do_key(partial_arg$1, param);
  };
  for_in(a, do_key_a);
  for_in(b, do_key_b);
  const match = min_key_lhs.contents;
  const match$1 = min_key_rhs.contents;
  if (match !== undefined) {
    if (match$1 !== undefined) {
      return Caml.caml_string_compare(match, match$1);
    } else {
      return -1;
    }
  } else if (match$1 !== undefined) {
    return 1;
  } else {
    return 0;
  }
}

function caml_equal(a, b) {
  if (a === b) {
    return true;
  }
  const a_type = typeof a;
  if (a_type === "string" || a_type === "number" || a_type === "bigint" || a_type === "boolean" || a_type === "undefined" || a === null) {
    return false;
  }
  const b_type = typeof b;
  if (a_type === "function" || b_type === "function") {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "equal: functional value"
      });
  }
  if (b_type === "number" || b_type === "bigint" || b_type === "undefined" || b === null) {
    return false;
  }
  const tag_a = a.TAG;
  const tag_b = b.TAG;
  if (tag_a === 248) {
    return a[1] === b[1];
  }
  if (tag_a === 251) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "equal: abstract value"
      });
  }
  if (tag_a !== tag_b) {
    return false;
  }
  const len_a = a.length | 0;
  const len_b = b.length | 0;
  if (len_a === len_b) {
    if (Array.isArray(a)) {
      let _i = 0;
      while (true) {
        const i = _i;
        if (i === len_a) {
          return true;
        }
        if (!caml_equal(a[i], b[i])) {
          return false;
        }
        _i = i + 1 | 0;
        continue;
      };
    } else if ((a instanceof Date && b instanceof Date
      )) {
      return !(a > b || a < b);
    } else {
      const result = {
        contents: true
      };
      const do_key_a = function (key) {
        if (!Object.prototype.hasOwnProperty.call(b, key)) {
          result.contents = false;
          return;
        }
        
      };
      const do_key_b = function (key) {
        if (!Object.prototype.hasOwnProperty.call(a, key) || !caml_equal(b[key], a[key])) {
          result.contents = false;
          return;
        }
        
      };
      for_in(a, do_key_a);
      if (result.contents) {
        for_in(b, do_key_b);
      }
      return result.contents;
    }
  } else {
    return false;
  }
}

function caml_equal_null(x, y) {
  if (y !== null) {
    return caml_equal(x, y);
  } else {
    return x === y;
  }
}

function caml_equal_undefined(x, y) {
  if (y !== undefined) {
    return caml_equal(x, y);
  } else {
    return x === y;
  }
}

function caml_equal_nullable(x, y) {
  if (y == null) {
    return x === y;
  } else {
    return caml_equal(x, y);
  }
}

function caml_notequal(a, b) {
  if ((typeof a === "number" || typeof a === "bigint") && (typeof b === "number" || typeof b === "bigint")) {
    return a !== b;
  } else {
    return !caml_equal(a, b);
  }
}

function caml_greaterequal(a, b) {
  if ((typeof a === "number" || typeof a === "bigint") && (typeof b === "number" || typeof b === "bigint")) {
    return a >= b;
  } else {
    return caml_compare(a, b) >= 0;
  }
}

function caml_greaterthan(a, b) {
  if ((typeof a === "number" || typeof a === "bigint") && (typeof b === "number" || typeof b === "bigint")) {
    return a > b;
  } else {
    return caml_compare(a, b) > 0;
  }
}

function caml_lessequal(a, b) {
  if ((typeof a === "number" || typeof a === "bigint") && (typeof b === "number" || typeof b === "bigint")) {
    return a <= b;
  } else {
    return caml_compare(a, b) <= 0;
  }
}

function caml_lessthan(a, b) {
  if ((typeof a === "number" || typeof a === "bigint") && (typeof b === "number" || typeof b === "bigint")) {
    return a < b;
  } else {
    return caml_compare(a, b) < 0;
  }
}

function caml_min(x, y) {
  if (caml_compare(x, y) <= 0) {
    return x;
  } else {
    return y;
  }
}

function caml_max(x, y) {
  if (caml_compare(x, y) >= 0) {
    return x;
  } else {
    return y;
  }
}

export {
  caml_obj_dup,
  update_dummy,
  caml_compare,
  caml_equal,
  caml_equal_null,
  caml_equal_undefined,
  caml_equal_nullable,
  caml_notequal,
  caml_greaterequal,
  caml_greaterthan,
  caml_lessthan,
  caml_lessequal,
  caml_min,
  caml_max,
}
/* No side effect */
