// melange.js/caml_float.js
// Generated by Melange
'use strict';


const caml_int32_float_of_bits = (function(x){
    return new Float32Array(new Int32Array([x]).buffer)[0]
    }
);

const caml_int32_bits_of_float = (function(x){
  return new Int32Array(new Float32Array([x]).buffer)[0]
}
);

function caml_modf_float(x) {
  if (!isFinite(x)) {
    if (isNaN(x)) {
      return [
        NaN,
        NaN
      ];
    } else {
      return [
        1 / x,
        x
      ];
    }
  }
  const neg = 1 / x < 0;
  const x$1 = Math.abs(x);
  const i = Math.floor(x$1);
  const f = x$1 - i;
  if (neg) {
    return [
      - f,
      - i
    ];
  } else {
    return [
      f,
      i
    ];
  }
}

function caml_ldexp_float(x, exp) {
  let x$p = x;
  let exp$p = exp;
  if (exp$p > 1023) {
    exp$p = exp$p - 1023;
    x$p = x$p * Math.pow(2, 1023);
    if (exp$p > 1023) {
      exp$p = exp$p - 1023;
      x$p = x$p * Math.pow(2, 1023);
    }
    
  } else if (exp$p < -1023) {
    exp$p = exp$p + 1023;
    x$p = x$p * Math.pow(2, -1023);
  }
  return x$p * Math.pow(2, exp$p);
}

function caml_frexp_float(x) {
  if (x === 0 || !isFinite(x)) {
    return [
      x,
      0
    ];
  }
  const neg = x < 0;
  let x$p = Math.abs(x);
  let exp = Math.floor(Math.LOG2E * Math.log(x$p)) + 1;
  x$p = x$p * Math.pow(2, - exp);
  if (x$p < 0.5) {
    x$p = x$p * 2;
    exp = exp - 1;
  }
  if (neg) {
    x$p = - x$p;
  }
  return [
    x$p,
    exp | 0
  ];
}

function caml_copysign_float(x, y) {
  const x$1 = Math.abs(x);
  const y$1 = y === 0 ? 1 / y : y;
  if (y$1 < 0) {
    return - x$1;
  } else {
    return x$1;
  }
}

function caml_signbit_float(x) {
  if (x === 0 && 1 / x === -Infinity) {
    return true;
  } else if (x === 0 && 1 / x === Infinity) {
    return false;
  } else {
    return x < 0;
  }
}

function caml_expm1_float(x) {
  const y = Math.exp(x);
  const z = y - 1;
  if (Math.abs(x) > 1) {
    return z;
  } else if (z === 0) {
    return x;
  } else {
    return x * z / Math.log(y);
  }
}

function caml_hypot_float(x, y) {
  const x0 = Math.abs(x);
  const y0 = Math.abs(y);
  const a = x0 > y0 ? x0 : y0;
  const b = (
    x0 < y0 ? x0 : y0
  ) / (
    a !== 0 ? a : 1
  );
  return a * Math.sqrt(1 + b * b);
}

function caml_log10_float(x) {
  return Math.LOG10E * Math.log(x);
}

module.exports = {
  caml_int32_float_of_bits,
  caml_int32_bits_of_float,
  caml_modf_float,
  caml_ldexp_float,
  caml_frexp_float,
  caml_copysign_float,
  caml_expm1_float,
  caml_hypot_float,
  caml_log10_float,
  caml_signbit_float,
}
/* No side effect */
