// melange.js/caml_bigstring.mjs
// Generated by Melange


const caml_bigstring_get16 = (function(ba, i) {
    if (i < 0 || i + 1 >= ba.dims[0])
      throw [0, -3, "index out of bounds"];
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    return dv.getInt16(i, true);
  }
);

const caml_bigstring_get16u = (function(ba, i) {
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    return dv.getInt16(i, true);
  }
);

const caml_bigstring_set16 = (function(ba, i, v) {
    if (i < 0 || i + 1 >= ba.dims[0])
      throw [0, -3, "index out of bounds"];
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    dv.setInt16(i, v, true);
  }
);

const caml_bigstring_set16u = (function(ba, i, v) {
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    dv.setInt16(i, v, true);
  }
);

const caml_bigstring_get32 = (function(ba, i) {
    if (i < 0 || i + 3 >= ba.dims[0])
      throw [0, -3, "index out of bounds"];
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    return dv.getInt32(i, true);
  }
);

const caml_bigstring_get32u = (function(ba, i) {
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    return dv.getInt32(i, true);
  }
);

const caml_bigstring_set32 = (function(ba, i, v) {
    if (i < 0 || i + 3 >= ba.dims[0])
      throw [0, -3, "index out of bounds"];
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    dv.setInt32(i, v, true);
  }
);

const caml_bigstring_set32u = (function(ba, i, v) {
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    dv.setInt32(i, v, true);
  }
);

const caml_bigstring_get64 = (function(ba, i) {
    if (i < 0 || i + 7 >= ba.dims[0])
      throw [0, -3, "index out of bounds"];
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    var lo = dv.getInt32(i, true);
    var hi = dv.getInt32(i + 4, true);
    return [lo, hi];
  }
);

const caml_bigstring_get64u = (function(ba, i) {
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    var lo = dv.getInt32(i, true);
    var hi = dv.getInt32(i + 4, true);
    return [lo, hi];
  }
);

const caml_bigstring_set64 = (function(ba, i, v) {
    if (i < 0 || i + 7 >= ba.dims[0])
      throw [0, -3, "index out of bounds"];
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    dv.setInt32(i, v[0], true);
    dv.setInt32(i + 4, v[1], true);
  }
);

const caml_bigstring_set64u = (function(ba, i, v) {
    var dv = new DataView(ba.data.buffer, ba.data.byteOffset, ba.data.byteLength);
    dv.setInt32(i, v[0], true);
    dv.setInt32(i + 4, v[1], true);
  }
);

const caml_bigstring_blit_ba_to_bytes = (function(ba, ba_off, bytes, bytes_off, len) {
    for (var i = 0; i < len; i++) {
      bytes[bytes_off + i] = ba.data[ba_off + i];
    }
  }
);

const caml_bigstring_blit_bytes_to_ba = (function(bytes, bytes_off, ba, ba_off, len) {
    for (var i = 0; i < len; i++) {
      ba.data[ba_off + i] = bytes[bytes_off + i];
    }
  }
);

const caml_bigstring_blit_ba_to_ba = (function(src, src_off, dst, dst_off, len) {
    if (src.data === dst.data && dst_off > src_off) {
      for (var i = len - 1; i >= 0; i--) {
        dst.data[dst_off + i] = src.data[src_off + i];
      }
    } else {
      for (var i = 0; i < len; i++) {
        dst.data[dst_off + i] = src.data[src_off + i];
      }
    }
  }
);

export {
  caml_bigstring_get16,
  caml_bigstring_get16u,
  caml_bigstring_set16,
  caml_bigstring_set16u,
  caml_bigstring_get32,
  caml_bigstring_get32u,
  caml_bigstring_set32,
  caml_bigstring_set32u,
  caml_bigstring_get64,
  caml_bigstring_get64u,
  caml_bigstring_set64,
  caml_bigstring_set64u,
  caml_bigstring_blit_ba_to_bytes,
  caml_bigstring_blit_bytes_to_ba,
  caml_bigstring_blit_ba_to_ba,
}
/* No side effect */
