// melange.js/caml_option.js
// Generated by Melange
'use strict';


function isNested(x) {
  return x.MEL_PRIVATE_NESTED_SOME_NONE !== undefined;
}

function some(x) {
  if (x === undefined) {
    return {
      MEL_PRIVATE_NESTED_SOME_NONE: 0
    };
  } else if (x !== null && x.MEL_PRIVATE_NESTED_SOME_NONE !== undefined) {
    return {
      MEL_PRIVATE_NESTED_SOME_NONE: x.MEL_PRIVATE_NESTED_SOME_NONE + 1 | 0
    };
  } else {
    return x;
  }
}

function nullable_to_opt(x) {
  if (x == null) {
    return;
  } else {
    return some(x);
  }
}

function undefined_to_opt(x) {
  if (x === undefined) {
    return;
  } else {
    return some(x);
  }
}

function null_to_opt(x) {
  if (x === null) {
    return;
  } else {
    return some(x);
  }
}

function valFromOption(x) {
  if (!(x !== null && x.MEL_PRIVATE_NESTED_SOME_NONE !== undefined)) {
    return x;
  }
  const depth = x.MEL_PRIVATE_NESTED_SOME_NONE;
  if (depth === 0) {
    return;
  } else {
    return {
      MEL_PRIVATE_NESTED_SOME_NONE: depth - 1 | 0
    };
  }
}

function option_get(x) {
  if (x === undefined) {
    return;
  } else {
    return valFromOption(x);
  }
}

function option_unwrap(x) {
  if (x !== undefined) {
    return x.VAL;
  } else {
    return x;
  }
}

module.exports = {
  nullable_to_opt,
  undefined_to_opt,
  null_to_opt,
  valFromOption,
  some,
  isNested,
  option_get,
  option_unwrap,
}
/* No side effect */
