// melange.js/js_exn.mjs
// Generated by Melange

import * as Caml_exceptions from "melange.js/caml_exceptions.mjs";
import * as Caml_js_exceptions from "./caml_js_exceptions.mjs";
import * as Caml_option from "melange.js/caml_option.mjs";

const $$Error = /* @__PURE__ */ Caml_exceptions.create("Js__Js_exn.Error");

function asJsExn(exn) {
  if (exn.MEL_EXN_ID === $$Error) {
    return Caml_option.some(exn._1);
  }
  
}

const anyToExnInternal = Caml_js_exceptions.internalToOCamlException;

function raiseError(str) {
  throw new Error(str);
}

function raiseEvalError(str) {
  throw new EvalError(str);
}

function raiseRangeError(str) {
  throw new RangeError(str);
}

function raiseReferenceError(str) {
  throw new ReferenceError(str);
}

function raiseSyntaxError(str) {
  throw new SyntaxError(str);
}

function raiseTypeError(str) {
  throw new TypeError(str);
}

function raiseUriError(str) {
  throw new URIError(str);
}

export {
  $$Error,
  asJsExn,
  anyToExnInternal,
  raiseError,
  raiseEvalError,
  raiseRangeError,
  raiseReferenceError,
  raiseSyntaxError,
  raiseTypeError,
  raiseUriError,
}
/* Caml_js_exceptions Not a pure module */
