// melange/int.js
// Generated by Melange
'use strict';

const Caml = require("melange.js/caml.js");
const Caml_format = require("melange.js/caml_format.js");
const Caml_hash = require("melange.js/caml_hash.js");

function abs(x) {
  if (x >= 0) {
    return x;
  } else {
    return -x | 0;
  }
}

const min_int = -2147483648;

function lognot(x) {
  return x ^ -1;
}

function equal(prim0, prim1) {
  return prim0 === prim1;
}

const compare = Caml.caml_int_compare;

function min(x, y) {
  if (x <= y) {
    return x;
  } else {
    return y;
  }
}

function max(x, y) {
  if (x >= y) {
    return x;
  } else {
    return y;
  }
}

function to_string(x) {
  return Caml_format.caml_format_int("%d", x);
}

function seeded_hash(seed, x) {
  return Caml_hash.caml_hash(10, 100, seed, x);
}

function hash(x) {
  return Caml_hash.caml_hash(10, 100, 0, x);
}

const zero = 0;

const one = 1;

const minus_one = -1;

const max_int = 2147483647;

module.exports = {
  zero,
  one,
  minus_one,
  abs,
  max_int,
  min_int,
  lognot,
  equal,
  compare,
  min,
  max,
  to_string,
  seeded_hash,
  hash,
}
/* No side effect */
