// melange/queue.js
// Generated by Melange
'use strict';

const Caml_exceptions = require("melange.js/caml_exceptions.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Caml_option = require("melange.js/caml_option.js");
const Curry = require("melange.js/curry.js");
const Stdlib__Seq = require("./seq.js");

const Empty = /* @__PURE__ */ Caml_exceptions.create("Stdlib.Queue.Empty");

function create(param) {
  return {
    length: 0,
    first: /* Nil */ 0,
    last: /* Nil */ 0
  };
}

function clear(q) {
  q.length = 0;
  q.first = /* Nil */ 0;
  q.last = /* Nil */ 0;
}

function add(x, q) {
  const cell = {
    TAG: /* Cons */ 0,
    content: x,
    next: /* Nil */ 0
  };
  const last = q.last;
  if (typeof last !== "object" && typeof last !== "function") {
    q.length = 1;
    q.first = cell;
    q.last = cell;
    return;
  }
  q.length = q.length + 1 | 0;
  last.next = cell;
  q.last = cell;
}

function peek(q) {
  const match = q.first;
  if (!(typeof match !== "object" && typeof match !== "function")) {
    return match.content;
  }
  throw new Caml_js_exceptions.MelangeError(Empty, {
      MEL_EXN_ID: Empty
    });
}

function peek_opt(q) {
  const match = q.first;
  if (typeof match !== "object" && typeof match !== "function") {
    return;
  } else {
    return Caml_option.some(match.content);
  }
}

function take(q) {
  const match = q.first;
  if (typeof match !== "object" && typeof match !== "function") {
    throw new Caml_js_exceptions.MelangeError(Empty, {
        MEL_EXN_ID: Empty
      });
  }
  const content = match.content;
  const match$1 = match.next;
  if (typeof match$1 !== "object" && typeof match$1 !== "function") {
    clear(q);
    return content;
  }
  const next = match.next;
  q.length = q.length - 1 | 0;
  q.first = next;
  return content;
}

function take_opt(q) {
  const match = q.first;
  if (typeof match !== "object" && typeof match !== "function") {
    return;
  }
  const content = match.content;
  const match$1 = match.next;
  if (typeof match$1 !== "object" && typeof match$1 !== "function") {
    clear(q);
    return Caml_option.some(content);
  }
  const next = match.next;
  q.length = q.length - 1 | 0;
  q.first = next;
  return Caml_option.some(content);
}

function drop(q) {
  const match = q.first;
  if (typeof match !== "object" && typeof match !== "function") {
    throw new Caml_js_exceptions.MelangeError(Empty, {
        MEL_EXN_ID: Empty
      });
  }
  const match$1 = match.next;
  if (typeof match$1 !== "object" && typeof match$1 !== "function") {
    return clear(q);
  }
  const next = match.next;
  q.length = q.length - 1 | 0;
  q.first = next;
}

function copy(q) {
  let q_res = {
    length: q.length,
    first: /* Nil */ 0,
    last: /* Nil */ 0
  };
  let _prev = /* Nil */ 0;
  let _cell = q.first;
  while (true) {
    const cell = _cell;
    const prev = _prev;
    if (typeof cell !== "object" && typeof cell !== "function") {
      q_res.last = prev;
      return q_res;
    }
    const next = cell.next;
    const res = {
      TAG: /* Cons */ 0,
      content: cell.content,
      next: /* Nil */ 0
    };
    if (typeof prev !== "object" && typeof prev !== "function") {
      q_res.first = res;
    } else {
      prev.next = res;
    }
    _cell = next;
    _prev = res;
    continue;
  };
}

function is_empty(q) {
  return q.length === 0;
}

function length(q) {
  return q.length;
}

function iter(f, q) {
  let _cell = q.first;
  while (true) {
    const cell = _cell;
    if (typeof cell !== "object" && typeof cell !== "function") {
      return;
    }
    const next = cell.next;
    Curry._1(f, cell.content);
    _cell = next;
    continue;
  };
}

function fold(f, accu, q) {
  let _accu = accu;
  let _cell = q.first;
  while (true) {
    const cell = _cell;
    const accu$1 = _accu;
    if (typeof cell !== "object" && typeof cell !== "function") {
      return accu$1;
    }
    const next = cell.next;
    const accu$2 = Curry._2(f, accu$1, cell.content);
    _cell = next;
    _accu = accu$2;
    continue;
  };
}

function transfer(q1, q2) {
  if (q1.length <= 0) {
    return;
  }
  const last = q2.last;
  if (typeof last !== "object" && typeof last !== "function") {
    q2.length = q1.length;
    q2.first = q1.first;
    q2.last = q1.last;
    return clear(q1);
  }
  q2.length = q2.length + q1.length | 0;
  last.next = q1.first;
  q2.last = q1.last;
  clear(q1);
}

function to_seq(q) {
  const aux = function (c, param) {
    if (typeof c !== "object" && typeof c !== "function") {
      return /* Nil */ 0;
    }
    const next = c.next;
    return {
      TAG: /* Cons */ 0,
      _0: c.content,
      _1: (function (param) {
        return aux(next, param);
      })
    };
  };
  const partial_arg = q.first;
  return function (param) {
    return aux(partial_arg, param);
  };
}

function add_seq(q, i) {
  Stdlib__Seq.iter((function (x) {
    add(x, q);
  }), i);
}

function of_seq(g) {
  const q = {
    length: 0,
    first: /* Nil */ 0,
    last: /* Nil */ 0
  };
  add_seq(q, g);
  return q;
}

const push = add;

const pop = take;

const top = peek;

module.exports = {
  Empty,
  create,
  add,
  push,
  take,
  take_opt,
  pop,
  peek,
  peek_opt,
  top,
  drop,
  clear,
  copy,
  is_empty,
  length,
  iter,
  fold,
  transfer,
  to_seq,
  add_seq,
  of_seq,
}
/* No side effect */
