// melange/stack.js
// Generated by Melange
'use strict';

const Caml_exceptions = require("melange.js/caml_exceptions.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Caml_obj = require("melange.js/caml_obj.js");
const Caml_option = require("melange.js/caml_option.js");
const Stdlib__List = require("./list.js");
const Stdlib__Seq = require("./seq.js");

const Empty = /* @__PURE__ */ Caml_exceptions.create("Stdlib.Stack.Empty");

function create(param) {
  return {
    c: /* [] */ 0,
    len: 0
  };
}

function clear(s) {
  s.c = /* [] */ 0;
  s.len = 0;
}

function copy(s) {
  return {
    c: s.c,
    len: s.len
  };
}

function push(x, s) {
  s.c = {
    hd: x,
    tl: s.c
  };
  s.len = s.len + 1 | 0;
}

function pop(s) {
  const match = s.c;
  if (match) {
    s.c = match.tl;
    s.len = s.len - 1 | 0;
    return match.hd;
  }
  throw new Caml_js_exceptions.MelangeError(Empty, {
      MEL_EXN_ID: Empty
    });
}

function pop_opt(s) {
  const match = s.c;
  if (match) {
    s.c = match.tl;
    s.len = s.len - 1 | 0;
    return Caml_option.some(match.hd);
  }
  
}

function drop(s) {
  const match = s.c;
  if (match) {
    s.c = match.tl;
    s.len = s.len - 1 | 0;
    return;
  }
  throw new Caml_js_exceptions.MelangeError(Empty, {
      MEL_EXN_ID: Empty
    });
}

function top(s) {
  const match = s.c;
  if (match) {
    return match.hd;
  }
  throw new Caml_js_exceptions.MelangeError(Empty, {
      MEL_EXN_ID: Empty
    });
}

function top_opt(s) {
  const match = s.c;
  if (match) {
    return Caml_option.some(match.hd);
  }
  
}

function is_empty(s) {
  return Caml_obj.caml_equal(s.c, /* [] */ 0);
}

function length(s) {
  return s.len;
}

function iter(f, s) {
  Stdlib__List.iter(f, s.c);
}

function fold(f, acc, s) {
  return Stdlib__List.fold_left(f, acc, s.c);
}

function to_seq(s) {
  return Stdlib__List.to_seq(s.c);
}

function add_seq(q, i) {
  Stdlib__Seq.iter((function (x) {
    push(x, q);
  }), i);
}

function of_seq(g) {
  const s = {
    c: /* [] */ 0,
    len: 0
  };
  add_seq(s, g);
  return s;
}

module.exports = {
  Empty,
  create,
  push,
  pop,
  pop_opt,
  drop,
  top,
  top_opt,
  clear,
  copy,
  is_empty,
  length,
  iter,
  fold,
  to_seq,
  add_seq,
  of_seq,
}
/* No side effect */
