// melange/atomic.mjs
// Generated by Melange

import * as Caml_external_polyfill from "melange.js/caml_external_polyfill.mjs";
import * as CamlinternalAtomic from "./camlinternalAtomic.mjs";

function set(t, v) {
  Caml_external_polyfill.resolve("caml_atomic_exchange_field")(t[0], t[1], v);
}

function incr(t) {
  Caml_external_polyfill.resolve("caml_atomic_fetch_add_field")(t[0], t[1], 1);
}

function decr(t) {
  Caml_external_polyfill.resolve("caml_atomic_fetch_add_field")(t[0], t[1], -1);
}

const Loc = {
  set: set,
  incr: incr,
  decr: decr
};

const make = CamlinternalAtomic.make;

const get = CamlinternalAtomic.get;

const set$1 = CamlinternalAtomic.set;

const exchange = CamlinternalAtomic.exchange;

const compare_and_set = CamlinternalAtomic.compare_and_set;

const fetch_and_add = CamlinternalAtomic.fetch_and_add;

const incr$1 = CamlinternalAtomic.incr;

const decr$1 = CamlinternalAtomic.decr;

export {
  make,
  get,
  set$1 as set,
  exchange,
  compare_and_set,
  fetch_and_add,
  incr$1 as incr,
  decr$1 as decr,
  Loc,
}
/* No side effect */
