// melange/bigarray.mjs
// Generated by Melange

import * as Caml_array from "melange.js/caml_array.mjs";
import * as Caml_bigarray from "melange.js/caml_bigarray.mjs";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";
import * as Curry from "melange.js/curry.mjs";
import * as Stdlib__Array from "./array.mjs";
import * as Stdlib__Sys from "./sys.mjs";

function kind_size_in_bytes(param) {
  switch (param) {
    case /* Float32 */ 0 :
    case /* Int32 */ 6 :
      return 4;
    case /* Int */ 8 :
    case /* Nativeint */ 9 :
      return Stdlib__Sys.word_size / 8 | 0;
    case /* Complex64 */ 11 :
      return 16;
    case /* Int8_signed */ 2 :
    case /* Int8_unsigned */ 3 :
    case /* Char */ 12 :
      return 1;
    case /* Int16_signed */ 4 :
    case /* Int16_unsigned */ 5 :
    case /* Float16 */ 13 :
      return 2;
    default:
      return 8;
  }
}

function cloop(arr, idx, f, col, max) {
  if (col === idx.length) {
    return Caml_bigarray.caml_ba_set_generic(arr, idx, Curry._1(f, idx));
  }
  for (let j = 0, j_finish = Caml_array.get(max, col); j < j_finish; ++j) {
    Caml_array.set(idx, col, j);
    cloop(arr, idx, f, col + 1 | 0, max);
  }
}

function floop(arr, idx, f, col, max) {
  if (col < 0) {
    return Caml_bigarray.caml_ba_set_generic(arr, idx, Curry._1(f, idx));
  }
  for (let j = 1, j_finish = Caml_array.get(max, col); j <= j_finish; ++j) {
    Caml_array.set(idx, col, j);
    floop(arr, idx, f, col - 1 | 0, max);
  }
}

function init(kind, layout, dims, f) {
  const arr = Caml_bigarray.caml_ba_create(kind, layout, dims);
  const dlen = dims.length;
  if (layout === /* C_layout */ 0) {
    cloop(arr, Caml_array.make(dlen, 0), f, 0, dims);
    return arr;
  }
  floop(arr, Caml_array.make(dlen, 1), f, dlen - 1 | 0, dims);
  return arr;
}

function dims(a) {
  const n = Caml_bigarray.caml_ba_num_dims(a);
  const d = Caml_array.make(n, 0);
  for (let i = 0; i < n; ++i) {
    Caml_array.set(d, i, Caml_bigarray.caml_ba_dim(a, i));
  }
  return d;
}

function size_in_bytes(arr) {
  return Math.imul(kind_size_in_bytes(Caml_bigarray.caml_ba_kind(arr)), Stdlib__Array.fold_left((function (prim0, prim1) {
    return Math.imul(prim0, prim1);
  }), 1, dims(arr)));
}

const Genarray = {
  cloop: cloop,
  floop: floop,
  init: init,
  dims: dims,
  size_in_bytes: size_in_bytes
};

function create(kind, layout) {
  return Caml_bigarray.caml_ba_create(kind, layout, []);
}

function get(arr) {
  return Caml_bigarray.caml_ba_get_generic(arr, []);
}

function set(arr) {
  const partial_arg = [];
  return function (param) {
    return Caml_bigarray.caml_ba_set_generic(arr, partial_arg, param);
  };
}

function size_in_bytes$1(arr) {
  return kind_size_in_bytes(Caml_bigarray.caml_ba_kind(arr));
}

function of_value(kind, layout, v) {
  const a = Caml_bigarray.caml_ba_create(kind, layout, []);
  set(a)(v);
  return a;
}

const Array0 = {
  create: create,
  get: get,
  set: set,
  size_in_bytes: size_in_bytes$1,
  of_value: of_value,
  init: of_value
};

function create$1(kind, layout, dim) {
  return Caml_bigarray.caml_ba_create(kind, layout, [dim]);
}

function size_in_bytes$2(arr) {
  return Math.imul(kind_size_in_bytes(Caml_bigarray.caml_ba_kind(arr)), Caml_bigarray.caml_ba_dim_1(arr));
}

function slice(a, n) {
  Caml_bigarray.caml_ba_layout(a);
  return Caml_bigarray.caml_ba_slice(a, [n]);
}

function c_init(arr, dim, f) {
  for (let i = 0; i < dim; ++i) {
    Caml_bigarray.caml_ba_set_1(arr, i, Curry._1(f, i));
  }
}

function fortran_init(arr, dim, f) {
  for (let i = 1; i <= dim; ++i) {
    Caml_bigarray.caml_ba_set_1(arr, i, Curry._1(f, i));
  }
}

function init$1(kind, layout, dim, f) {
  const arr = create$1(kind, layout, dim);
  if (layout === /* C_layout */ 0) {
    c_init(arr, dim, f);
    return arr;
  }
  fortran_init(arr, dim, f);
  return arr;
}

function of_array(kind, layout, data) {
  const ba = create$1(kind, layout, data.length);
  let ofs;
  ofs = layout === /* C_layout */ 0 ? 0 : 1;
  for (let i = 0, i_finish = data.length; i < i_finish; ++i) {
    Caml_bigarray.caml_ba_set_1(ba, i + ofs | 0, Caml_array.get(data, i));
  }
  return ba;
}

const Array1 = {
  create: create$1,
  size_in_bytes: size_in_bytes$2,
  slice: slice,
  c_init: c_init,
  fortran_init: fortran_init,
  init: init$1,
  of_array: of_array
};

function create$2(kind, layout, dim1, dim2) {
  return Caml_bigarray.caml_ba_create(kind, layout, [
    dim1,
    dim2
  ]);
}

function size_in_bytes$3(arr) {
  return Math.imul(Math.imul(kind_size_in_bytes(Caml_bigarray.caml_ba_kind(arr)), Caml_bigarray.caml_ba_dim_1(arr)), Caml_bigarray.caml_ba_dim_2(arr));
}

function slice_left(a, n) {
  return Caml_bigarray.caml_ba_slice(a, [n]);
}

function slice_right(a, n) {
  return Caml_bigarray.caml_ba_slice(a, [n]);
}

function c_init$1(arr, dim1, dim2, f) {
  for (let i = 0; i < dim1; ++i) {
    for (let j = 0; j < dim2; ++j) {
      Caml_bigarray.caml_ba_set_2(arr, i, j, Curry._2(f, i, j));
    }
  }
}

function fortran_init$1(arr, dim1, dim2, f) {
  for (let j = 1; j <= dim2; ++j) {
    for (let i = 1; i <= dim1; ++i) {
      Caml_bigarray.caml_ba_set_2(arr, i, j, Curry._2(f, i, j));
    }
  }
}

function init$2(kind, layout, dim1, dim2, f) {
  const arr = create$2(kind, layout, dim1, dim2);
  if (layout === /* C_layout */ 0) {
    c_init$1(arr, dim1, dim2, f);
    return arr;
  }
  fortran_init$1(arr, dim1, dim2, f);
  return arr;
}

function of_array$1(kind, layout, data) {
  const dim1 = data.length;
  const dim2 = dim1 === 0 ? 0 : Caml_array.get(data, 0).length;
  const ba = create$2(kind, layout, dim1, dim2);
  let ofs;
  ofs = layout === /* C_layout */ 0 ? 0 : 1;
  for (let i = 0; i < dim1; ++i) {
    const row = Caml_array.get(data, i);
    if (row.length !== dim2) {
      throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
          MEL_EXN_ID: "Invalid_argument",
          _1: "Bigarray.Array2.of_array: non-rectangular data"
        });
    }
    for (let j = 0; j < dim2; ++j) {
      Caml_bigarray.caml_ba_set_2(ba, i + ofs | 0, j + ofs | 0, Caml_array.get(row, j));
    }
  }
  return ba;
}

const Array2 = {
  create: create$2,
  size_in_bytes: size_in_bytes$3,
  slice_left: slice_left,
  slice_right: slice_right,
  c_init: c_init$1,
  fortran_init: fortran_init$1,
  init: init$2,
  of_array: of_array$1
};

function create$3(kind, layout, dim1, dim2, dim3) {
  return Caml_bigarray.caml_ba_create(kind, layout, [
    dim1,
    dim2,
    dim3
  ]);
}

function size_in_bytes$4(arr) {
  return Math.imul(Math.imul(Math.imul(kind_size_in_bytes(Caml_bigarray.caml_ba_kind(arr)), Caml_bigarray.caml_ba_dim_1(arr)), Caml_bigarray.caml_ba_dim_2(arr)), Caml_bigarray.caml_ba_dim_3(arr));
}

function slice_left_1(a, n, m) {
  return Caml_bigarray.caml_ba_slice(a, [
    n,
    m
  ]);
}

function slice_right_1(a, n, m) {
  return Caml_bigarray.caml_ba_slice(a, [
    n,
    m
  ]);
}

function slice_left_2(a, n) {
  return Caml_bigarray.caml_ba_slice(a, [n]);
}

function slice_right_2(a, n) {
  return Caml_bigarray.caml_ba_slice(a, [n]);
}

function c_init$2(arr, dim1, dim2, dim3, f) {
  for (let i = 0; i < dim1; ++i) {
    for (let j = 0; j < dim2; ++j) {
      for (let k = 0; k < dim3; ++k) {
        Caml_bigarray.caml_ba_set_3(arr, i, j, k, Curry._3(f, i, j, k));
      }
    }
  }
}

function fortran_init$2(arr, dim1, dim2, dim3, f) {
  for (let k = 1; k <= dim3; ++k) {
    for (let j = 1; j <= dim2; ++j) {
      for (let i = 1; i <= dim1; ++i) {
        Caml_bigarray.caml_ba_set_3(arr, i, j, k, Curry._3(f, i, j, k));
      }
    }
  }
}

function init$3(kind, layout, dim1, dim2, dim3, f) {
  const arr = create$3(kind, layout, dim1, dim2, dim3);
  if (layout === /* C_layout */ 0) {
    c_init$2(arr, dim1, dim2, dim3, f);
    return arr;
  }
  fortran_init$2(arr, dim1, dim2, dim3, f);
  return arr;
}

function of_array$2(kind, layout, data) {
  const dim1 = data.length;
  const dim2 = dim1 === 0 ? 0 : Caml_array.get(data, 0).length;
  const dim3 = dim2 === 0 ? 0 : Caml_array.get(Caml_array.get(data, 0), 0).length;
  const ba = create$3(kind, layout, dim1, dim2, dim3);
  let ofs;
  ofs = layout === /* C_layout */ 0 ? 0 : 1;
  for (let i = 0; i < dim1; ++i) {
    const row = Caml_array.get(data, i);
    if (row.length !== dim2) {
      throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
          MEL_EXN_ID: "Invalid_argument",
          _1: "Bigarray.Array3.of_array: non-cubic data"
        });
    }
    for (let j = 0; j < dim2; ++j) {
      const col = Caml_array.get(row, j);
      if (col.length !== dim3) {
        throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
            MEL_EXN_ID: "Invalid_argument",
            _1: "Bigarray.Array3.of_array: non-cubic data"
          });
      }
      for (let k = 0; k < dim3; ++k) {
        Caml_bigarray.caml_ba_set_3(ba, i + ofs | 0, j + ofs | 0, k + ofs | 0, Caml_array.get(col, k));
      }
    }
  }
  return ba;
}

const Array3 = {
  create: create$3,
  size_in_bytes: size_in_bytes$4,
  slice_left_1: slice_left_1,
  slice_right_1: slice_right_1,
  slice_left_2: slice_left_2,
  slice_right_2: slice_right_2,
  c_init: c_init$2,
  fortran_init: fortran_init$2,
  init: init$3,
  of_array: of_array$2
};

function array0_of_genarray(a) {
  if (Caml_bigarray.caml_ba_num_dims(a) === 0) {
    return a;
  }
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: "Bigarray.array0_of_genarray"
    });
}

function array1_of_genarray(a) {
  if (Caml_bigarray.caml_ba_num_dims(a) === 1) {
    return a;
  }
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: "Bigarray.array1_of_genarray"
    });
}

function array2_of_genarray(a) {
  if (Caml_bigarray.caml_ba_num_dims(a) === 2) {
    return a;
  }
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: "Bigarray.array2_of_genarray"
    });
}

function array3_of_genarray(a) {
  if (Caml_bigarray.caml_ba_num_dims(a) === 3) {
    return a;
  }
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: "Bigarray.array3_of_genarray"
    });
}

function reshape_0(a) {
  return Caml_bigarray.caml_ba_reshape(a, []);
}

function reshape_1(a, dim1) {
  return Caml_bigarray.caml_ba_reshape(a, [dim1]);
}

function reshape_2(a, dim1, dim2) {
  return Caml_bigarray.caml_ba_reshape(a, [
    dim1,
    dim2
  ]);
}

function reshape_3(a, dim1, dim2, dim3) {
  return Caml_bigarray.caml_ba_reshape(a, [
    dim1,
    dim2,
    dim3
  ]);
}

const float16 = /* Float16 */ 13;

const float32 = /* Float32 */ 0;

const float64 = /* Float64 */ 1;

const int8_signed = /* Int8_signed */ 2;

const int8_unsigned = /* Int8_unsigned */ 3;

const int16_signed = /* Int16_signed */ 4;

const int16_unsigned = /* Int16_unsigned */ 5;

const int32 = /* Int32 */ 6;

const int64 = /* Int64 */ 7;

const $$int = /* Int */ 8;

const nativeint = /* Nativeint */ 9;

const complex32 = /* Complex32 */ 10;

const complex64 = /* Complex64 */ 11;

const $$char = /* Char */ 12;

const c_layout = /* C_layout */ 0;

const fortran_layout = /* Fortran_layout */ 1;

export {
  float16,
  float32,
  float64,
  int8_signed,
  int8_unsigned,
  int16_signed,
  int16_unsigned,
  int32,
  int64,
  $$int,
  nativeint,
  complex32,
  complex64,
  $$char,
  kind_size_in_bytes,
  c_layout,
  fortran_layout,
  Genarray,
  Array0,
  Array1,
  Array2,
  Array3,
  array0_of_genarray,
  array1_of_genarray,
  array2_of_genarray,
  array3_of_genarray,
  reshape_0,
  reshape_1,
  reshape_2,
  reshape_3,
}
/* No side effect */
