// melange/bytesLabels.mjs
// Generated by Melange

import * as Stdlib__Bytes from "./bytes.mjs";

const make = Stdlib__Bytes.make;

const init = Stdlib__Bytes.init;

const empty = Stdlib__Bytes.empty;

const copy = Stdlib__Bytes.copy;

const of_string = Stdlib__Bytes.of_string;

const to_string = Stdlib__Bytes.to_string;

const sub = Stdlib__Bytes.sub;

const sub_string = Stdlib__Bytes.sub_string;

const extend = Stdlib__Bytes.extend;

const fill = Stdlib__Bytes.fill;

const blit = Stdlib__Bytes.blit;

const blit_string = Stdlib__Bytes.blit_string;

const concat = Stdlib__Bytes.concat;

const cat = Stdlib__Bytes.cat;

const iter = Stdlib__Bytes.iter;

const iteri = Stdlib__Bytes.iteri;

const map = Stdlib__Bytes.map;

const mapi = Stdlib__Bytes.mapi;

const fold_left = Stdlib__Bytes.fold_left;

const fold_right = Stdlib__Bytes.fold_right;

const for_all = Stdlib__Bytes.for_all;

const exists = Stdlib__Bytes.exists;

const trim = Stdlib__Bytes.trim;

const escaped = Stdlib__Bytes.escaped;

const index = Stdlib__Bytes.index;

const index_opt = Stdlib__Bytes.index_opt;

const rindex = Stdlib__Bytes.rindex;

const rindex_opt = Stdlib__Bytes.rindex_opt;

const index_from = Stdlib__Bytes.index_from;

const index_from_opt = Stdlib__Bytes.index_from_opt;

const rindex_from = Stdlib__Bytes.rindex_from;

const rindex_from_opt = Stdlib__Bytes.rindex_from_opt;

const contains = Stdlib__Bytes.contains;

const contains_from = Stdlib__Bytes.contains_from;

const rcontains_from = Stdlib__Bytes.rcontains_from;

const uppercase_ascii = Stdlib__Bytes.uppercase_ascii;

const lowercase_ascii = Stdlib__Bytes.lowercase_ascii;

const capitalize_ascii = Stdlib__Bytes.capitalize_ascii;

const uncapitalize_ascii = Stdlib__Bytes.uncapitalize_ascii;

const compare = Stdlib__Bytes.compare;

const equal = Stdlib__Bytes.equal;

const starts_with = Stdlib__Bytes.starts_with;

const ends_with = Stdlib__Bytes.ends_with;

const unsafe_to_string = Stdlib__Bytes.unsafe_to_string;

const unsafe_of_string = Stdlib__Bytes.unsafe_of_string;

const split_on_char = Stdlib__Bytes.split_on_char;

const to_seq = Stdlib__Bytes.to_seq;

const to_seqi = Stdlib__Bytes.to_seqi;

const of_seq = Stdlib__Bytes.of_seq;

const get_utf_8_uchar = Stdlib__Bytes.get_utf_8_uchar;

const set_utf_8_uchar = Stdlib__Bytes.set_utf_8_uchar;

const is_valid_utf_8 = Stdlib__Bytes.is_valid_utf_8;

const get_utf_16be_uchar = Stdlib__Bytes.get_utf_16be_uchar;

const set_utf_16be_uchar = Stdlib__Bytes.set_utf_16be_uchar;

const is_valid_utf_16be = Stdlib__Bytes.is_valid_utf_16be;

const get_utf_16le_uchar = Stdlib__Bytes.get_utf_16le_uchar;

const set_utf_16le_uchar = Stdlib__Bytes.set_utf_16le_uchar;

const is_valid_utf_16le = Stdlib__Bytes.is_valid_utf_16le;

const get_uint8 = Stdlib__Bytes.get_uint8;

const get_int8 = Stdlib__Bytes.get_int8;

const get_uint16_ne = Stdlib__Bytes.get_uint16_ne;

const get_uint16_be = Stdlib__Bytes.get_uint16_be;

const get_uint16_le = Stdlib__Bytes.get_uint16_le;

const get_int16_ne = Stdlib__Bytes.get_int16_ne;

const get_int16_be = Stdlib__Bytes.get_int16_be;

const get_int16_le = Stdlib__Bytes.get_int16_le;

const get_int32_ne = Stdlib__Bytes.get_int32_ne;

const get_int32_be = Stdlib__Bytes.get_int32_be;

const get_int32_le = Stdlib__Bytes.get_int32_le;

const get_int64_ne = Stdlib__Bytes.get_int64_ne;

const get_int64_be = Stdlib__Bytes.get_int64_be;

const get_int64_le = Stdlib__Bytes.get_int64_le;

const set_uint8 = Stdlib__Bytes.set_uint8;

const set_int8 = Stdlib__Bytes.set_int8;

const set_uint16_ne = Stdlib__Bytes.set_uint16_ne;

const set_uint16_be = Stdlib__Bytes.set_uint16_be;

const set_uint16_le = Stdlib__Bytes.set_uint16_le;

const set_int16_ne = Stdlib__Bytes.set_int16_ne;

const set_int16_be = Stdlib__Bytes.set_int16_be;

const set_int16_le = Stdlib__Bytes.set_int16_le;

const set_int32_ne = Stdlib__Bytes.set_int32_ne;

const set_int32_be = Stdlib__Bytes.set_int32_be;

const set_int32_le = Stdlib__Bytes.set_int32_le;

const set_int64_ne = Stdlib__Bytes.set_int64_ne;

const set_int64_be = Stdlib__Bytes.set_int64_be;

const set_int64_le = Stdlib__Bytes.set_int64_le;

const unsafe_escape = Stdlib__Bytes.unsafe_escape;

export {
  make,
  init,
  empty,
  copy,
  of_string,
  to_string,
  sub,
  sub_string,
  extend,
  fill,
  blit,
  blit_string,
  concat,
  cat,
  iter,
  iteri,
  map,
  mapi,
  fold_left,
  fold_right,
  for_all,
  exists,
  trim,
  escaped,
  index,
  index_opt,
  rindex,
  rindex_opt,
  index_from,
  index_from_opt,
  rindex_from,
  rindex_from_opt,
  contains,
  contains_from,
  rcontains_from,
  uppercase_ascii,
  lowercase_ascii,
  capitalize_ascii,
  uncapitalize_ascii,
  compare,
  equal,
  starts_with,
  ends_with,
  unsafe_to_string,
  unsafe_of_string,
  split_on_char,
  to_seq,
  to_seqi,
  of_seq,
  get_utf_8_uchar,
  set_utf_8_uchar,
  is_valid_utf_8,
  get_utf_16be_uchar,
  set_utf_16be_uchar,
  is_valid_utf_16be,
  get_utf_16le_uchar,
  set_utf_16le_uchar,
  is_valid_utf_16le,
  get_uint8,
  get_int8,
  get_uint16_ne,
  get_uint16_be,
  get_uint16_le,
  get_int16_ne,
  get_int16_be,
  get_int16_le,
  get_int32_ne,
  get_int32_be,
  get_int32_le,
  get_int64_ne,
  get_int64_be,
  get_int64_le,
  set_uint8,
  set_int8,
  set_uint16_ne,
  set_uint16_be,
  set_uint16_le,
  set_int16_ne,
  set_int16_be,
  set_int16_le,
  set_int32_ne,
  set_int32_be,
  set_int32_le,
  set_int64_ne,
  set_int64_be,
  set_int64_le,
  unsafe_escape,
}
/* No side effect */
