// melange/camlinternalLazy.js
// Generated by Melange
'use strict';

const Caml_exceptions = require("melange.js/caml_exceptions.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Stdlib__Obj = require("./obj.js");

const Undefined = /* @__PURE__ */ Caml_exceptions.create("CamlinternalLazy.Undefined");

function is_val(l) {
  return l.LAZY_DONE;
}

function forward_with_closure(blk, closure) {
  const result = closure();
  blk.VAL = result;
  blk.LAZY_DONE = true;
  return result;
}

function raise_undefined() {
  throw new Caml_js_exceptions.MelangeError(Undefined, {
      MEL_EXN_ID: Undefined
    });
}

function force_lazy_block(blk) {
  const closure = blk.VAL;
  blk.VAL = raise_undefined;
  try {
    return forward_with_closure(blk, closure);
  }
  catch (e){
    blk.VAL = (function () {
      throw e;
    });
    throw e;
  }
}

function force_val_lazy_block(blk) {
  const closure = blk.VAL;
  blk.VAL = raise_undefined;
  return forward_with_closure(blk, closure);
}

function force(lzv) {
  const lzv$1 = lzv;
  if (lzv$1.LAZY_DONE) {
    return lzv$1.VAL;
  } else {
    return force_lazy_block(lzv$1);
  }
}

function force_val(lzv) {
  if (lzv.LAZY_DONE) {
    return lzv.VAL;
  } else {
    return force_val_lazy_block(lzv);
  }
}

function indirect(lzv) {
  const lzv$1 = lzv;
  const t = lzv$1.TAG;
  if (t === Stdlib__Obj.lazy_tag || t === Stdlib__Obj.forcing_tag) {
    return {
      LAZY_DONE: false,
      VAL: (function () {
        return force_lazy_block(lzv$1);
      })
    };
  } else {
    return lzv$1;
  }
}

module.exports = {
  Undefined,
  force_lazy_block,
  force_val_lazy_block,
  force,
  force_val,
  is_val,
  indirect,
}
/* No side effect */
