// melange/char.js
// Generated by Melange
'use strict';

const Caml_bytes = require("melange.js/caml_bytes.js");
const Caml_hash = require("melange.js/caml_hash.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Stdlib = require("./stdlib.js");

function chr(n) {
  if (n < 0 || n > 255) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "Char.chr"
      });
  }
  return n;
}

function escaped(c) {
  let exit = 0;
  if (c >= 40) {
    if (c === 92) {
      return "\\\\";
    }
    exit = c >= 127 ? 1 : 2;
  } else if (c >= 32) {
    if (c >= 39) {
      return "\\'";
    }
    exit = 2;
  } else if (c >= 14) {
    exit = 1;
  } else {
    switch (c) {
      case 8 : {
        return "\\b";
        }
      case 9 : {
        return "\\t";
        }
      case 10 : {
        return "\\n";
        }
      case 0 : {
        }
      case 1 : {
        }
      case 2 : {
        }
      case 3 : {
        }
      case 4 : {
        }
      case 5 : {
        }
      case 6 : {
        }
      case 7 : {
        }
      case 11 : {
        }
      case 12 : {
        exit = 1;
        break;
        }
      case 13 : {
        return "\\r";
        }
    }
  }
  switch (exit) {
    case 1 : {
      const s = [
        0,
        0,
        0,
        0
      ];
      s[0] = /* '\\' */92;
      s[1] = 48 + (c / 100 | 0) | 0;
      s[2] = 48 + (c / 10 | 0) % 10 | 0;
      s[3] = 48 + c % 10 | 0;
      return Caml_bytes.bytes_to_string(s);
      }
    case 2 : {
      const s$1 = [0];
      s$1[0] = c;
      return Caml_bytes.bytes_to_string(s$1);
      }
  }
}

function lowercase_ascii(c) {
  if (c > 90 || c < 65) {
    return c;
  } else {
    return c + 32 | 0;
  }
}

function uppercase_ascii(c) {
  if (c > 122 || c < 97) {
    return c;
  } else {
    return c - 32 | 0;
  }
}

function compare(c1, c2) {
  return c1 - c2 | 0;
}

function equal(c1, c2) {
  return (c1 - c2 | 0) === 0;
}

function seeded_hash(seed, x) {
  return Caml_hash.caml_hash(10, 100, seed, x);
}

function hash(x) {
  return Caml_hash.caml_hash(10, 100, 0, x);
}

function is_valid(param) {
  return param < 128;
}

function is_upper(param) {
  return !(param > 90 || param < 65);
}

function is_lower(param) {
  return !(param > 122 || param < 97);
}

function is_letter(param) {
  if (param >= 91) {
    return !(param > 122 || param < 97);
  } else {
    return param >= 65;
  }
}

function is_alphanum(param) {
  if (param > 90 || param < 48) {
    return !(param > 122 || param < 97);
  } else {
    return param > 64 || param < 58;
  }
}

function is_white(param) {
  if (param >= 14) {
    return param === 32;
  } else {
    return param >= 9;
  }
}

function is_blank(param) {
  if (param !== 9) {
    return param === 32;
  } else {
    return true;
  }
}

function is_graphic(param) {
  return !(param > 126 || param < 33);
}

function is_print(param) {
  return !(param > 126 || param < 32);
}

function is_control(param) {
  if (param !== 127) {
    return param < 32;
  } else {
    return true;
  }
}

function is_digit(param) {
  return !(param > 57 || param < 48);
}

function digit_to_int(c) {
  if (!(c > 57 || c < 48)) {
    return c - 48 | 0;
  }
  const s = escaped(c) + ": not a decimal digit";
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: s
    });
}

function digit_of_int(n) {
  return 48 + Stdlib.abs(n % 10) | 0;
}

function is_hex_digit(param) {
  if (param > 70 || param < 48) {
    return !(param > 102 || param < 97);
  } else {
    return param > 64 || param < 58;
  }
}

function hex_digit_to_int(c) {
  if (c >= 65) {
    if (c >= 97) {
      if (c < 103) {
        return (10 + c | 0) - 97 | 0;
      }
      
    } else if (c < 71) {
      return (10 + c | 0) - 65 | 0;
    }
    
  } else if (!(c > 57 || c < 48)) {
    return c - 48 | 0;
  }
  const s = escaped(c) + ": not a hexadecimal digit";
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: s
    });
}

function lower_hex_digit_of_int(n) {
  const d = Stdlib.abs(n % 16);
  if (d < 10) {
    return 48 + d | 0;
  } else {
    return 87 + d | 0;
  }
}

function upper_hex_digit_of_int(n) {
  const d = Stdlib.abs(n % 16);
  if (d < 10) {
    return 48 + d | 0;
  } else {
    return 55 + d | 0;
  }
}

const Ascii = {
  min: /* '\000' */0,
  max: /* '\127' */127,
  is_valid: is_valid,
  is_upper: is_upper,
  is_lower: is_lower,
  is_letter: is_letter,
  is_alphanum: is_alphanum,
  is_white: is_white,
  is_blank: is_blank,
  is_graphic: is_graphic,
  is_print: is_print,
  is_control: is_control,
  is_digit: is_digit,
  digit_to_int: digit_to_int,
  digit_of_int: digit_of_int,
  is_hex_digit: is_hex_digit,
  hex_digit_to_int: hex_digit_to_int,
  lower_hex_digit_of_int: lower_hex_digit_of_int,
  upper_hex_digit_of_int: upper_hex_digit_of_int,
  uppercase: uppercase_ascii,
  lowercase: lowercase_ascii
};

module.exports = {
  chr,
  escaped,
  compare,
  equal,
  Ascii,
  lowercase_ascii,
  uppercase_ascii,
  seeded_hash,
  hash,
}
/* No side effect */
