// melange/fun.js
// Generated by Melange
'use strict';

const Caml_exceptions = require("melange.js/caml_exceptions.js");
const Caml_external_polyfill = require("melange.js/caml_external_polyfill.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Curry = require("melange.js/curry.js");
const Stdlib__Printexc = require("./printexc.js");

function $$const(c, param) {
  return c;
}

function compose(f, g, x) {
  return Curry._1(f, Curry._1(g, x));
}

function flip(f, x, y) {
  return Curry._2(f, y, x);
}

function negate(p, v) {
  return !Curry._1(p, v);
}

const Finally_raised = /* @__PURE__ */ Caml_exceptions.create("Stdlib.Fun.Finally_raised");

Stdlib__Printexc.register_printer(function (exn) {
  if (exn.MEL_EXN_ID === Finally_raised) {
    return "Fun.Finally_raised: " + Stdlib__Printexc.to_string(exn._1);
  }
  
});

function protect($$finally, work) {
  const finally_no_exn = function (param) {
    try {
      return Curry._1($$finally, undefined);
    }
    catch (raw_e){
      const e = Caml_js_exceptions.internalToOCamlException(raw_e);
      let bt;
      const exn = new Caml_js_exceptions.MelangeError(Finally_raised, {
          MEL_EXN_ID: Finally_raised,
          _1: e
        });
      Caml_external_polyfill.resolve("caml_restore_raw_backtrace")(exn, bt);
      throw exn;
    }
  };
  let result;
  try {
    result = Curry._1(work, undefined);
  }
  catch (raw_work_exn){
    const work_exn = Caml_js_exceptions.internalToOCamlException(raw_work_exn);
    let work_bt;
    finally_no_exn();
    Caml_external_polyfill.resolve("caml_restore_raw_backtrace")(work_exn, work_bt);
    throw work_exn;
  }
  finally_no_exn();
  return result;
}

module.exports = {
  $$const,
  compose,
  flip,
  negate,
  protect,
  Finally_raised,
}
/*  Not a pure module */
