// melange/int.mjs
// Generated by Melange

import * as Caml from "melange.js/caml.mjs";
import * as Caml_format from "melange.js/caml_format.mjs";
import * as Caml_hash from "melange.js/caml_hash.mjs";
import * as Caml_int64 from "melange.js/caml_int64.mjs";
import * as Stdlib__Sys from "./sys.mjs";

function abs(x) {
  if (x >= 0) {
    return x;
  } else {
    return -x | 0;
  }
}

const min_int = -2147483648;

function lognot(x) {
  return x ^ -1;
}

function equal(prim0, prim1) {
  return prim0 === prim1;
}

const compare = Caml.caml_int_compare;

function min(x, y) {
  if (x <= y) {
    return x;
  } else {
    return y;
  }
}

function max(x, y) {
  if (x >= y) {
    return x;
  } else {
    return y;
  }
}

function leading_zeros(x) {
  let x$1 = x;
  let n = Stdlib__Sys.int_size;
  if (Stdlib__Sys.int_size > 32) {
    const y = (x$1 >>> 32);
    if (y !== 0) {
      n = n - 32 | 0;
      x$1 = y;
    }
    
  }
  const y$1 = (x$1 >>> 16);
  if (y$1 !== 0) {
    n = n - 16 | 0;
    x$1 = y$1;
  }
  const y$2 = (x$1 >>> 8);
  if (y$2 !== 0) {
    n = n - 8 | 0;
    x$1 = y$2;
  }
  const y$3 = (x$1 >>> 4);
  if (y$3 !== 0) {
    n = n - 4 | 0;
    x$1 = y$3;
  }
  const y$4 = (x$1 >>> 2);
  if (y$4 !== 0) {
    n = n - 2 | 0;
    x$1 = y$4;
  }
  const y$5 = (x$1 >>> 1);
  if (y$5 !== 0) {
    return n - 2 | 0;
  } else {
    return n - x$1 | 0;
  }
}

function unsigned_bitsize(x) {
  return Stdlib__Sys.int_size - leading_zeros(x) | 0;
}

function leading_sign_bits(x) {
  if (x >= 0) {
    return leading_zeros(x) - 1 | 0;
  } else {
    return leading_zeros(x ^ -1) - 1 | 0;
  }
}

function signed_bitsize(x) {
  return Stdlib__Sys.int_size - leading_sign_bits(x) | 0;
}

function trailing_zeros(x) {
  if (x === 0) {
    return Stdlib__Sys.int_size;
  }
  let x$1 = x;
  let n = Stdlib__Sys.int_size - 1 | 0;
  if (Stdlib__Sys.int_size > 32) {
    const y = (x$1 << 32);
    if (y !== 0) {
      n = n - 32 | 0;
      x$1 = y;
    }
    
  }
  const y$1 = (x$1 << 16);
  if (y$1 !== 0) {
    n = n - 16 | 0;
    x$1 = y$1;
  }
  const y$2 = (x$1 << 8);
  if (y$2 !== 0) {
    n = n - 8 | 0;
    x$1 = y$2;
  }
  const y$3 = (x$1 << 4);
  if (y$3 !== 0) {
    n = n - 4 | 0;
    x$1 = y$3;
  }
  const y$4 = (x$1 << 2);
  if (y$4 !== 0) {
    n = n - 2 | 0;
    x$1 = y$4;
  }
  const y$5 = (x$1 << 1);
  if (y$5 !== 0) {
    return n - 1 | 0;
  } else {
    return n;
  }
}

const cst1 = Caml_int64.to_int32([
  1431655765,
  1431655765
]);

const cst2 = Caml_int64.to_int32([
  858993459,
  858993459
]);

const cst3 = Caml_int64.to_int32([
  252645135,
  252645135
]);

function popcount(x) {
  const x$1 = x - ((x >>> 1) & cst1) | 0;
  const x$2 = (x$1 & cst2) + ((x$1 >>> 2) & cst2) | 0;
  const x$3 = x$2 + (x$2 >>> 4) & cst3;
  const x$4 = x$3 + (x$3 >>> 8) | 0;
  const x$5 = x$4 + (x$4 >>> 16) | 0;
  if (Stdlib__Sys.int_size <= 32) {
    return x$5 & 63;
  }
  const x$6 = x$5 + (x$5 >>> 32) | 0;
  return x$6 & 127;
}

function to_string(x) {
  return Caml_format.caml_format_int("%d", x);
}

function seeded_hash(seed, x) {
  return Caml_hash.caml_hash(10, 100, seed, x);
}

function hash(x) {
  return Caml_hash.caml_hash(10, 100, 0, x);
}

function fdiv(n, d) {
  const q = n / d | 0;
  if ((n ^ d) >= 0 || n === Math.imul(q, d)) {
    return q;
  } else {
    return q - 1 | 0;
  }
}

function cdiv(n, d) {
  const q = n / d | 0;
  if ((n ^ d) < 0 || n === Math.imul(q, d)) {
    return q;
  } else {
    return q + 1 | 0;
  }
}

function erem(n, d) {
  const r = n % d;
  if (r >= 0) {
    return r;
  } else if (d >= 0) {
    return r + d | 0;
  } else {
    return r - d | 0;
  }
}

function ediv(n, d) {
  const q = n / d | 0;
  const r = n - Math.imul(q, d) | 0;
  if (r >= 0) {
    return q;
  } else if (d >= 0) {
    return q - 1 | 0;
  } else {
    return q + 1 | 0;
  }
}

const zero = 0;

const one = 1;

const minus_one = -1;

const max_int = 2147483647;

export {
  zero,
  one,
  minus_one,
  fdiv,
  cdiv,
  ediv,
  erem,
  abs,
  max_int,
  min_int,
  lognot,
  equal,
  compare,
  min,
  max,
  popcount,
  unsigned_bitsize,
  signed_bitsize,
  leading_zeros,
  leading_sign_bits,
  trailing_zeros,
  to_string,
  seeded_hash,
  hash,
}
/* No side effect */
