// melange/int32.mjs
// Generated by Melange

import * as Caml from "melange.js/caml.mjs";
import * as Caml_format from "melange.js/caml_format.mjs";
import * as Caml_hash from "melange.js/caml_hash.mjs";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";
import * as Stdlib from "./stdlib.mjs";
import * as Stdlib__Sys from "./sys.mjs";

function succ(n) {
  return n + 1 | 0;
}

function pred(n) {
  return n - 1 | 0;
}

function abs(n) {
  if (n >= 0) {
    return n;
  } else {
    return -n | 0;
  }
}

function lognot(n) {
  return n ^ -1;
}

let unsigned_to_int;

if (Stdlib__Sys.word_size !== 32) {
  if (Stdlib__Sys.word_size !== 64) {
    throw new Caml_js_exceptions.MelangeError("Assert_failure", {
        MEL_EXN_ID: "Assert_failure",
        _1: [
          "jscomp/stdlib/int32.ml",
          69,
          6
        ]
      });
  }
  unsigned_to_int = (function (n) {
    return n & -1;
  });
} else {
  unsigned_to_int = (function (n) {
    if (n >= 0 && n <= Stdlib.max_int) {
      return n;
    }
    
  });
}

function to_string(n) {
  return Caml_format.caml_int32_format("%d", n);
}

function of_string_opt(s) {
  try {
    return Caml_format.caml_int32_of_string(s);
  }
  catch (raw_exn){
    const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.MEL_EXN_ID === Stdlib.Failure) {
      return;
    }
    throw exn;
  }
}

const compare = Caml.caml_int_compare;

function equal(x, y) {
  return x === y;
}

function unsigned_compare(n, m) {
  return Caml.caml_int_compare(n - -2147483648 | 0, m - -2147483648 | 0);
}

function unsigned_lt(n, m) {
  return (n - -2147483648 | 0) < (m - -2147483648 | 0);
}

function min(x, y) {
  if (x <= y) {
    return x;
  } else {
    return y;
  }
}

function max(x, y) {
  if (x >= y) {
    return x;
  } else {
    return y;
  }
}

function unsigned_div(n, d) {
  if (d < 0) {
    if (unsigned_lt(n, d)) {
      return 0;
    } else {
      return 1;
    }
  }
  const q = (((n >>> 1) / d | 0) << 1);
  const r = n - Math.imul(q, d) | 0;
  if (unsigned_lt(r, d)) {
    return q;
  } else {
    return q + 1 | 0;
  }
}

function unsigned_rem(n, d) {
  return n - Math.imul(unsigned_div(n, d), d) | 0;
}

function fdiv(n, d) {
  const q = n / d | 0;
  if ((n ^ d) >= 0 || n === Math.imul(q, d)) {
    return q;
  } else {
    return q - 1 | 0;
  }
}

function cdiv(n, d) {
  const q = n / d | 0;
  if ((n ^ d) < 0 || n === Math.imul(q, d)) {
    return q;
  } else {
    return q + 1 | 0;
  }
}

function erem(n, d) {
  const r = n % d;
  if (r >= 0) {
    return r;
  } else if (d >= 0) {
    return r + d | 0;
  } else {
    return r - d | 0;
  }
}

function ediv(n, d) {
  const q = n / d | 0;
  const r = n - Math.imul(q, d) | 0;
  if (r >= 0) {
    return q;
  } else if (d >= 0) {
    return q - 1 | 0;
  } else {
    return q + 1 | 0;
  }
}

function leading_zeros(x) {
  let x$1 = x;
  let n = 32;
  const y = (x$1 >>> 16);
  if (y !== 0) {
    n = n - 16 | 0;
    x$1 = y;
  }
  const y$1 = (x$1 >>> 8);
  if (y$1 !== 0) {
    n = n - 8 | 0;
    x$1 = y$1;
  }
  const y$2 = (x$1 >>> 4);
  if (y$2 !== 0) {
    n = n - 4 | 0;
    x$1 = y$2;
  }
  const y$3 = (x$1 >>> 2);
  if (y$3 !== 0) {
    n = n - 2 | 0;
    x$1 = y$3;
  }
  const y$4 = (x$1 >>> 1);
  if (y$4 !== 0) {
    return n - 2 | 0;
  } else {
    return n - x$1 | 0;
  }
}

function unsigned_bitsize(x) {
  return 32 - leading_zeros(x) | 0;
}

function leading_sign_bits(x) {
  if (x >= 0) {
    return leading_zeros(x) - 1 | 0;
  } else {
    return leading_zeros(x ^ -1) - 1 | 0;
  }
}

function signed_bitsize(x) {
  return 32 - leading_sign_bits(x) | 0;
}

function trailing_zeros(x) {
  if (x === 0) {
    return 32;
  }
  let x$1 = x;
  let n = 31;
  const y = (x$1 << 16);
  if (y !== 0) {
    n = n - 16 | 0;
    x$1 = y;
  }
  const y$1 = (x$1 << 8);
  if (y$1 !== 0) {
    n = n - 8 | 0;
    x$1 = y$1;
  }
  const y$2 = (x$1 << 4);
  if (y$2 !== 0) {
    n = n - 4 | 0;
    x$1 = y$2;
  }
  const y$3 = (x$1 << 2);
  if (y$3 !== 0) {
    n = n - 2 | 0;
    x$1 = y$3;
  }
  const y$4 = (x$1 << 1);
  if (y$4 !== 0) {
    return n - 1 | 0;
  } else {
    return n;
  }
}

function popcount(x) {
  const x$1 = x - ((x >>> 1) & 1431655765) | 0;
  const x$2 = (x$1 & 858993459) + ((x$1 >>> 2) & 858993459) | 0;
  const x$3 = x$2 + (x$2 >>> 4) & 252645135;
  const x$4 = x$3 + (x$3 >>> 8) | 0;
  const x$5 = x$4 + (x$4 >>> 16) | 0;
  return x$5 & 63;
}

function seeded_hash(seed, x) {
  return Caml_hash.caml_hash(10, 100, seed, x);
}

function hash(x) {
  return Caml_hash.caml_hash(10, 100, 0, x);
}

const zero = 0;

const one = 1;

const minus_one = -1;

const max_int = 2147483647;

const min_int = -2147483648;

export {
  zero,
  one,
  minus_one,
  unsigned_div,
  unsigned_rem,
  fdiv,
  cdiv,
  ediv,
  erem,
  succ,
  pred,
  abs,
  max_int,
  min_int,
  lognot,
  unsigned_to_int,
  of_string_opt,
  to_string,
  compare,
  unsigned_compare,
  equal,
  min,
  max,
  popcount,
  unsigned_bitsize,
  signed_bitsize,
  leading_zeros,
  leading_sign_bits,
  trailing_zeros,
  seeded_hash,
  hash,
}
/* unsigned_to_int Not a pure module */
