// melange/int64.js
// Generated by Melange
'use strict';

const Caml = require("melange.js/caml.js");
const Caml_format = require("melange.js/caml_format.js");
const Caml_hash = require("melange.js/caml_hash.js");
const Caml_int64 = require("melange.js/caml_int64.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Stdlib = require("./stdlib.js");

const zero = Caml_int64.zero;

const one = Caml_int64.one;

function succ(n) {
  return Caml_int64.add(n, Caml_int64.one);
}

function pred(n) {
  return Caml_int64.sub(n, Caml_int64.one);
}

function abs(n) {
  if (Caml.i64_ge(n, Caml_int64.zero)) {
    return n;
  } else {
    return Caml_int64.neg(n);
  }
}

const min_int = Caml_int64.min_int;

function lognot(n) {
  return Caml_int64.xor(n, Caml_int64.neg_one);
}

const max_int = Caml_int64.of_int32(Stdlib.max_int);

function unsigned_to_int(n) {
  if (Caml.i64_ge(n, Caml_int64.zero) && Caml.i64_le(n, max_int)) {
    return Caml_int64.to_int32(n);
  }
  
}

function to_string(n) {
  return Caml_format.caml_int64_format("%d", n);
}

function of_string_opt(s) {
  try {
    return Caml_format.caml_int64_of_string(s);
  }
  catch (raw_exn){
    const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.MEL_EXN_ID === Stdlib.Failure) {
      return;
    }
    throw exn;
  }
}

const compare = Caml_int64.compare;

const equal = Caml.i64_eq;

function unsigned_compare(n, m) {
  return Caml_int64.compare(Caml_int64.sub(n, min_int), Caml_int64.sub(m, min_int));
}

function unsigned_lt(n, m) {
  return Caml.i64_lt(Caml_int64.sub(n, min_int), Caml_int64.sub(m, min_int));
}

function min(x, y) {
  if (Caml.i64_le(x, y)) {
    return x;
  } else {
    return y;
  }
}

function max(x, y) {
  if (Caml.i64_ge(x, y)) {
    return x;
  } else {
    return y;
  }
}

function unsigned_div(n, d) {
  if (Caml.i64_lt(d, zero)) {
    if (unsigned_lt(n, d)) {
      return zero;
    } else {
      return one;
    }
  }
  const q = Caml_int64.lsl_(Caml_int64.div(Caml_int64.lsr_(n, 1), d), 1);
  const r = Caml_int64.sub(n, Caml_int64.mul(q, d));
  if (unsigned_lt(r, d)) {
    return q;
  } else {
    return Caml_int64.add(q, Caml_int64.one);
  }
}

function unsigned_rem(n, d) {
  return Caml_int64.sub(n, Caml_int64.mul(unsigned_div(n, d), d));
}

function seeded_hash(seed, x) {
  return Caml_hash.caml_hash(10, 100, seed, x);
}

function hash(x) {
  return Caml_hash.caml_hash(10, 100, 0, x);
}

const minus_one = Caml_int64.neg_one;

const max_int$1 = Caml_int64.max_int;

module.exports = {
  zero,
  one,
  minus_one,
  unsigned_div,
  unsigned_rem,
  succ,
  pred,
  abs,
  max_int: max_int$1,
  min_int,
  lognot,
  unsigned_to_int,
  of_string_opt,
  to_string,
  compare,
  unsigned_compare,
  equal,
  min,
  max,
  seeded_hash,
  hash,
}
/* No side effect */
