// melange/int64.mjs
// Generated by Melange

import * as Caml from "melange.js/caml.mjs";
import * as Caml_format from "melange.js/caml_format.mjs";
import * as Caml_hash from "melange.js/caml_hash.mjs";
import * as Caml_int64 from "melange.js/caml_int64.mjs";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";
import * as Stdlib from "./stdlib.mjs";

const zero = Caml_int64.zero;

const one = Caml_int64.one;

function succ(n) {
  return Caml_int64.add(n, Caml_int64.one);
}

function pred(n) {
  return Caml_int64.sub(n, Caml_int64.one);
}

function abs(n) {
  if (Caml.i64_ge(n, Caml_int64.zero)) {
    return n;
  } else {
    return Caml_int64.neg(n);
  }
}

const min_int = Caml_int64.min_int;

function lognot(n) {
  return Caml_int64.xor(n, Caml_int64.neg_one);
}

const max_int = Caml_int64.of_int32(Stdlib.max_int);

function unsigned_to_int(n) {
  if (Caml.i64_ge(n, Caml_int64.zero) && Caml.i64_le(n, max_int)) {
    return Caml_int64.to_int32(n);
  }
  
}

function to_string(n) {
  return Caml_format.caml_int64_format("%d", n);
}

function of_string_opt(s) {
  try {
    return Caml_format.caml_int64_of_string(s);
  }
  catch (raw_exn){
    const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.MEL_EXN_ID === Stdlib.Failure) {
      return;
    }
    throw exn;
  }
}

const compare = Caml_int64.compare;

const equal = Caml.i64_eq;

function unsigned_compare(n, m) {
  return Caml_int64.compare(Caml_int64.sub(n, min_int), Caml_int64.sub(m, min_int));
}

function unsigned_lt(n, m) {
  return Caml.i64_lt(Caml_int64.sub(n, min_int), Caml_int64.sub(m, min_int));
}

function min(x, y) {
  if (Caml.i64_le(x, y)) {
    return x;
  } else {
    return y;
  }
}

function max(x, y) {
  if (Caml.i64_ge(x, y)) {
    return x;
  } else {
    return y;
  }
}

function unsigned_div(n, d) {
  if (Caml.i64_lt(d, zero)) {
    if (unsigned_lt(n, d)) {
      return zero;
    } else {
      return one;
    }
  }
  const q = Caml_int64.lsl_(Caml_int64.div(Caml_int64.lsr_(n, 1), d), 1);
  const r = Caml_int64.sub(n, Caml_int64.mul(q, d));
  if (unsigned_lt(r, d)) {
    return q;
  } else {
    return Caml_int64.add(q, Caml_int64.one);
  }
}

function unsigned_rem(n, d) {
  return Caml_int64.sub(n, Caml_int64.mul(unsigned_div(n, d), d));
}

function fdiv(n, d) {
  const q = Caml_int64.div(n, d);
  if (Caml.i64_ge(Caml_int64.xor(n, d), Caml_int64.zero) || Caml.i64_eq(n, Caml_int64.mul(q, d))) {
    return q;
  } else {
    return Caml_int64.sub(q, Caml_int64.one);
  }
}

function cdiv(n, d) {
  const q = Caml_int64.div(n, d);
  if (Caml.i64_lt(Caml_int64.xor(n, d), Caml_int64.zero) || Caml.i64_eq(n, Caml_int64.mul(q, d))) {
    return q;
  } else {
    return Caml_int64.add(q, Caml_int64.one);
  }
}

function erem(n, d) {
  const r = Caml_int64.mod_(n, d);
  if (Caml.i64_ge(r, Caml_int64.zero)) {
    return r;
  } else if (Caml.i64_ge(d, Caml_int64.zero)) {
    return Caml_int64.add(r, d);
  } else {
    return Caml_int64.sub(r, d);
  }
}

function ediv(n, d) {
  const q = Caml_int64.div(n, d);
  const r = Caml_int64.sub(n, Caml_int64.mul(q, d));
  if (Caml.i64_ge(r, Caml_int64.zero)) {
    return q;
  } else if (Caml.i64_ge(d, Caml_int64.zero)) {
    return Caml_int64.sub(q, Caml_int64.one);
  } else {
    return Caml_int64.add(q, Caml_int64.one);
  }
}

function leading_zeros(x) {
  let x$1 = x;
  let n = 64;
  const y = Caml_int64.lsr_(x$1, 32);
  if (Caml.i64_neq(y, Caml_int64.zero)) {
    n = n - 32 | 0;
    x$1 = y;
  }
  const y$1 = Caml_int64.lsr_(x$1, 16);
  if (Caml.i64_neq(y$1, Caml_int64.zero)) {
    n = n - 16 | 0;
    x$1 = y$1;
  }
  const y$2 = Caml_int64.lsr_(x$1, 8);
  if (Caml.i64_neq(y$2, Caml_int64.zero)) {
    n = n - 8 | 0;
    x$1 = y$2;
  }
  const y$3 = Caml_int64.lsr_(x$1, 4);
  if (Caml.i64_neq(y$3, Caml_int64.zero)) {
    n = n - 4 | 0;
    x$1 = y$3;
  }
  const y$4 = Caml_int64.lsr_(x$1, 2);
  if (Caml.i64_neq(y$4, Caml_int64.zero)) {
    n = n - 2 | 0;
    x$1 = y$4;
  }
  const y$5 = Caml_int64.lsr_(x$1, 1);
  if (Caml.i64_neq(y$5, Caml_int64.zero)) {
    return n - 2 | 0;
  } else {
    return n - Caml_int64.to_int32(x$1) | 0;
  }
}

function unsigned_bitsize(x) {
  return 64 - leading_zeros(x) | 0;
}

function leading_sign_bits(x) {
  if (Caml.i64_ge(x, Caml_int64.zero)) {
    return leading_zeros(x) - 1 | 0;
  } else {
    return leading_zeros(Caml_int64.xor(x, Caml_int64.neg_one)) - 1 | 0;
  }
}

function signed_bitsize(x) {
  return 64 - leading_sign_bits(x) | 0;
}

function trailing_zeros(x) {
  if (Caml.i64_eq(x, Caml_int64.zero)) {
    return 64;
  }
  let x$1 = x;
  let n = 63;
  const y = Caml_int64.lsl_(x$1, 32);
  if (Caml.i64_neq(y, Caml_int64.zero)) {
    n = n - 32 | 0;
    x$1 = y;
  }
  const y$1 = Caml_int64.lsl_(x$1, 16);
  if (Caml.i64_neq(y$1, Caml_int64.zero)) {
    n = n - 16 | 0;
    x$1 = y$1;
  }
  const y$2 = Caml_int64.lsl_(x$1, 8);
  if (Caml.i64_neq(y$2, Caml_int64.zero)) {
    n = n - 8 | 0;
    x$1 = y$2;
  }
  const y$3 = Caml_int64.lsl_(x$1, 4);
  if (Caml.i64_neq(y$3, Caml_int64.zero)) {
    n = n - 4 | 0;
    x$1 = y$3;
  }
  const y$4 = Caml_int64.lsl_(x$1, 2);
  if (Caml.i64_neq(y$4, Caml_int64.zero)) {
    n = n - 2 | 0;
    x$1 = y$4;
  }
  const y$5 = Caml_int64.lsl_(x$1, 1);
  if (Caml.i64_neq(y$5, Caml_int64.zero)) {
    return n - 1 | 0;
  } else {
    return n;
  }
}

function popcount(x) {
  const x$1 = Caml_int64.sub(x, Caml_int64.and_(Caml_int64.lsr_(x, 1), [
    1431655765,
    1431655765
  ]));
  const x$2 = Caml_int64.add(Caml_int64.and_(x$1, [
    858993459,
    858993459
  ]), Caml_int64.and_(Caml_int64.lsr_(x$1, 2), [
    858993459,
    858993459
  ]));
  const x$3 = Caml_int64.and_(Caml_int64.add(x$2, Caml_int64.lsr_(x$2, 4)), [
    252645135,
    252645135
  ]);
  const x$4 = Caml_int64.add(x$3, Caml_int64.lsr_(x$3, 8));
  const x$5 = Caml_int64.add(x$4, Caml_int64.lsr_(x$4, 16));
  const x$6 = Caml_int64.add(x$5, Caml_int64.lsr_(x$5, 32));
  return Caml_int64.to_int32(x$6) & 127;
}

function seeded_hash(seed, x) {
  return Caml_hash.caml_hash(10, 100, seed, x);
}

function hash(x) {
  return Caml_hash.caml_hash(10, 100, 0, x);
}

const minus_one = Caml_int64.neg_one;

const max_int$1 = Caml_int64.max_int;

export {
  zero,
  one,
  minus_one,
  unsigned_div,
  unsigned_rem,
  fdiv,
  cdiv,
  ediv,
  erem,
  succ,
  pred,
  abs,
  max_int$1 as max_int,
  min_int,
  lognot,
  unsigned_to_int,
  of_string_opt,
  to_string,
  compare,
  unsigned_compare,
  equal,
  min,
  max,
  popcount,
  unsigned_bitsize,
  signed_bitsize,
  leading_zeros,
  leading_sign_bits,
  trailing_zeros,
  seeded_hash,
  hash,
}
/* No side effect */
