// melange/lazy.mjs
// Generated by Melange

import * as Caml_external_polyfill from "melange.js/caml_external_polyfill.mjs";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";
import * as CamlinternalLazy from "./camlinternalLazy.mjs";
import * as Curry from "melange.js/curry.mjs";
import * as Stdlib__Atomic from "./atomic.mjs";

function from_fun(f) {
  return {
    LAZY_DONE: false,
    VAL: (function () {
      return Curry._1(f, undefined);
    })
  };
}

function from_val(v) {
  return {
    LAZY_DONE: true,
    VAL: v
  };
}

function map(f, x) {
  return {
    LAZY_DONE: false,
    VAL: (function () {
      return Curry._1(f, CamlinternalLazy.force(x));
    })
  };
}

function map_val(f, x) {
  if (x.LAZY_DONE) {
    return {
      LAZY_DONE: true,
      VAL: Curry._1(f, CamlinternalLazy.force(x))
    };
  } else {
    return {
      LAZY_DONE: false,
      VAL: (function () {
        return Curry._1(f, CamlinternalLazy.force(x));
      })
    };
  }
}

function from_val$1(v) {
  return {
    v: {
      TAG: /* Result */ 2,
      _0: {
        TAG: /* Ok */ 0,
        _0: v
      }
    }
  };
}

function from_fun$1(f) {
  return {
    v: {
      TAG: /* Thunk */ 0,
      _0: f
    }
  };
}

function is_val(th) {
  const match = th.v;
  switch (match.TAG) {
    case /* Thunk */ 0 :
    case /* Forcing */ 1 :
      return false;
    case /* Result */ 2 :
      if (match._0.TAG === /* Ok */ 0) {
        return true;
      } else {
        return false;
      }
  }
}

function force(th) {
  while (true) {
    const mut = th.v;
    switch (mut.TAG) {
      case /* Thunk */ 0 :
        const mut$1 = Caml_external_polyfill.resolve("caml_ml_mutex_new")();
        Caml_external_polyfill.resolve("caml_ml_mutex_lock")(mut$1);
        if (Stdlib__Atomic.compare_and_set(th, mut, {
            TAG: /* Forcing */ 1,
            _0: mut$1
          })) {
          let exit = 0;
          let v;
          try {
            v = Curry._1(mut._0, undefined);
            exit = 1;
          }
          catch (raw_exn){
            const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
            let bt;
            th.v = {
              TAG: /* Result */ 2,
              _0: {
                TAG: /* Error */ 1,
                _0: [
                  exn,
                  bt
                ]
              }
            };
          }
          if (exit === 1) {
            th.v = {
              TAG: /* Result */ 2,
              _0: {
                TAG: /* Ok */ 0,
                _0: v
              }
            };
          }
          Caml_external_polyfill.resolve("caml_ml_mutex_unlock")(mut$1);
          continue;
        }
        Caml_external_polyfill.resolve("caml_ml_mutex_unlock")(mut$1);
        continue;
        break;
      case /* Forcing */ 1 :
        const mut$2 = mut._0;
        Caml_external_polyfill.resolve("caml_ml_mutex_lock")(mut$2);
        Caml_external_polyfill.resolve("caml_ml_mutex_unlock")(mut$2);
        continue;
      case /* Result */ 2 :
        const v$1 = mut._0;
        if (v$1.TAG === /* Ok */ 0) {
          return v$1._0;
        }
        const match = v$1._0;
        const exn$1 = match[0];
        Caml_external_polyfill.resolve("caml_restore_raw_backtrace")(exn$1, match[1]);
        throw exn$1;
    }
  };
}

const Undefined = CamlinternalLazy.Undefined;

const is_val$1 = CamlinternalLazy.is_val;

const force_val = CamlinternalLazy.force_val;

const Mutexed = {
  is_val: is_val,
  from_val: from_val$1,
  from_fun: from_fun$1,
  force: force
};

export {
  Undefined,
  map,
  is_val$1 as is_val,
  from_val,
  map_val,
  from_fun,
  force_val,
  Mutexed,
}
/* No side effect */
