// melange/obj.mjs
// Generated by Melange

import * as Caml_external_polyfill from "melange.js/caml_external_polyfill.mjs";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";
import * as Stdlib__Sys from "./sys.mjs";

function is_block(a) {
  return typeof a !== "number";
}

function double_field(x, i) {
  return Caml_external_polyfill.resolve("caml_floatarray_get")(x, i);
}

function set_double_field(x, i, v) {
  Caml_external_polyfill.resolve("caml_floatarray_set")(x, i, v);
}

function of_val(x) {
  const slot = typeof x !== "number" && x.TAG !== 248 && (x.length | 0) >= 1 ? x[0] : x;
  let name;
  if (typeof slot !== "number" && slot.TAG === 248) {
    name = slot[0];
  } else {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "Obj.extension_constructor"
      });
  }
  if (name.TAG === 252) {
    return slot;
  }
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: "Obj.extension_constructor"
    });
}

function name(slot) {
  return slot[0];
}

function id(slot) {
  return slot[1];
}

const Extension_constructor = {
  of_val: of_val,
  name: name,
  id: id
};

const max_ephe_length = Stdlib__Sys.max_array_length - 2 | 0;

function create(l) {
  if (!(0 <= l && l <= max_ephe_length)) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "Obj.Ephemeron.create"
      });
  }
  return Caml_external_polyfill.resolve("caml_ephe_create")(l);
}

function length(x) {
  return (x.length | 0) - 2 | 0;
}

function raise_if_invalid_offset(e, o, msg) {
  if (0 <= o && o < ((e.length | 0) - 2 | 0)) {
    return;
  }
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: msg
    });
}

function get_key(e, o) {
  raise_if_invalid_offset(e, o, "Obj.Ephemeron.get_key");
  return Caml_external_polyfill.resolve("caml_ephe_get_key")(e, o);
}

function get_key_copy(e, o) {
  raise_if_invalid_offset(e, o, "Obj.Ephemeron.get_key_copy");
  return Caml_external_polyfill.resolve("caml_ephe_get_key_copy")(e, o);
}

function set_key(e, o, x) {
  raise_if_invalid_offset(e, o, "Obj.Ephemeron.set_key");
  Caml_external_polyfill.resolve("caml_ephe_set_key")(e, o, x);
}

function unset_key(e, o) {
  raise_if_invalid_offset(e, o, "Obj.Ephemeron.unset_key");
  Caml_external_polyfill.resolve("caml_ephe_unset_key")(e, o);
}

function check_key(e, o) {
  raise_if_invalid_offset(e, o, "Obj.Ephemeron.check_key");
  return Caml_external_polyfill.resolve("caml_ephe_check_key")(e, o);
}

function blit_key(e1, o1, e2, o2, l) {
  if (l < 0 || o1 < 0 || o1 > (((e1.length | 0) - 2 | 0) - l | 0) || o2 < 0 || o2 > (((e2.length | 0) - 2 | 0) - l | 0)) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "Obj.Ephemeron.blit_key"
      });
  }
  if (l !== 0) {
    return Caml_external_polyfill.resolve("caml_ephe_blit_key")(e1, o1, e2, o2, l);
  }
  
}

const first_non_constant_constructor_tag = 0;

const last_non_constant_constructor_tag = 243;

const forcing_tag = 244;

const cont_tag = 245;

const lazy_tag = 246;

const closure_tag = 247;

const object_tag = 248;

const infix_tag = 249;

const forward_tag = 250;

const no_scan_tag = 251;

const abstract_tag = 251;

const string_tag = 252;

const double_tag = 253;

const double_array_tag = 254;

const custom_tag = 255;

const int_tag = 1000;

const out_of_heap_tag = 1001;

const unaligned_tag = 1002;

function Ephemeron_get_data(prim) {
  return Caml_external_polyfill.resolve("caml_ephe_get_data")(prim);
}

function Ephemeron_get_data_copy(prim) {
  return Caml_external_polyfill.resolve("caml_ephe_get_data_copy")(prim);
}

function Ephemeron_set_data(prim0, prim1) {
  return Caml_external_polyfill.resolve("caml_ephe_set_data")(prim0, prim1);
}

function Ephemeron_unset_data(prim) {
  return Caml_external_polyfill.resolve("caml_ephe_unset_data")(prim);
}

function Ephemeron_check_data(prim) {
  return Caml_external_polyfill.resolve("caml_ephe_check_data")(prim);
}

function Ephemeron_blit_data(prim0, prim1) {
  return Caml_external_polyfill.resolve("caml_ephe_blit_data")(prim0, prim1);
}

const Ephemeron = {
  create: create,
  length: length,
  get_key: get_key,
  get_key_copy: get_key_copy,
  set_key: set_key,
  unset_key: unset_key,
  check_key: check_key,
  blit_key: blit_key,
  get_data: Ephemeron_get_data,
  get_data_copy: Ephemeron_get_data_copy,
  set_data: Ephemeron_set_data,
  unset_data: Ephemeron_unset_data,
  check_data: Ephemeron_check_data,
  blit_data: Ephemeron_blit_data,
  max_ephe_length: max_ephe_length
};

export {
  is_block,
  double_field,
  set_double_field,
  first_non_constant_constructor_tag,
  last_non_constant_constructor_tag,
  forcing_tag,
  cont_tag,
  lazy_tag,
  closure_tag,
  object_tag,
  infix_tag,
  forward_tag,
  no_scan_tag,
  abstract_tag,
  string_tag,
  double_tag,
  double_array_tag,
  custom_tag,
  int_tag,
  out_of_heap_tag,
  unaligned_tag,
  Extension_constructor,
  Ephemeron,
}
/* No side effect */
