// melange/out_channel.js
// Generated by Melange
'use strict';

const Caml_external_polyfill = require("melange.js/caml_external_polyfill.js");
const Curry = require("melange.js/curry.js");
const Stdlib = require("./stdlib.js");
const Stdlib__Fun = require("./fun.js");

function with_open(openfun, s, f) {
  const oc = Curry._1(openfun, s);
  return Stdlib__Fun.protect((function (param) {
    Stdlib.close_out_noerr(oc);
  }), (function (param) {
    return Curry._1(f, oc);
  }));
}

function with_open_bin(s, f) {
  return with_open(Stdlib.open_out_bin, s, f);
}

function with_open_text(s, f) {
  return with_open(Stdlib.open_out, s, f);
}

function with_open_gen(flags, perm, s, f) {
  return with_open((function (param) {
    return Stdlib.open_out_gen(flags, perm, param);
  }), s, f);
}

const seek = Stdlib.LargeFile.seek_out;

const pos = Stdlib.LargeFile.pos_out;

const length = Stdlib.LargeFile.out_channel_length;

const stdout = Stdlib.stdout;

const stderr = Stdlib.stderr;

const open_bin = Stdlib.open_out_bin;

const open_text = Stdlib.open_out;

const open_gen = Stdlib.open_out_gen;

const close = Stdlib.close_out;

const close_noerr = Stdlib.close_out_noerr;

const output_char = Stdlib.output_char;

const output_byte = Stdlib.output_byte;

const output_string = Stdlib.output_string;

const output_bytes = Stdlib.output_bytes;

const output = Stdlib.output;

const output_substring = Stdlib.output_substring;

const flush = Stdlib.flush;

const flush_all = Stdlib.flush_all;

const set_binary_mode = Stdlib.set_binary_mode_out;

function is_binary_mode(prim) {
  return Caml_external_polyfill.resolve("caml_ml_is_binary_mode")(prim);
}

function set_buffered(prim0, prim1) {
  return Caml_external_polyfill.resolve("caml_ml_set_buffered")(prim0, prim1);
}

function is_buffered(prim) {
  return Caml_external_polyfill.resolve("caml_ml_is_buffered")(prim);
}

function isatty(prim) {
  return Caml_external_polyfill.resolve("caml_sys_isatty")(prim);
}

module.exports = {
  stdout,
  stderr,
  open_bin,
  open_text,
  open_gen,
  with_open_bin,
  with_open_text,
  with_open_gen,
  close,
  close_noerr,
  output_char,
  output_byte,
  output_string,
  output_bytes,
  output,
  output_substring,
  flush,
  flush_all,
  seek,
  pos,
  length,
  set_binary_mode,
  is_binary_mode,
  set_buffered,
  is_buffered,
  isatty,
}
/* Stdlib__Fun Not a pure module */
