// melange/pair.mjs
// Generated by Melange

import * as Curry from "melange.js/curry.mjs";

function make(a, b) {
  return [
    a,
    b
  ];
}

function fst(param) {
  return param[0];
}

function snd(param) {
  return param[1];
}

function swap(param) {
  return [
    param[1],
    param[0]
  ];
}

function fold(f, param) {
  return Curry._2(f, param[0], param[1]);
}

function map(f, g, param) {
  return [
    Curry._1(f, param[0]),
    Curry._1(g, param[1])
  ];
}

function iter(f, g, param) {
  Curry._1(f, param[0]);
  return Curry._1(g, param[1]);
}

function map_fst(f, param) {
  return [
    Curry._1(f, param[0]),
    param[1]
  ];
}

function map_snd(g, param) {
  return [
    param[0],
    Curry._1(g, param[1])
  ];
}

function equal(eqa, eqb, param, param$1) {
  if (Curry._2(eqa, param[0], param$1[0])) {
    return Curry._2(eqb, param[1], param$1[1]);
  } else {
    return false;
  }
}

function compare(cmpa, cmpb, param, param$1) {
  const c = Curry._2(cmpa, param[0], param$1[0]);
  if (c !== 0) {
    return c;
  } else {
    return Curry._2(cmpb, param[1], param$1[1]);
  }
}

export {
  make,
  fst,
  snd,
  swap,
  fold,
  map,
  iter,
  map_fst,
  map_snd,
  equal,
  compare,
}
/* No side effect */
