// melange/set.mjs
// Generated by Melange

import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";
import * as Caml_obj from "melange.js/caml_obj.mjs";
import * as Caml_option from "melange.js/caml_option.mjs";
import * as Curry from "melange.js/curry.mjs";
import * as Stdlib from "./stdlib.mjs";
import * as Stdlib__List from "./list.mjs";
import * as Stdlib__Seq from "./seq.mjs";

function Make(funarg) {
  const height = function (param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return 0;
    } else {
      return param.h;
    }
  };
  const create = function (l, v, r) {
    let hl;
    hl = /* tag */ typeof l !== "object" && typeof l !== "function" ? 0 : l.h;
    let hr;
    hr = /* tag */ typeof r !== "object" && typeof r !== "function" ? 0 : r.h;
    return {
      TAG: /* Node */ 0,
      l: l,
      v: v,
      r: r,
      h: hl >= hr ? hl + 1 | 0 : hr + 1 | 0
    };
  };
  const bal = function (l, v, r) {
    let hl;
    hl = /* tag */ typeof l !== "object" && typeof l !== "function" ? 0 : l.h;
    let hr;
    hr = /* tag */ typeof r !== "object" && typeof r !== "function" ? 0 : r.h;
    if (hl > (hr + 2 | 0)) {
      if (/* tag */ typeof l !== "object" && typeof l !== "function") {
        throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
            MEL_EXN_ID: "Invalid_argument",
            _1: "Set.bal"
          });
      }
      const lr = l.r;
      const lv = l.v;
      const ll = l.l;
      if (height(ll) >= height(lr)) {
        return create(ll, lv, create(lr, v, r));
      }
      if (!/* tag */ (typeof lr !== "object" && typeof lr !== "function")) {
        return create(create(ll, lv, lr.l), lr.v, create(lr.r, v, r));
      }
      throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
          MEL_EXN_ID: "Invalid_argument",
          _1: "Set.bal"
        });
    }
    if (hr <= (hl + 2 | 0)) {
      return {
        TAG: /* Node */ 0,
        l: l,
        v: v,
        r: r,
        h: hl >= hr ? hl + 1 | 0 : hr + 1 | 0
      };
    }
    if (/* tag */ typeof r !== "object" && typeof r !== "function") {
      throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
          MEL_EXN_ID: "Invalid_argument",
          _1: "Set.bal"
        });
    }
    const rr = r.r;
    const rv = r.v;
    const rl = r.l;
    if (height(rr) >= height(rl)) {
      return create(create(l, v, rl), rv, rr);
    }
    if (!/* tag */ (typeof rl !== "object" && typeof rl !== "function")) {
      return create(create(l, v, rl.l), rl.v, create(rl.r, rv, rr));
    }
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "Set.bal"
      });
  };
  const add = function (x, t) {
    if (/* tag */ typeof t !== "object" && typeof t !== "function") {
      return {
        TAG: /* Node */ 0,
        l: /* Empty */ 0,
        v: x,
        r: /* Empty */ 0,
        h: 1
      };
    }
    const r = t.r;
    const v = t.v;
    const l = t.l;
    const c = Curry._2(funarg.compare, x, v);
    if (c === 0) {
      return t;
    }
    if (c < 0) {
      const ll = add(x, l);
      if (l === ll) {
        return t;
      } else {
        return bal(ll, v, r);
      }
    }
    const rr = add(x, r);
    if (r === rr) {
      return t;
    } else {
      return bal(l, v, rr);
    }
  };
  const singleton = function (x) {
    return {
      TAG: /* Node */ 0,
      l: /* Empty */ 0,
      v: x,
      r: /* Empty */ 0,
      h: 1
    };
  };
  const add_min_element = function (x, param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return singleton(x);
    } else {
      return bal(add_min_element(x, param.l), param.v, param.r);
    }
  };
  const add_max_element = function (x, param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return singleton(x);
    } else {
      return bal(param.l, param.v, add_max_element(x, param.r));
    }
  };
  const join = function (l, v, r) {
    if (/* tag */ typeof l !== "object" && typeof l !== "function") {
      return add_min_element(v, r);
    }
    const lh = l.h;
    if (/* tag */ typeof r !== "object" && typeof r !== "function") {
      return add_max_element(v, l);
    }
    const rh = r.h;
    if (lh > (rh + 2 | 0)) {
      return bal(l.l, l.v, join(l.r, v, r));
    } else if (rh > (lh + 2 | 0)) {
      return bal(join(l, v, r.l), r.v, r.r);
    } else {
      return create(l, v, r);
    }
  };
  const min_elt = function (_param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
            MEL_EXN_ID: Stdlib.Not_found
          });
      }
      const l = param.l;
      if (/* tag */ typeof l !== "object" && typeof l !== "function") {
        return param.v;
      }
      _param = l;
      continue;
    };
  };
  const min_elt_opt = function (_param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return;
      }
      const l = param.l;
      if (/* tag */ typeof l !== "object" && typeof l !== "function") {
        return Caml_option.some(param.v);
      }
      _param = l;
      continue;
    };
  };
  const max_elt = function (_param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
            MEL_EXN_ID: Stdlib.Not_found
          });
      }
      let tmp = param.r;
      if (/* tag */ typeof tmp !== "object" && typeof tmp !== "function") {
        return param.v;
      }
      _param = param.r;
      continue;
    };
  };
  const max_elt_opt = function (_param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return;
      }
      let tmp = param.r;
      if (/* tag */ typeof tmp !== "object" && typeof tmp !== "function") {
        return Caml_option.some(param.v);
      }
      _param = param.r;
      continue;
    };
  };
  const remove_min_elt = function (param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
          MEL_EXN_ID: "Invalid_argument",
          _1: "Set.remove_min_elt"
        });
    }
    const l = param.l;
    if (/* tag */ typeof l !== "object" && typeof l !== "function") {
      return param.r;
    } else {
      return bal(remove_min_elt(l), param.v, param.r);
    }
  };
  const concat = function (t1, t2) {
    if (/* tag */ typeof t1 !== "object" && typeof t1 !== "function") {
      return t2;
    } else if (/* tag */ typeof t2 !== "object" && typeof t2 !== "function") {
      return t1;
    } else {
      return join(t1, min_elt(t2), remove_min_elt(t2));
    }
  };
  const split = function (x, param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return [
        /* Empty */ 0,
        false,
        /* Empty */ 0
      ];
    }
    const r = param.r;
    const v = param.v;
    const l = param.l;
    const c = Curry._2(funarg.compare, x, v);
    if (c === 0) {
      return [
        l,
        true,
        r
      ];
    }
    if (c < 0) {
      const match = split(x, l);
      return [
        match[0],
        match[1],
        join(match[2], v, r)
      ];
    }
    const match$1 = split(x, r);
    return [
      join(l, v, match$1[0]),
      match$1[1],
      match$1[2]
    ];
  };
  const is_empty = function (param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return true;
    } else {
      return false;
    }
  };
  const mem = function (x, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return false;
      }
      const c = Curry._2(funarg.compare, x, param.v);
      if (c === 0) {
        return true;
      }
      _param = c < 0 ? param.l : param.r;
      continue;
    };
  };
  const remove = function (x, t) {
    if (/* tag */ typeof t !== "object" && typeof t !== "function") {
      return /* Empty */ 0;
    }
    const r = t.r;
    const v = t.v;
    const l = t.l;
    const c = Curry._2(funarg.compare, x, v);
    if (c === 0) {
      if (/* tag */ typeof l !== "object" && typeof l !== "function") {
        return r;
      } else if (/* tag */ typeof r !== "object" && typeof r !== "function") {
        return l;
      } else {
        return bal(l, min_elt(r), remove_min_elt(r));
      }
    }
    if (c < 0) {
      const ll = remove(x, l);
      if (l === ll) {
        return t;
      } else {
        return bal(ll, v, r);
      }
    }
    const rr = remove(x, r);
    if (r === rr) {
      return t;
    } else {
      return bal(l, v, rr);
    }
  };
  const union = function (s1, s2) {
    if (/* tag */ typeof s1 !== "object" && typeof s1 !== "function") {
      return s2;
    }
    const h1 = s1.h;
    const v1 = s1.v;
    if (/* tag */ typeof s2 !== "object" && typeof s2 !== "function") {
      return s1;
    }
    const h2 = s2.h;
    const v2 = s2.v;
    if (h1 >= h2) {
      if (h2 === 1) {
        return add(v2, s1);
      }
      const match = split(v1, s2);
      return join(union(s1.l, match[0]), v1, union(s1.r, match[2]));
    }
    if (h1 === 1) {
      return add(v1, s2);
    }
    const match$1 = split(v2, s1);
    return join(union(match$1[0], s2.l), v2, union(match$1[2], s2.r));
  };
  const inter = function (s1, s2) {
    if (/* tag */ typeof s1 !== "object" && typeof s1 !== "function") {
      return /* Empty */ 0;
    }
    if (/* tag */ typeof s2 !== "object" && typeof s2 !== "function") {
      return /* Empty */ 0;
    }
    const r1 = s1.r;
    const v1 = s1.v;
    const l1 = s1.l;
    const match = split(v1, s2);
    const l2 = match[0];
    if (match[1]) {
      return join(inter(l1, l2), v1, inter(r1, match[2]));
    } else {
      return concat(inter(l1, l2), inter(r1, match[2]));
    }
  };
  const split_bis = function (x, param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return {
        TAG: /* NotFound */ 0,
        _0: /* Empty */ 0,
        _1: (function (param) {
          return /* Empty */ 0;
        })
      };
    }
    const r = param.r;
    const v = param.v;
    const l = param.l;
    const c = Curry._2(funarg.compare, x, v);
    if (c === 0) {
      return /* Found */ 0;
    }
    if (c < 0) {
      const match = split_bis(x, l);
      if (/* tag */ typeof match !== "object" && typeof match !== "function") {
        return /* Found */ 0;
      }
      const rl = match._1;
      return {
        TAG: /* NotFound */ 0,
        _0: match._0,
        _1: (function (param) {
          return join(Curry._1(rl, undefined), v, r);
        })
      };
    }
    const match$1 = split_bis(x, r);
    if (/* tag */ typeof match$1 !== "object" && typeof match$1 !== "function") {
      return /* Found */ 0;
    } else {
      return {
        TAG: /* NotFound */ 0,
        _0: join(l, v, match$1._0),
        _1: match$1._1
      };
    }
  };
  const disjoint = function (_s1, _s2) {
    while (true) {
      const s2 = _s2;
      const s1 = _s1;
      if (/* tag */ typeof s1 !== "object" && typeof s1 !== "function") {
        return true;
      }
      if (/* tag */ typeof s2 !== "object" && typeof s2 !== "function") {
        return true;
      }
      if (s1 === s2) {
        return false;
      }
      const match = split_bis(s1.v, s2);
      if (/* tag */ typeof match !== "object" && typeof match !== "function") {
        return false;
      }
      if (!disjoint(s1.l, match._0)) {
        return false;
      }
      _s2 = Curry._1(match._1, undefined);
      _s1 = s1.r;
      continue;
    };
  };
  const diff = function (s1, s2) {
    if (/* tag */ typeof s1 !== "object" && typeof s1 !== "function") {
      return /* Empty */ 0;
    }
    if (/* tag */ typeof s2 !== "object" && typeof s2 !== "function") {
      return s1;
    }
    const r1 = s1.r;
    const v1 = s1.v;
    const l1 = s1.l;
    const match = split(v1, s2);
    const l2 = match[0];
    if (match[1]) {
      return concat(diff(l1, l2), diff(r1, match[2]));
    } else {
      return join(diff(l1, l2), v1, diff(r1, match[2]));
    }
  };
  const cons_enum = function (_s, _e) {
    while (true) {
      const e = _e;
      const s = _s;
      if (/* tag */ typeof s !== "object" && typeof s !== "function") {
        return e;
      }
      _e = {
        TAG: /* More */ 0,
        _0: s.v,
        _1: s.r,
        _2: e
      };
      _s = s.l;
      continue;
    };
  };
  const compare = function (s1, s2) {
    let _e1 = cons_enum(s1, /* End */ 0);
    let _e2 = cons_enum(s2, /* End */ 0);
    while (true) {
      const e2 = _e2;
      const e1 = _e1;
      if (/* tag */ typeof e1 !== "object" && typeof e1 !== "function") {
        if (/* tag */ typeof e2 !== "object" && typeof e2 !== "function") {
          return 0;
        } else {
          return -1;
        }
      }
      if (/* tag */ typeof e2 !== "object" && typeof e2 !== "function") {
        return 1;
      }
      const c = Curry._2(funarg.compare, e1._0, e2._0);
      if (c !== 0) {
        return c;
      }
      _e2 = cons_enum(e2._1, e2._2);
      _e1 = cons_enum(e1._1, e1._2);
      continue;
    };
  };
  const equal = function (s1, s2) {
    return compare(s1, s2) === 0;
  };
  const subset = function (_s1, _s2) {
    while (true) {
      const s2 = _s2;
      const s1 = _s1;
      if (/* tag */ typeof s1 !== "object" && typeof s1 !== "function") {
        return true;
      }
      const r1 = s1.r;
      const v1 = s1.v;
      const l1 = s1.l;
      if (/* tag */ typeof s2 !== "object" && typeof s2 !== "function") {
        return false;
      }
      const r2 = s2.r;
      const l2 = s2.l;
      const c = Curry._2(funarg.compare, v1, s2.v);
      if (c === 0) {
        if (!subset(l1, l2)) {
          return false;
        }
        _s2 = r2;
        _s1 = r1;
        continue;
      }
      if (c < 0) {
        if (!subset({
            TAG: /* Node */ 0,
            l: l1,
            v: v1,
            r: /* Empty */ 0,
            h: 0
          }, l2)) {
          return false;
        }
        _s1 = r1;
        continue;
      }
      if (!subset({
          TAG: /* Node */ 0,
          l: /* Empty */ 0,
          v: v1,
          r: r1,
          h: 0
        }, r2)) {
        return false;
      }
      _s1 = l1;
      continue;
    };
  };
  const iter = function (f, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return;
      }
      iter(f, param.l);
      Curry._1(f, param.v);
      _param = param.r;
      continue;
    };
  };
  const fold = function (f, _s, _accu) {
    while (true) {
      const accu = _accu;
      const s = _s;
      if (/* tag */ typeof s !== "object" && typeof s !== "function") {
        return accu;
      }
      _accu = Curry._2(f, s.v, fold(f, s.l, accu));
      _s = s.r;
      continue;
    };
  };
  const for_all = function (p, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return true;
      }
      if (!Curry._1(p, param.v)) {
        return false;
      }
      if (!for_all(p, param.l)) {
        return false;
      }
      _param = param.r;
      continue;
    };
  };
  const exists = function (p, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return false;
      }
      if (Curry._1(p, param.v)) {
        return true;
      }
      if (exists(p, param.l)) {
        return true;
      }
      _param = param.r;
      continue;
    };
  };
  const filter = function (p, t) {
    if (/* tag */ typeof t !== "object" && typeof t !== "function") {
      return /* Empty */ 0;
    }
    const r = t.r;
    const v = t.v;
    const l = t.l;
    const l$p = filter(p, l);
    const pv = Curry._1(p, v);
    const r$p = filter(p, r);
    if (pv) {
      if (l === l$p && r === r$p) {
        return t;
      } else {
        return join(l$p, v, r$p);
      }
    } else {
      return concat(l$p, r$p);
    }
  };
  const partition = function (p, param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return [
        /* Empty */ 0,
        /* Empty */ 0
      ];
    }
    const v = param.v;
    const match = partition(p, param.l);
    const lf = match[1];
    const lt = match[0];
    const pv = Curry._1(p, v);
    const match$1 = partition(p, param.r);
    const rf = match$1[1];
    const rt = match$1[0];
    if (pv) {
      return [
        join(lt, v, rt),
        concat(lf, rf)
      ];
    } else {
      return [
        concat(lt, rt),
        join(lf, v, rf)
      ];
    }
  };
  const cardinal = function (param) {
    if (/* tag */ typeof param !== "object" && typeof param !== "function") {
      return 0;
    } else {
      return (cardinal(param.l) + 1 | 0) + cardinal(param.r) | 0;
    }
  };
  const elements_aux = function (_accu, _param) {
    while (true) {
      const param = _param;
      const accu = _accu;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return accu;
      }
      _param = param.l;
      _accu = {
        hd: param.v,
        tl: elements_aux(accu, param.r)
      };
      continue;
    };
  };
  const elements = function (s) {
    return elements_aux(/* [] */ 0, s);
  };
  const find = function (x, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
            MEL_EXN_ID: Stdlib.Not_found
          });
      }
      const v = param.v;
      const c = Curry._2(funarg.compare, x, v);
      if (c === 0) {
        return v;
      }
      _param = c < 0 ? param.l : param.r;
      continue;
    };
  };
  const find_first = function (f, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
            MEL_EXN_ID: Stdlib.Not_found
          });
      }
      const v = param.v;
      if (Curry._1(f, v)) {
        let _v0 = v;
        let _param$1 = param.l;
        while (true) {
          const param$1 = _param$1;
          const v0 = _v0;
          if (/* tag */ typeof param$1 !== "object" && typeof param$1 !== "function") {
            return v0;
          }
          const v$1 = param$1.v;
          if (Curry._1(f, v$1)) {
            _param$1 = param$1.l;
            _v0 = v$1;
            continue;
          }
          _param$1 = param$1.r;
          continue;
        };
      }
      _param = param.r;
      continue;
    };
  };
  const find_first_opt = function (f, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return;
      }
      const v = param.v;
      if (Curry._1(f, v)) {
        let _v0 = v;
        let _param$1 = param.l;
        while (true) {
          const param$1 = _param$1;
          const v0 = _v0;
          if (/* tag */ typeof param$1 !== "object" && typeof param$1 !== "function") {
            return Caml_option.some(v0);
          }
          const v$1 = param$1.v;
          if (Curry._1(f, v$1)) {
            _param$1 = param$1.l;
            _v0 = v$1;
            continue;
          }
          _param$1 = param$1.r;
          continue;
        };
      }
      _param = param.r;
      continue;
    };
  };
  const find_last = function (f, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
            MEL_EXN_ID: Stdlib.Not_found
          });
      }
      const v = param.v;
      if (Curry._1(f, v)) {
        let _v0 = v;
        let _param$1 = param.r;
        while (true) {
          const param$1 = _param$1;
          const v0 = _v0;
          if (/* tag */ typeof param$1 !== "object" && typeof param$1 !== "function") {
            return v0;
          }
          const v$1 = param$1.v;
          if (Curry._1(f, v$1)) {
            _param$1 = param$1.r;
            _v0 = v$1;
            continue;
          }
          _param$1 = param$1.l;
          continue;
        };
      }
      _param = param.l;
      continue;
    };
  };
  const find_last_opt = function (f, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return;
      }
      const v = param.v;
      if (Curry._1(f, v)) {
        let _v0 = v;
        let _param$1 = param.r;
        while (true) {
          const param$1 = _param$1;
          const v0 = _v0;
          if (/* tag */ typeof param$1 !== "object" && typeof param$1 !== "function") {
            return Caml_option.some(v0);
          }
          const v$1 = param$1.v;
          if (Curry._1(f, v$1)) {
            _param$1 = param$1.r;
            _v0 = v$1;
            continue;
          }
          _param$1 = param$1.l;
          continue;
        };
      }
      _param = param.l;
      continue;
    };
  };
  const find_opt = function (x, _param) {
    while (true) {
      const param = _param;
      if (/* tag */ typeof param !== "object" && typeof param !== "function") {
        return;
      }
      const v = param.v;
      const c = Curry._2(funarg.compare, x, v);
      if (c === 0) {
        return Caml_option.some(v);
      }
      _param = c < 0 ? param.l : param.r;
      continue;
    };
  };
  const try_join = function (l, v, r) {
    if ((Caml_obj.caml_equal(l, /* Empty */ 0) || Curry._2(funarg.compare, max_elt(l), v) < 0) && (Caml_obj.caml_equal(r, /* Empty */ 0) || Curry._2(funarg.compare, v, min_elt(r)) < 0)) {
      return join(l, v, r);
    } else {
      return union(l, add(v, r));
    }
  };
  const map = function (f, t) {
    if (/* tag */ typeof t !== "object" && typeof t !== "function") {
      return /* Empty */ 0;
    }
    const r = t.r;
    const v = t.v;
    const l = t.l;
    const l$p = map(f, l);
    const v$p = Curry._1(f, v);
    const r$p = map(f, r);
    if (l === l$p && v === v$p && r === r$p) {
      return t;
    } else {
      return try_join(l$p, v$p, r$p);
    }
  };
  const filter_map = function (f, t) {
    if (/* tag */ typeof t !== "object" && typeof t !== "function") {
      return /* Empty */ 0;
    }
    const r = t.r;
    const v = t.v;
    const l = t.l;
    const l$p = filter_map(f, l);
    const v$p = Curry._1(f, v);
    const r$p = filter_map(f, r);
    if (v$p === undefined) {
      if (/* tag */ typeof l$p !== "object" && typeof l$p !== "function") {
        return r$p;
      } else if (/* tag */ typeof r$p !== "object" && typeof r$p !== "function") {
        return l$p;
      } else {
        return try_join(l$p, min_elt(r$p), remove_min_elt(r$p));
      }
    }
    const v$p$1 = Caml_option.valFromOption(v$p);
    if (l === l$p && v === v$p$1 && r === r$p) {
      return t;
    } else {
      return try_join(l$p, v$p$1, r$p);
    }
  };
  const of_list = function (l) {
    if (!l) {
      return /* Empty */ 0;
    }
    const match = l.tl;
    const x0 = l.hd;
    if (!match) {
      return singleton(x0);
    }
    const match$1 = match.tl;
    const x1 = match.hd;
    if (!match$1) {
      return add(x1, singleton(x0));
    }
    const match$2 = match$1.tl;
    const x2 = match$1.hd;
    if (!match$2) {
      return add(x2, add(x1, singleton(x0)));
    }
    const match$3 = match$2.tl;
    const x3 = match$2.hd;
    if (match$3) {
      if (match$3.tl) {
        let l$1 = Stdlib__List.sort_uniq(funarg.compare, l);
        const sub = function (n, l) {
          switch (n) {
            case 0 : {
              return [
                /* Empty */ 0,
                l
              ];
              }
            case 1 : {
              if (l) {
                return [
                  {
                    TAG: /* Node */ 0,
                    l: /* Empty */ 0,
                    v: l.hd,
                    r: /* Empty */ 0,
                    h: 1
                  },
                  l.tl
                ];
              }
              break;
              }
            case 2 : {
              if (l) {
                const match = l.tl;
                if (match) {
                  return [
                    {
                      TAG: /* Node */ 0,
                      l: {
                        TAG: /* Node */ 0,
                        l: /* Empty */ 0,
                        v: l.hd,
                        r: /* Empty */ 0,
                        h: 1
                      },
                      v: match.hd,
                      r: /* Empty */ 0,
                      h: 2
                    },
                    match.tl
                  ];
                }
                
              }
              break;
              }
            case 3 : {
              if (l) {
                const match$1 = l.tl;
                if (match$1) {
                  const match$2 = match$1.tl;
                  if (match$2) {
                    return [
                      {
                        TAG: /* Node */ 0,
                        l: {
                          TAG: /* Node */ 0,
                          l: /* Empty */ 0,
                          v: l.hd,
                          r: /* Empty */ 0,
                          h: 1
                        },
                        v: match$1.hd,
                        r: {
                          TAG: /* Node */ 0,
                          l: /* Empty */ 0,
                          v: match$2.hd,
                          r: /* Empty */ 0,
                          h: 1
                        },
                        h: 2
                      },
                      match$2.tl
                    ];
                  }
                  
                }
                
              }
              break;
              }
          }
          const nl = n / 2 | 0;
          const match$3 = sub(nl, l);
          const l$1 = match$3[1];
          if (l$1) {
            const match$4 = sub((n - nl | 0) - 1 | 0, l$1.tl);
            return [
              create(match$3[0], l$1.hd, match$4[0]),
              match$4[1]
            ];
          }
          throw new Caml_js_exceptions.MelangeError("Assert_failure", {
              MEL_EXN_ID: "Assert_failure",
              _1: [
                "jscomp/stdlib/set.ml",
                571,
                18
              ]
            });
        };
        return sub(Stdlib__List.length(l$1), l$1)[0];
      } else {
        return add(match$3.hd, add(x3, add(x2, add(x1, singleton(x0)))));
      }
    } else {
      return add(x3, add(x2, add(x1, singleton(x0))));
    }
  };
  const add_seq = function (i, m) {
    return Stdlib__Seq.fold_left((function (s, x) {
      return add(x, s);
    }), m, i);
  };
  const of_seq = function (i) {
    return add_seq(i, /* Empty */ 0);
  };
  const seq_of_enum_ = function (c, param) {
    if (/* tag */ typeof c !== "object" && typeof c !== "function") {
      return /* Nil */ 0;
    }
    const partial_arg = cons_enum(c._1, c._2);
    return {
      TAG: /* Cons */ 0,
      _0: c._0,
      _1: (function (param) {
        return seq_of_enum_(partial_arg, param);
      })
    };
  };
  const to_seq = function (c) {
    const partial_arg = cons_enum(c, /* End */ 0);
    return function (param) {
      return seq_of_enum_(partial_arg, param);
    };
  };
  const snoc_enum = function (_s, _e) {
    while (true) {
      const e = _e;
      const s = _s;
      if (/* tag */ typeof s !== "object" && typeof s !== "function") {
        return e;
      }
      _e = {
        TAG: /* More */ 0,
        _0: s.v,
        _1: s.l,
        _2: e
      };
      _s = s.r;
      continue;
    };
  };
  const rev_seq_of_enum_ = function (c, param) {
    if (/* tag */ typeof c !== "object" && typeof c !== "function") {
      return /* Nil */ 0;
    }
    const partial_arg = snoc_enum(c._1, c._2);
    return {
      TAG: /* Cons */ 0,
      _0: c._0,
      _1: (function (param) {
        return rev_seq_of_enum_(partial_arg, param);
      })
    };
  };
  const to_rev_seq = function (c) {
    const partial_arg = snoc_enum(c, /* End */ 0);
    return function (param) {
      return rev_seq_of_enum_(partial_arg, param);
    };
  };
  const to_seq_from = function (low, s) {
    const aux = function (low, _s, _c) {
      while (true) {
        const c = _c;
        const s = _s;
        if (/* tag */ typeof s !== "object" && typeof s !== "function") {
          return c;
        }
        const r = s.r;
        const v = s.v;
        const n = Curry._2(funarg.compare, v, low);
        if (n === 0) {
          return {
            TAG: /* More */ 0,
            _0: v,
            _1: r,
            _2: c
          };
        }
        if (n < 0) {
          _s = r;
          continue;
        }
        _c = {
          TAG: /* More */ 0,
          _0: v,
          _1: r,
          _2: c
        };
        _s = s.l;
        continue;
      };
    };
    const partial_arg = aux(low, s, /* End */ 0);
    return function (param) {
      return seq_of_enum_(partial_arg, param);
    };
  };
  return {
    empty: /* Empty */ 0,
    add: add,
    singleton: singleton,
    remove: remove,
    union: union,
    inter: inter,
    disjoint: disjoint,
    diff: diff,
    cardinal: cardinal,
    elements: elements,
    min_elt: min_elt,
    min_elt_opt: min_elt_opt,
    max_elt: max_elt,
    max_elt_opt: max_elt_opt,
    choose: min_elt,
    choose_opt: min_elt_opt,
    find: find,
    find_opt: find_opt,
    find_first: find_first,
    find_first_opt: find_first_opt,
    find_last: find_last,
    find_last_opt: find_last_opt,
    iter: iter,
    fold: fold,
    map: map,
    filter: filter,
    filter_map: filter_map,
    partition: partition,
    split: split,
    is_empty: is_empty,
    mem: mem,
    equal: equal,
    compare: compare,
    subset: subset,
    for_all: for_all,
    exists: exists,
    to_list: elements,
    of_list: of_list,
    to_seq_from: to_seq_from,
    to_seq: to_seq,
    to_rev_seq: to_rev_seq,
    add_seq: add_seq,
    of_seq: of_seq
  };
}

export {
  Make,
}
/* No side effect */
