// melange/string.mjs
// Generated by Melange

import * as Caml from "melange.js/caml.mjs";
import * as Caml_array from "melange.js/caml_array.mjs";
import * as Caml_bytes from "melange.js/caml_bytes.mjs";
import * as Caml_hash from "melange.js/caml_hash.mjs";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.mjs";
import * as Caml_string from "melange.js/caml_string.mjs";
import * as Curry from "melange.js/curry.mjs";
import * as Stdlib from "./stdlib.mjs";
import * as Stdlib__Array from "./array.mjs";
import * as Stdlib__Bytes from "./bytes.mjs";
import * as Stdlib__Char from "./char.mjs";
import * as Stdlib__Int from "./int.mjs";
import * as Stdlib__List from "./list.mjs";
import * as Stdlib__Uchar from "./uchar.mjs";

function make(n, c) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.make(n, c));
}

function init(n, f) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.init(n, f));
}

function of_char(c) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.make(1, c));
}

function sub(s, ofs, len) {
  if (ofs === 0 && s.length === len) {
    return s;
  } else {
    return Caml_bytes.bytes_to_string(Stdlib__Bytes.sub(Caml_bytes.bytes_of_string(s), ofs, len));
  }
}

function ensure_ge(x, y) {
  if (x >= y) {
    return x;
  }
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: "String.concat"
    });
}

function sum_lengths(_acc, seplen, _param) {
  while (true) {
    const param = _param;
    const acc = _acc;
    if (!param) {
      return acc;
    }
    const hd = param.hd;
    if (!param.tl) {
      return hd.length + acc | 0;
    }
    _param = param.tl;
    _acc = ensure_ge((hd.length + seplen | 0) + acc | 0, acc);
    continue;
  };
}

function unsafe_blits(dst, _pos, sep, seplen, _param) {
  while (true) {
    const param = _param;
    const pos = _pos;
    if (!param) {
      return dst;
    }
    const hd = param.hd;
    if (param.tl) {
      Caml_bytes.caml_blit_string(hd, 0, dst, pos, hd.length);
      Caml_bytes.caml_blit_string(sep, 0, dst, pos + hd.length | 0, seplen);
      _param = param.tl;
      _pos = (pos + hd.length | 0) + seplen | 0;
      continue;
    }
    Caml_bytes.caml_blit_string(hd, 0, dst, pos, hd.length);
    return dst;
  };
}

function concat(sep, l) {
  if (!l) {
    return "";
  }
  if (!l.tl) {
    return l.hd;
  }
  const seplen = sep.length;
  return Caml_bytes.bytes_to_string(unsafe_blits(Caml_bytes.caml_create_bytes(sum_lengths(0, seplen, l)), 0, sep, seplen, l));
}

function cat(prim0, prim1) {
  return prim0 + prim1;
}

function iter(f, s) {
  for (let i = 0, i_finish = s.length; i < i_finish; ++i) {
    Curry._1(f, s.charCodeAt(i));
  }
}

function iteri(f, s) {
  for (let i = 0, i_finish = s.length; i < i_finish; ++i) {
    Curry._2(f, i, s.charCodeAt(i));
  }
}

function map(f, s) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.map(f, Caml_bytes.bytes_of_string(s)));
}

function mapi(f, s) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.mapi(f, Caml_bytes.bytes_of_string(s)));
}

function fold_right(f, x, a) {
  return Stdlib__Bytes.fold_right(f, Caml_bytes.bytes_of_string(x), a);
}

function fold_left(f, a, x) {
  return Stdlib__Bytes.fold_left(f, a, Caml_bytes.bytes_of_string(x));
}

function exists(f, s) {
  return Stdlib__Bytes.exists(f, Caml_bytes.bytes_of_string(s));
}

function for_all(f, s) {
  return Stdlib__Bytes.for_all(f, Caml_bytes.bytes_of_string(s));
}

function is_space(param) {
  if (param > 13 || param < 9) {
    return param === 32;
  } else {
    return param !== 11;
  }
}

function trim(s) {
  if (s === "" || !(is_space(s.charCodeAt(0)) || is_space(s.charCodeAt(s.length - 1 | 0)))) {
    return s;
  } else {
    return Caml_bytes.bytes_to_string(Stdlib__Bytes.trim(Caml_bytes.bytes_of_string(s)));
  }
}

function escaped(s) {
  const b = Caml_bytes.bytes_of_string(s);
  const b$p = Stdlib__Bytes.unsafe_escape(b);
  if (b === b$p) {
    return s;
  } else {
    return Caml_bytes.bytes_to_string(b$p);
  }
}

function invalid_start(start, len) {
  const s = concat("", {
    hd: "start: ",
    tl: {
      hd: String(start),
      tl: {
        hd: " not in range [0;",
        tl: {
          hd: String(len),
          tl: {
            hd: "]",
            tl: /* [] */ 0
          }
        }
      }
    }
  });
  throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
      MEL_EXN_ID: "Invalid_argument",
      _1: s
    });
}

function find_first_index(sat, startOpt, s) {
  const start = startOpt !== undefined ? startOpt : 0;
  const len = s.length;
  if (!(0 <= start && start <= len)) {
    return invalid_start(start, len);
  }
  let i = start;
  while (i < len && !Curry._1(sat, s.charCodeAt(i))) {
    i = i + 1 | 0;
  };
  if (i < len) {
    return i;
  }
  
}

function find_last_index(sat, start, s) {
  const len = s.length;
  const start$1 = start !== undefined ? start : len;
  if (!(0 <= start$1 && start$1 <= len)) {
    return invalid_start(start$1, len);
  }
  let i = start$1 === len ? len - 1 | 0 : start$1;
  while (i >= 0 && !Curry._1(sat, s.charCodeAt(i))) {
    i = i - 1 | 0;
  };
  if (i < 0) {
    return;
  } else {
    return i;
  }
}

function index_rec(s, lim, _i, c) {
  while (true) {
    const i = _i;
    if (i >= lim) {
      throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
          MEL_EXN_ID: Stdlib.Not_found
        });
    }
    if (s.charCodeAt(i) === c) {
      return i;
    }
    _i = i + 1 | 0;
    continue;
  };
}

function index(s, c) {
  return index_rec(s, s.length, 0, c);
}

function index_rec_opt(s, lim, _i, c) {
  while (true) {
    const i = _i;
    if (i >= lim) {
      return;
    }
    if (s.charCodeAt(i) === c) {
      return i;
    }
    _i = i + 1 | 0;
    continue;
  };
}

function index_opt(s, c) {
  return index_rec_opt(s, s.length, 0, c);
}

function index_from(s, i, c) {
  const l = s.length;
  if (i < 0 || i > l) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "String.index_from / Bytes.index_from"
      });
  }
  return index_rec(s, l, i, c);
}

function index_from_opt(s, i, c) {
  const l = s.length;
  if (i < 0 || i > l) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "String.index_from_opt / Bytes.index_from_opt"
      });
  }
  return index_rec_opt(s, l, i, c);
}

function rindex_rec(s, _i, c) {
  while (true) {
    const i = _i;
    if (i < 0) {
      throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
          MEL_EXN_ID: Stdlib.Not_found
        });
    }
    if (s.charCodeAt(i) === c) {
      return i;
    }
    _i = i - 1 | 0;
    continue;
  };
}

function rindex(s, c) {
  return rindex_rec(s, s.length - 1 | 0, c);
}

function rindex_from(s, i, c) {
  if (i < -1 || i >= s.length) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "String.rindex_from / Bytes.rindex_from"
      });
  }
  return rindex_rec(s, i, c);
}

function rindex_rec_opt(s, _i, c) {
  while (true) {
    const i = _i;
    if (i < 0) {
      return;
    }
    if (s.charCodeAt(i) === c) {
      return i;
    }
    _i = i - 1 | 0;
    continue;
  };
}

function rindex_opt(s, c) {
  return rindex_rec_opt(s, s.length - 1 | 0, c);
}

function rindex_from_opt(s, i, c) {
  if (i < -1 || i >= s.length) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "String.rindex_from_opt / Bytes.rindex_from_opt"
      });
  }
  return rindex_rec_opt(s, i, c);
}

function contains_from(s, i, c) {
  const l = s.length;
  if (i < 0 || i > l) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "String.contains_from / Bytes.contains_from"
      });
  }
  try {
    index_rec(s, l, i, c);
    return true;
  }
  catch (raw_exn){
    const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.MEL_EXN_ID === Stdlib.Not_found) {
      return false;
    }
    throw exn;
  }
}

function contains(s, c) {
  return contains_from(s, 0, c);
}

function rcontains_from(s, i, c) {
  if (i < 0 || i >= s.length) {
    throw new Caml_js_exceptions.MelangeError("Invalid_argument", {
        MEL_EXN_ID: "Invalid_argument",
        _1: "String.rcontains_from / Bytes.rcontains_from"
      });
  }
  try {
    rindex_rec(s, i, c);
    return true;
  }
  catch (raw_exn){
    const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.MEL_EXN_ID === Stdlib.Not_found) {
      return false;
    }
    throw exn;
  }
}

function find_maximal_suffix_and_period(sub) {
  const sublen = sub.length;
  let i = -1;
  let j = 0;
  let k = 1;
  let p = 1;
  while ((j + k | 0) < sublen) {
    const c = ((Caml_string.get(sub, j + k | 0) - Caml_string.get(sub, i + k | 0) | 0) << 0);
    if (c < 0) {
      j = j + k | 0;
      k = 1;
      p = j - i | 0;
    } else if (c > 0) {
      i = j;
      j = i + 1 | 0;
      k = 1;
      p = 1;
    } else if (k === p) {
      j = j + p | 0;
      k = 1;
    } else {
      k = k + 1 | 0;
    }
  };
  const l0 = i;
  const p0 = p;
  i = -1;
  j = 0;
  k = 1;
  p = 1;
  while ((j + k | 0) < sublen) {
    const c$1 = Math.imul(-1, Caml_string.get(sub, j + k | 0) - Caml_string.get(sub, i + k | 0) | 0);
    if (c$1 < 0) {
      j = j + k | 0;
      k = 1;
      p = j - i | 0;
    } else if (c$1 > 0) {
      i = j;
      j = i + 1 | 0;
      k = 1;
      p = 1;
    } else if (k === p) {
      j = j + p | 0;
      k = 1;
    } else {
      k = k + 1 | 0;
    }
  };
  const l1 = i;
  const p1 = p;
  if (l0 > l1) {
    return [
      l0,
      p0
    ];
  } else {
    return [
      l1,
      p1
    ];
  }
}

function is_sub_periodic(sub, param) {
  const p = param[1];
  const l = param[0];
  if (l > p) {
    return false;
  }
  let i = 0;
  while (i <= l && Stdlib__Char.equal(Caml_string.get(sub, i), Caml_string.get(sub, i + p | 0))) {
    i = i + 1 | 0;
  };
  return i > l;
}

function find(start, sub, param, sub_periodic, s) {
  const p = param[1];
  const l = param[0];
  const slen = s.length;
  const sublen = sub.length;
  if (!(0 <= start && start <= slen)) {
    return invalid_start(start, slen);
  }
  const smax = slen - sublen | 0;
  let j = start;
  try {
    if (sub_periodic) {
      let memory = -1;
      while (j <= smax) {
        let i = 1 + Stdlib__Int.max(l, memory) | 0;
        while (i < sublen && Stdlib__Char.equal(Caml_string.get(sub, i), Caml_string.get(s, i + j | 0))) {
          i = i + 1 | 0;
        };
        if (i < sublen) {
          j = j + (i - l | 0) | 0;
          memory = -1;
        } else {
          i = l;
          while (i > memory && Stdlib__Char.equal(Caml_string.get(sub, i), Caml_string.get(s, i + j | 0))) {
            i = i - 1 | 0;
          };
          if (i <= memory) {
            throw new Caml_js_exceptions.MelangeError(Stdlib.Exit, {
                MEL_EXN_ID: Stdlib.Exit
              });
          }
          j = j + p | 0;
          memory = (sublen - p | 0) - 1 | 0;
        }
      };
      return -1;
    }
    const q = 1 + Stdlib__Int.max(l + 1 | 0, (sublen - l | 0) - 1 | 0) | 0;
    while (j <= smax) {
      let i$1 = l + 1 | 0;
      while (i$1 < sublen && Stdlib__Char.equal(Caml_string.get(sub, i$1), Caml_string.get(s, i$1 + j | 0))) {
        i$1 = i$1 + 1 | 0;
      };
      if (i$1 < sublen) {
        j = j + (i$1 - l | 0) | 0;
      } else {
        i$1 = l;
        while (i$1 >= 0 && Stdlib__Char.equal(Caml_string.get(sub, i$1), Caml_string.get(s, i$1 + j | 0))) {
          i$1 = i$1 - 1 | 0;
        };
        if (i$1 < 0) {
          throw new Caml_js_exceptions.MelangeError(Stdlib.Exit, {
              MEL_EXN_ID: Stdlib.Exit
            });
        }
        j = j + q | 0;
      }
    };
    return -1;
  }
  catch (raw_exn){
    const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.MEL_EXN_ID === Stdlib.Exit) {
      return j;
    }
    throw exn;
  }
}

function rfind_maximal_suffix_and_period(sub) {
  const sublen = sub.length;
  let i = -1;
  let j = 0;
  let k = 1;
  let p = 1;
  while ((j + k | 0) < sublen) {
    const c = ((Caml_string.get(sub, (sub.length - 1 | 0) - (j + k | 0) | 0) - Caml_string.get(sub, (sub.length - 1 | 0) - (i + k | 0) | 0) | 0) << 0);
    if (c < 0) {
      j = j + k | 0;
      k = 1;
      p = j - i | 0;
    } else if (c > 0) {
      i = j;
      j = i + 1 | 0;
      k = 1;
      p = 1;
    } else if (k === p) {
      j = j + p | 0;
      k = 1;
    } else {
      k = k + 1 | 0;
    }
  };
  const l0 = i;
  const p0 = p;
  i = -1;
  j = 0;
  k = 1;
  p = 1;
  while ((j + k | 0) < sublen) {
    const c$1 = Math.imul(-1, Caml_string.get(sub, (sub.length - 1 | 0) - (j + k | 0) | 0) - Caml_string.get(sub, (sub.length - 1 | 0) - (i + k | 0) | 0) | 0);
    if (c$1 < 0) {
      j = j + k | 0;
      k = 1;
      p = j - i | 0;
    } else if (c$1 > 0) {
      i = j;
      j = i + 1 | 0;
      k = 1;
      p = 1;
    } else if (k === p) {
      j = j + p | 0;
      k = 1;
    } else {
      k = k + 1 | 0;
    }
  };
  const l1 = i;
  const p1 = p;
  if (l0 > l1) {
    return [
      l0,
      p0
    ];
  } else {
    return [
      l1,
      p1
    ];
  }
}

function ris_sub_periodic(sub, param) {
  const p = param[1];
  const l = param[0];
  if (l > p) {
    return false;
  }
  let i = 0;
  while (i <= l && Stdlib__Char.equal(Caml_string.get(sub, (sub.length - 1 | 0) - i | 0), Caml_string.get(sub, (sub.length - 1 | 0) - (i + p | 0) | 0))) {
    i = i + 1 | 0;
  };
  return i > l;
}

function rfind(start, sub, param, rsub_periodic, s) {
  const p = param[1];
  const l = param[0];
  const slen = s.length;
  const sublen = sub.length;
  if (!(0 <= start && start <= slen)) {
    return invalid_start(start, slen);
  }
  const start$1 = Stdlib__Int.max(0, (slen - 1 | 0) - (start + (sublen - 1 | 0) | 0) | 0);
  const smax = slen - sublen | 0;
  let j = start$1;
  try {
    if (rsub_periodic) {
      let memory = -1;
      while (j <= smax) {
        let i = 1 + Stdlib__Int.max(l, memory) | 0;
        while (i < sublen && Stdlib__Char.equal(Caml_string.get(sub, (sub.length - 1 | 0) - i | 0), Caml_string.get(s, (s.length - 1 | 0) - (i + j | 0) | 0))) {
          i = i + 1 | 0;
        };
        if (i < sublen) {
          j = j + (i - l | 0) | 0;
          memory = -1;
        } else {
          i = l;
          while (i > memory && Stdlib__Char.equal(Caml_string.get(sub, (sub.length - 1 | 0) - i | 0), Caml_string.get(s, (s.length - 1 | 0) - (i + j | 0) | 0))) {
            i = i - 1 | 0;
          };
          if (i <= memory) {
            throw new Caml_js_exceptions.MelangeError(Stdlib.Exit, {
                MEL_EXN_ID: Stdlib.Exit
              });
          }
          j = j + p | 0;
          memory = (sublen - p | 0) - 1 | 0;
        }
      };
      return -1;
    }
    const q = 1 + Stdlib__Int.max(l + 1 | 0, (sublen - l | 0) - 1 | 0) | 0;
    while (j <= smax) {
      let i$1 = l + 1 | 0;
      while (i$1 < sublen && Stdlib__Char.equal(Caml_string.get(sub, (sub.length - 1 | 0) - i$1 | 0), Caml_string.get(s, (s.length - 1 | 0) - (i$1 + j | 0) | 0))) {
        i$1 = i$1 + 1 | 0;
      };
      if (i$1 < sublen) {
        j = j + (i$1 - l | 0) | 0;
      } else {
        i$1 = l;
        while (i$1 >= 0 && Stdlib__Char.equal(Caml_string.get(sub, (sub.length - 1 | 0) - i$1 | 0), Caml_string.get(s, (s.length - 1 | 0) - (i$1 + j | 0) | 0))) {
          i$1 = i$1 - 1 | 0;
        };
        if (i$1 < 0) {
          throw new Caml_js_exceptions.MelangeError(Stdlib.Exit, {
              MEL_EXN_ID: Stdlib.Exit
            });
        }
        j = j + q | 0;
      }
    };
    return -1;
  }
  catch (raw_exn){
    const exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.MEL_EXN_ID === Stdlib.Exit) {
      return (slen - 1 | 0) - (j + (sublen - 1 | 0) | 0) | 0;
    }
    throw exn;
  }
}

function find_first(sub) {
  const sub_lp = find_maximal_suffix_and_period(sub);
  const sub_periodic = is_sub_periodic(sub, sub_lp);
  return function (startOpt, s) {
    const start = startOpt !== undefined ? startOpt : 0;
    const i = find(start, sub, sub_lp, sub_periodic, s);
    if (i !== -1) {
      return i;
    }
    
  };
}

function find_last(sub) {
  const rsub_lp = rfind_maximal_suffix_and_period(sub);
  const rsub_periodic = ris_sub_periodic(sub, rsub_lp);
  return function (start, s) {
    const start$1 = start !== undefined ? start : s.length;
    const i = rfind(start$1, sub, rsub_lp, rsub_periodic, s);
    if (i !== -1) {
      return i;
    }
    
  };
}

function find_all(sub) {
  const sub_lp = find_maximal_suffix_and_period(sub);
  const sub_periodic = is_sub_periodic(sub, sub_lp);
  return function (f, startOpt, s, acc) {
    const start = startOpt !== undefined ? startOpt : 0;
    const slen = s.length;
    if (0 <= start && start <= slen) {
      let _acc = acc;
      let _start = start;
      while (true) {
        const start$1 = _start;
        const acc$1 = _acc;
        if (start$1 > slen) {
          return acc$1;
        }
        const i = find(start$1, sub, sub_lp, sub_periodic, s);
        if (i === -1) {
          return acc$1;
        }
        const start$2 = i + Stdlib__Int.max(sub.length, 1) | 0;
        _start = start$2;
        _acc = Curry._2(f, i, acc$1);
        continue;
      };
    } else {
      return invalid_start(start, slen);
    }
  };
}

function rfind_all(sub) {
  const rsub_lp = rfind_maximal_suffix_and_period(sub);
  const rsub_periodic = ris_sub_periodic(sub, rsub_lp);
  return function (f, start, s, acc) {
    const slen = s.length;
    const start$1 = start !== undefined ? start : s.length;
    if (0 <= start$1 && start$1 <= slen) {
      let _acc = acc;
      let _start = start$1;
      while (true) {
        const start$2 = _start;
        const acc$1 = _acc;
        if (start$2 < 0) {
          return acc$1;
        }
        const i = rfind(start$2, sub, rsub_lp, rsub_periodic, s);
        if (i === -1) {
          return acc$1;
        }
        const start$3 = i - Stdlib__Int.max(sub.length, 1) | 0;
        _start = start$3;
        _acc = Curry._2(f, i, acc$1);
        continue;
      };
    } else {
      return invalid_start(start$1, slen);
    }
  };
}

function replace_first(needle) {
  const find_first$1 = find_first(needle);
  return function (by, start, s) {
    const i = Curry._2(find_first$1, start, s);
    if (i === undefined) {
      return s;
    }
    const rest_first = i + needle.length | 0;
    const rest_len = (s.length - i | 0) - needle.length | 0;
    return concat(by, {
      hd: sub(s, 0, i),
      tl: {
        hd: sub(s, rest_first, rest_len),
        tl: /* [] */ 0
      }
    });
  };
}

function replace_last(needle) {
  const find_last$1 = find_last(needle);
  return function (by, start, s) {
    const i = Curry._2(find_last$1, start, s);
    if (i === undefined) {
      return s;
    }
    const rest_first = i + needle.length | 0;
    const rest_len = (s.length - i | 0) - needle.length | 0;
    return concat(by, {
      hd: sub(s, 0, i),
      tl: {
        hd: sub(s, rest_first, rest_len),
        tl: /* [] */ 0
      }
    });
  };
}

function replace_all(needle) {
  const find_all$1 = find_all(needle);
  return function (by, start, s) {
    const chunk_first = {
      contents: 0
    };
    const add_chunk = function (i, acc) {
      const acc_0 = sub(s, chunk_first.contents, i - chunk_first.contents | 0);
      const acc$1 = {
        hd: acc_0,
        tl: acc
      };
      chunk_first.contents = i + needle.length | 0;
      return acc$1;
    };
    const chunks = Curry._4(find_all$1, add_chunk, start, s, /* [] */ 0);
    if (!chunks) {
      return s;
    }
    const chunks_0 = sub(s, chunk_first.contents, s.length - chunk_first.contents | 0);
    const chunks$1 = {
      hd: chunks_0,
      tl: chunks
    };
    return concat(by, Stdlib__List.rev(chunks$1));
  };
}

function uppercase_ascii(s) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.uppercase_ascii(Caml_bytes.bytes_of_string(s)));
}

function lowercase_ascii(s) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.lowercase_ascii(Caml_bytes.bytes_of_string(s)));
}

function capitalize_ascii(s) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.capitalize_ascii(Caml_bytes.bytes_of_string(s)));
}

function uncapitalize_ascii(s) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.uncapitalize_ascii(Caml_bytes.bytes_of_string(s)));
}

function starts_with(prefix, s) {
  const len_s = s.length;
  const len_pre = prefix.length;
  if (len_s >= len_pre) {
    let _i = 0;
    while (true) {
      const i = _i;
      if (i === len_pre) {
        return true;
      }
      if (s.charCodeAt(i) !== prefix.charCodeAt(i)) {
        return false;
      }
      _i = i + 1 | 0;
      continue;
    };
  } else {
    return false;
  }
}

function ends_with(suffix, s) {
  const len_s = s.length;
  const len_suf = suffix.length;
  const diff = len_s - len_suf | 0;
  if (diff >= 0) {
    let _i = 0;
    while (true) {
      const i = _i;
      if (i === len_suf) {
        return true;
      }
      if (s.charCodeAt(diff + i | 0) !== suffix.charCodeAt(i)) {
        return false;
      }
      _i = i + 1 | 0;
      continue;
    };
  } else {
    return false;
  }
}

function includes(sub) {
  const sub_lp = find_maximal_suffix_and_period(sub);
  const sub_periodic = is_sub_periodic(sub, sub_lp);
  return function (s) {
    return find(0, sub, sub_lp, sub_periodic, s) !== -1;
  };
}

function hash(x) {
  return Caml_hash.caml_string_hash(0, x);
}

function take_first(n, s) {
  const lastOpt = n - 1 | 0;
  const first = 0;
  const max = s.length - 1 | 0;
  const first$1 = first < 0 ? 0 : first;
  const last = lastOpt > max ? max : lastOpt;
  if (first$1 > last) {
    return "";
  } else {
    return sub(s, first$1, (last - first$1 | 0) + 1 | 0);
  }
}

function drop_first(n, s) {
  const firstOpt = n;
  const last = Stdlib.max_int;
  const max = s.length - 1 | 0;
  const first = firstOpt < 0 ? 0 : firstOpt;
  const last$1 = last > max ? max : last;
  if (first > last$1) {
    return "";
  } else {
    return sub(s, first, (last$1 - first | 0) + 1 | 0);
  }
}

function cut_first(n, s) {
  return [
    take_first(n, s),
    drop_first(n, s)
  ];
}

function take_last(n, s) {
  const firstOpt = s.length - n | 0;
  const last = Stdlib.max_int;
  const max = s.length - 1 | 0;
  const first = firstOpt < 0 ? 0 : firstOpt;
  const last$1 = last > max ? max : last;
  if (first > last$1) {
    return "";
  } else {
    return sub(s, first, (last$1 - first | 0) + 1 | 0);
  }
}

function drop_last(n, s) {
  const lastOpt = (s.length - n | 0) - 1 | 0;
  const first = 0;
  const max = s.length - 1 | 0;
  const first$1 = first < 0 ? 0 : first;
  const last = lastOpt > max ? max : lastOpt;
  if (first$1 > last) {
    return "";
  } else {
    return sub(s, first$1, (last - first$1 | 0) + 1 | 0);
  }
}

function cut_last(n, s) {
  return [
    drop_last(n, s),
    take_last(n, s)
  ];
}

function take_first_while(sat, s) {
  const len = s.length;
  let i = 0;
  while (i < len && Curry._1(sat, s.charCodeAt(i))) {
    i = i + 1 | 0;
  };
  if (i === len) {
    return s;
  } else {
    return sub(s, 0, i);
  }
}

function drop_first_while(sat, s) {
  const len = s.length;
  let i = 0;
  while (i < len && Curry._1(sat, s.charCodeAt(i))) {
    i = i + 1 | 0;
  };
  if (i === 0) {
    return s;
  } else {
    return sub(s, i, len - i | 0);
  }
}

function cut_first_while(sat, s) {
  const len = s.length;
  let i = 0;
  while (i < len && Curry._1(sat, s.charCodeAt(i))) {
    i = i + 1 | 0;
  };
  if (i === len) {
    return [
      s,
      ""
    ];
  } else if (i === 0) {
    return [
      "",
      s
    ];
  } else {
    return [
      sub(s, 0, i),
      sub(s, i, len - i | 0)
    ];
  }
}

function take_last_while(sat, s) {
  const len = s.length;
  let i = len - 1 | 0;
  while (i >= 0 && Curry._1(sat, s.charCodeAt(i))) {
    i = i - 1 | 0;
  };
  if (i < 0) {
    return s;
  }
  const j = i + 1 | 0;
  return sub(s, j, len - j | 0);
}

function drop_last_while(sat, s) {
  const len = s.length;
  let i = len - 1 | 0;
  while (i >= 0 && Curry._1(sat, s.charCodeAt(i))) {
    i = i - 1 | 0;
  };
  if (i < 0) {
    return "";
  } else {
    return sub(s, 0, i + 1 | 0);
  }
}

function cut_last_while(sat, s) {
  const len = s.length;
  let i = len - 1 | 0;
  while (i >= 0 && Curry._1(sat, s.charCodeAt(i))) {
    i = i - 1 | 0;
  };
  if (i < 0) {
    return [
      "",
      s
    ];
  }
  if (i === (len - 1 | 0)) {
    return [
      s,
      ""
    ];
  }
  const j = i + 1 | 0;
  return [
    sub(s, 0, j),
    sub(s, j, len - j | 0)
  ];
}

function split_first(sep) {
  const find_first$1 = find_first(sep);
  return function (s) {
    const i = Curry._2(find_first$1, undefined, s);
    if (i === undefined) {
      return;
    }
    const lastOpt = i - 1 | 0;
    const first = 0;
    const max = s.length - 1 | 0;
    const first$1 = first < 0 ? 0 : first;
    const last = lastOpt > max ? max : lastOpt;
    const firstOpt = i + sep.length | 0;
    const last$1 = Stdlib.max_int;
    const max$1 = s.length - 1 | 0;
    const first$2 = firstOpt < 0 ? 0 : firstOpt;
    const last$2 = last$1 > max$1 ? max$1 : last$1;
    return [
      first$1 > last ? "" : sub(s, first$1, (last - first$1 | 0) + 1 | 0),
      first$2 > last$2 ? "" : sub(s, first$2, (last$2 - first$2 | 0) + 1 | 0)
    ];
  };
}

function split_last(sep) {
  const find_last$1 = find_last(sep);
  return function (s) {
    const i = Curry._2(find_last$1, undefined, s);
    if (i === undefined) {
      return;
    }
    const lastOpt = i - 1 | 0;
    const first = 0;
    const max = s.length - 1 | 0;
    const first$1 = first < 0 ? 0 : first;
    const last = lastOpt > max ? max : lastOpt;
    const firstOpt = i + sep.length | 0;
    const last$1 = Stdlib.max_int;
    const max$1 = s.length - 1 | 0;
    const first$2 = firstOpt < 0 ? 0 : firstOpt;
    const last$2 = last$1 > max$1 ? max$1 : last$1;
    return [
      first$1 > last ? "" : sub(s, first$1, (last - first$1 | 0) + 1 | 0),
      first$2 > last$2 ? "" : sub(s, first$2, (last$2 - first$2 | 0) + 1 | 0)
    ];
  };
}

function split_all(sep) {
  const find_all$1 = find_all(sep);
  return function (dropOpt, s) {
    const drop = dropOpt !== undefined ? dropOpt : (function (param) {
        return false;
      });
    const first = {
      contents: 0
    };
    const add_token = function (i, acc) {
      const lastOpt = i - 1 | 0;
      const firstOpt = first.contents;
      const max = s.length - 1 | 0;
      const first$1 = firstOpt < 0 ? 0 : firstOpt;
      const last = lastOpt > max ? max : lastOpt;
      const token = first$1 > last ? "" : sub(s, first$1, (last - first$1 | 0) + 1 | 0);
      first.contents = i + sep.length | 0;
      if (Curry._1(drop, token)) {
        return acc;
      } else {
        return {
          hd: token,
          tl: acc
        };
      }
    };
    const tokens = Curry._4(find_all$1, add_token, undefined, s, /* [] */ 0);
    const firstOpt = first.contents;
    const last = Stdlib.max_int;
    const max = s.length - 1 | 0;
    const first$1 = firstOpt < 0 ? 0 : firstOpt;
    const last$1 = last > max ? max : last;
    const last$2 = first$1 > last$1 ? "" : sub(s, first$1, (last$1 - first$1 | 0) + 1 | 0);
    return Stdlib__List.rev(Curry._1(drop, last$2) ? tokens : ({
        hd: last$2,
        tl: tokens
      }));
  };
}

function rsplit_all(sep) {
  const rfind_all$1 = rfind_all(sep);
  return function (dropOpt, s) {
    const drop = dropOpt !== undefined ? dropOpt : (function (param) {
        return false;
      });
    const last = {
      contents: s.length - 1 | 0
    };
    const add_token = function (i, acc) {
      const lastOpt = last.contents;
      const firstOpt = i + sep.length | 0;
      const max = s.length - 1 | 0;
      const first = firstOpt < 0 ? 0 : firstOpt;
      const last$1 = lastOpt > max ? max : lastOpt;
      const token = first > last$1 ? "" : sub(s, first, (last$1 - first | 0) + 1 | 0);
      last.contents = i - 1 | 0;
      if (Curry._1(drop, token)) {
        return acc;
      } else {
        return {
          hd: token,
          tl: acc
        };
      }
    };
    const tokens = Curry._4(rfind_all$1, add_token, undefined, s, /* [] */ 0);
    const lastOpt = last.contents;
    const first = 0;
    const max = s.length - 1 | 0;
    const first$1 = first < 0 ? 0 : first;
    const last$1 = lastOpt > max ? max : lastOpt;
    const last$2 = first$1 > last$1 ? "" : sub(s, first$1, (last$1 - first$1 | 0) + 1 | 0);
    if (Curry._1(drop, last$2)) {
      return tokens;
    } else {
      return {
        hd: last$2,
        tl: tokens
      };
    }
  };
}

function split_on_char(sep, s) {
  let r = /* [] */ 0;
  let j = s.length;
  for (let i = s.length - 1 | 0; i >= 0; --i) {
    if (s.charCodeAt(i) === sep) {
      r = {
        hd: sub(s, i + 1 | 0, (j - i | 0) - 1 | 0),
        tl: r
      };
      j = i;
    }
    
  }
  return {
    hd: sub(s, 0, j),
    tl: r
  };
}

const compare = Caml.caml_string_compare;

function is_empty(s) {
  return s.length === 0;
}

function to_seq(s) {
  return Stdlib__Bytes.to_seq(Caml_bytes.bytes_of_string(s));
}

function to_seqi(s) {
  return Stdlib__Bytes.to_seqi(Caml_bytes.bytes_of_string(s));
}

function of_seq(g) {
  return Caml_bytes.bytes_to_string(Stdlib__Bytes.of_seq(g));
}

function get_utf_8_uchar(s, i) {
  return Stdlib__Bytes.get_utf_8_uchar(Caml_bytes.bytes_of_string(s), i);
}

function is_valid_utf_8(s) {
  return Stdlib__Bytes.is_valid_utf_8(Caml_bytes.bytes_of_string(s));
}

function get_utf_16be_uchar(s, i) {
  return Stdlib__Bytes.get_utf_16be_uchar(Caml_bytes.bytes_of_string(s), i);
}

function is_valid_utf_16be(s) {
  return Stdlib__Bytes.is_valid_utf_16be(Caml_bytes.bytes_of_string(s));
}

function get_utf_16le_uchar(s, i) {
  return Stdlib__Bytes.get_utf_16le_uchar(Caml_bytes.bytes_of_string(s), i);
}

function is_valid_utf_16le(s) {
  return Stdlib__Bytes.is_valid_utf_16le(Caml_bytes.bytes_of_string(s));
}

function get_int8(s, i) {
  return Stdlib__Bytes.get_int8(Caml_bytes.bytes_of_string(s), i);
}

function get_uint16_le(s, i) {
  return Stdlib__Bytes.get_uint16_le(Caml_bytes.bytes_of_string(s), i);
}

function get_uint16_be(s, i) {
  return Stdlib__Bytes.get_uint16_be(Caml_bytes.bytes_of_string(s), i);
}

function get_int16_ne(s, i) {
  return Stdlib__Bytes.get_int16_ne(Caml_bytes.bytes_of_string(s), i);
}

function get_int16_le(s, i) {
  return Stdlib__Bytes.get_int16_le(Caml_bytes.bytes_of_string(s), i);
}

function get_int16_be(s, i) {
  return Stdlib__Bytes.get_int16_be(Caml_bytes.bytes_of_string(s), i);
}

function get_int32_le(s, i) {
  return Stdlib__Bytes.get_int32_le(Caml_bytes.bytes_of_string(s), i);
}

function get_int32_be(s, i) {
  return Stdlib__Bytes.get_int32_be(Caml_bytes.bytes_of_string(s), i);
}

function get_int64_le(s, i) {
  return Stdlib__Bytes.get_int64_le(Caml_bytes.bytes_of_string(s), i);
}

function get_int64_be(s, i) {
  return Stdlib__Bytes.get_int64_be(Caml_bytes.bytes_of_string(s), i);
}

function utf_8_uchar_length(s) {
  const slen = s.length;
  let i = 0;
  let ulen = 0;
  while (i < slen) {
    const dec_len = Stdlib__Uchar.utf_8_decode_length_of_byte(s.charCodeAt(i));
    i = i + (
      dec_len === 0 ? 1 : dec_len
    ) | 0;
    ulen = ulen + 1 | 0;
  };
  return ulen;
}

function uchar_array_of_utf_8_string(s) {
  const slen = s.length;
  const uchars = Caml_array.make(slen, Stdlib__Uchar.max);
  let k = 0;
  let i = 0;
  while (i < slen) {
    const dec = Stdlib__Bytes.get_utf_8_uchar(Caml_bytes.bytes_of_string(s), i);
    i = i + ((dec >>> 24) & 7) | 0;
    Caml_array.set(uchars, k, dec & 16777215);
    k = k + 1 | 0;
  };
  return [
    uchars,
    k
  ];
}

function edit_distance$p(limitOpt, s, param, s1) {
  const limit = limitOpt !== undefined ? limitOpt : Stdlib__Int.max_int;
  const len0 = param[1];
  const s0 = param[0];
  if (limit <= 1) {
    if (s === s1) {
      return 0;
    } else {
      return limit;
    }
  }
  const match = uchar_array_of_utf_8_string(s1);
  const len1 = match[1];
  const s1$1 = match[0];
  const limit$1 = Stdlib__Int.min(Stdlib__Int.max(len0, len1), limit);
  if (Stdlib__Int.abs(len1 - len0 | 0) >= limit$1) {
    return limit$1;
  }
  const match$1 = len0 > len1 ? [
      s0,
      s1$1
    ] : [
      s1$1,
      s0
    ];
  const match$2 = len0 > len1 ? [
      len0,
      len1
    ] : [
      len1,
      len0
    ];
  const len1$1 = match$2[1];
  const loop = function (_row_minus2, _row_minus1, _row, _i, len0, limit, s0, s1) {
    while (true) {
      const i = _i;
      const row = _row;
      const row_minus1 = _row_minus1;
      const row_minus2 = _row_minus2;
      if (i > len0) {
        return Caml_array.get(row_minus1, row_minus1.length - 1 | 0);
      }
      const len1 = row.length - 1 | 0;
      let row_min = Stdlib__Int.max_int;
      Caml_array.set(row, 0, i);
      const jmax = Stdlib__Int.min(len1, (i + limit | 0) - 1 | 0);
      const jmax$1 = jmax < 0 ? len1 : jmax;
      for (let j = Stdlib__Int.max(1, i - limit | 0); j <= jmax$1; ++j) {
        const cost = Caml_array.get(s0, i - 1 | 0) === Caml_array.get(s1, j - 1 | 0) ? 0 : 1;
        const min = Stdlib__Int.min(Caml_array.get(row_minus1, j - 1 | 0) + cost | 0, Stdlib__Int.min(Caml_array.get(row_minus1, j) + 1 | 0, Caml_array.get(row, j - 1 | 0) + 1 | 0));
        const min$1 = i > 1 && j > 1 && Caml_array.get(s0, i - 1 | 0) === Caml_array.get(s1, j - 2 | 0) && Caml_array.get(s0, i - 2 | 0) === Caml_array.get(s1, j - 1 | 0) ? Stdlib__Int.min(min, Caml_array.get(row_minus2, j - 2 | 0) + cost | 0) : min;
        Caml_array.set(row, j, min$1);
        row_min = Stdlib__Int.min(row_min, min$1);
      }
      if (row_min >= limit) {
        return limit;
      }
      _i = i + 1 | 0;
      _row = row_minus2;
      _row_minus1 = row;
      _row_minus2 = row_minus1;
      continue;
    };
  };
  const ignore = limit$1 + 1 | 0;
  const row_minus2 = Caml_array.make(len1$1 + 1 | 0, ignore);
  const row_minus1 = Stdlib__Array.init(len1$1 + 1 | 0, (function (x) {
    return x;
  }));
  const row = Caml_array.make(len1$1 + 1 | 0, ignore);
  const d = loop(row_minus2, row_minus1, row, 1, match$2[0], limit$1, match$1[0], match$1[1]);
  if (d > limit$1) {
    return limit$1;
  } else {
    return d;
  }
}

function edit_distance(limit, s0, s1) {
  const us0 = uchar_array_of_utf_8_string(s0);
  return edit_distance$p(limit, s0, us0, s1);
}

function default_max_dist(s) {
  const match = utf_8_uchar_length(s);
  if (match >= 3) {
    if (match >= 5) {
      return 2;
    } else {
      return 1;
    }
  } else if (match >= 0) {
    return 0;
  } else {
    return 2;
  }
}

function spellcheck(max_distOpt, iter_dict, s) {
  const max_dist = max_distOpt !== undefined ? max_distOpt : default_max_dist;
  const min = {
    contents: Curry._1(max_dist, s)
  };
  const acc = {
    contents: /* [] */ 0
  };
  const us = uchar_array_of_utf_8_string(s);
  Curry._1(iter_dict, (function (param) {
    const d = edit_distance$p(min.contents + 1 | 0, s, us, param);
    if (d === min.contents) {
      acc.contents = {
        hd: param,
        tl: acc.contents
      };
      return;
    } else if (d < min.contents) {
      min.contents = d;
      acc.contents = {
        hd: param,
        tl: /* [] */ 0
      };
      return;
    } else {
      return;
    }
  }));
  return Stdlib__List.rev(acc.contents);
}

const empty = "";

const of_bytes = Stdlib__Bytes.to_string;

const to_bytes = Stdlib__Bytes.of_string;

const blit = Stdlib__Bytes.blit_string;

function equal(prim0, prim1) {
  return prim0 === prim1;
}

const get_uint8 = Caml_string.get;

const get_uint16_ne = Caml_bytes.get16;

const get_int32_ne = Caml_bytes.get32;

const seeded_hash = Caml_hash.caml_string_hash;

const get_int64_ne = Caml_bytes.get64;

export {
  make,
  init,
  empty,
  of_char,
  of_bytes,
  to_bytes,
  blit,
  concat,
  cat,
  equal,
  compare,
  is_empty,
  starts_with,
  ends_with,
  includes,
  contains_from,
  rcontains_from,
  contains,
  sub,
  take_first,
  take_last,
  drop_first,
  drop_last,
  cut_first,
  cut_last,
  take_first_while,
  take_last_while,
  drop_first_while,
  drop_last_while,
  cut_first_while,
  cut_last_while,
  split_first,
  split_last,
  split_all,
  rsplit_all,
  split_on_char,
  map,
  mapi,
  fold_left,
  fold_right,
  for_all,
  exists,
  trim,
  escaped,
  uppercase_ascii,
  lowercase_ascii,
  capitalize_ascii,
  uncapitalize_ascii,
  iter,
  iteri,
  find_first_index,
  find_last_index,
  index_from,
  index_from_opt,
  rindex_from,
  rindex_from_opt,
  index,
  index_opt,
  rindex,
  rindex_opt,
  find_first,
  find_last,
  find_all,
  rfind_all,
  replace_first,
  replace_last,
  replace_all,
  to_seq,
  to_seqi,
  of_seq,
  get_utf_8_uchar,
  is_valid_utf_8,
  get_utf_16be_uchar,
  is_valid_utf_16be,
  get_utf_16le_uchar,
  is_valid_utf_16le,
  edit_distance,
  spellcheck,
  get_uint8,
  get_int8,
  get_uint16_ne,
  get_uint16_be,
  get_uint16_le,
  get_int16_ne,
  get_int16_be,
  get_int16_le,
  get_int32_ne,
  hash,
  seeded_hash,
  get_int32_be,
  get_int32_le,
  get_int64_ne,
  get_int64_be,
  get_int64_le,
}
/* No side effect */
