// melange/sys.mjs
// Generated by Melange

import * as Caml_exceptions from "melange.js/caml_exceptions.mjs";
import * as Caml_external_polyfill from "melange.js/caml_external_polyfill.mjs";
import * as Caml_sys from "melange.js/caml_sys.mjs";
import * as Stdlib from "./stdlib.mjs";

const executable_name = Caml_sys.caml_sys_executable_name();

const os_type = Caml_sys.os_type();

const backend_type = {
  TAG: /* Other */ 0,
  _0: "Melange"
};

const big_endian = false;

const unix = Caml_sys.os_type() === "Unix";

const win32 = Caml_sys.os_type() === "Win32";

function getenv_opt(s) {
  const x = typeof process === "undefined" ? undefined : process;
  if (x !== undefined) {
    return x.env[s];
  }
  
}

const interactive = {
  contents: false
};

function set_signal(sig_num, sig_beh) {
  
}

function signal_to_string(s) {
  if (s === -1) {
    return "SIGABRT";
  } else if (s === -2) {
    return "SIGALRM";
  } else if (s === -3) {
    return "SIGFPE";
  } else if (s === -4) {
    return "SIGHUP";
  } else if (s === -5) {
    return "SIGILL";
  } else if (s === -6) {
    return "SIGINT";
  } else if (s === -7) {
    return "SIGKILL";
  } else if (s === -8) {
    return "SIGPIPE";
  } else if (s === -9) {
    return "SIGQUIT";
  } else if (s === -10) {
    return "SIGSEGV";
  } else if (s === -11) {
    return "SIGTERM";
  } else if (s === -12) {
    return "SIGUSR1";
  } else if (s === -13) {
    return "SIGUSR2";
  } else if (s === -14) {
    return "SIGCHLD";
  } else if (s === -15) {
    return "SIGCONT";
  } else if (s === -16) {
    return "SIGSTOP";
  } else if (s === -17) {
    return "SIGTSTP";
  } else if (s === -18) {
    return "SIGTTIN";
  } else if (s === -19) {
    return "SIGTTOU";
  } else if (s === -20) {
    return "SIGVTALRM";
  } else if (s === -21) {
    return "SIGPROF";
  } else if (s === -22) {
    return "SIGBUS";
  } else if (s === -23) {
    return "SIGPOLL";
  } else if (s === -24) {
    return "SIGSYS";
  } else if (s === -25) {
    return "SIGTRAP";
  } else if (s === -26) {
    return "SIGURG";
  } else if (s === -27) {
    return "SIGXCPU";
  } else if (s === -28) {
    return "SIGXFSZ";
  } else if (s === -29) {
    return "SIGIO";
  } else if (s === -30) {
    return "SIGWINCH";
  } else if (s < -30) {
    return Stdlib.invalid_arg("Sys.signal_to_string");
  } else {
    return "SIG(" + (String(s) + ")");
  }
}

function signal_of_int(i) {
  if (i < 0) {
    return Stdlib.invalid_arg("Sys.signal_of_int");
  } else {
    return Caml_external_polyfill.resolve("caml_sys_rev_convert_signal_number")(i);
  }
}

function signal_to_int(i) {
  if (i < -30) {
    return Stdlib.invalid_arg("Sys.signal_to_int");
  } else {
    return Caml_external_polyfill.resolve("caml_sys_convert_signal_number")(i);
  }
}

const Break = /* @__PURE__ */ Caml_exceptions.create("Stdlib.Sys.Break");

function catch_break(on) {
  
}

function Make(Immediate, Non_immediate) {
  const repr = /* Non_immediate */ 1;
  return {
    repr: repr
  };
}

const Immediate64 = {
  Make: Make
};

const io_buffer_size = 65536;

const cygwin = false;

const word_size = 32;

const int_size = 32;

const max_string_length = 2147483647;

const max_array_length = 2147483647;

const max_floatarray_length = 2147483647;

const sigabrt = -1;

const sigalrm = -2;

const sigfpe = -3;

const sighup = -4;

const sigill = -5;

const sigint = -6;

const sigkill = -7;

const sigpipe = -8;

const sigquit = -9;

const sigsegv = -10;

const sigterm = -11;

const sigusr1 = -12;

const sigusr2 = -13;

const sigchld = -14;

const sigcont = -15;

const sigstop = -16;

const sigtstp = -17;

const sigttin = -18;

const sigttou = -19;

const sigvtalrm = -20;

const sigprof = -21;

const sigbus = -22;

const sigpoll = -23;

const sigsys = -24;

const sigtrap = -25;

const sigurg = -26;

const sigxcpu = -27;

const sigxfsz = -28;

const sigio = -29;

const sigwinch = -30;

const ocaml_version = "4.14.0+mel";

const development_version = false;

const ocaml_release = {
  major: 4,
  minor: 14,
  patchlevel: 0,
  extra: [
    /* Plus */ 0,
    "mel"
  ]
};

function enable_runtime_warnings(prim) {
  return Caml_external_polyfill.resolve("caml_ml_enable_runtime_warnings")(prim);
}

function runtime_warnings_enabled(prim) {
  return Caml_external_polyfill.resolve("caml_ml_runtime_warnings_enabled")(prim);
}

export {
  executable_name,
  getenv_opt,
  io_buffer_size,
  interactive,
  os_type,
  backend_type,
  unix,
  win32,
  cygwin,
  word_size,
  int_size,
  big_endian,
  max_string_length,
  max_array_length,
  max_floatarray_length,
  set_signal,
  sigabrt,
  sigalrm,
  sigfpe,
  sighup,
  sigill,
  sigint,
  sigkill,
  sigpipe,
  sigquit,
  sigsegv,
  sigterm,
  sigusr1,
  sigusr2,
  sigchld,
  sigcont,
  sigstop,
  sigtstp,
  sigttin,
  sigttou,
  sigvtalrm,
  sigprof,
  sigbus,
  sigpoll,
  sigsys,
  sigtrap,
  sigurg,
  sigxcpu,
  sigxfsz,
  sigio,
  sigwinch,
  signal_to_string,
  signal_of_int,
  signal_to_int,
  Break,
  catch_break,
  ocaml_version,
  development_version,
  ocaml_release,
  enable_runtime_warnings,
  runtime_warnings_enabled,
  Immediate64,
}
/* No side effect */
