// melange.belt/belt_SetDict.js
// Generated by Melange
'use strict';

const Belt__Belt_internalAVLset = require("./belt_internalAVLset.js");
const Caml_option = require("melange.js/caml_option.js");

function add(t, x, cmp) {
  if (t === undefined) {
    return Belt__Belt_internalAVLset.singleton(x);
  }
  const nt = Caml_option.valFromOption(t);
  const k = nt.v;
  const c = cmp(x, k);
  if (c === 0) {
    return t;
  }
  const l = nt.l;
  const r = nt.r;
  if (c < 0) {
    const ll = add(l, x, cmp);
    if (ll === l) {
      return t;
    } else {
      return Belt__Belt_internalAVLset.bal(ll, k, r);
    }
  }
  const rr = add(r, x, cmp);
  if (rr === r) {
    return t;
  } else {
    return Belt__Belt_internalAVLset.bal(l, k, rr);
  }
}

function remove(t, x, cmp) {
  if (t === undefined) {
    return t;
  }
  const n = Caml_option.valFromOption(t);
  const v = n.v;
  const l = n.l;
  const r = n.r;
  const c = cmp(x, v);
  if (c === 0) {
    if (l === undefined) {
      return r;
    }
    if (r === undefined) {
      return l;
    }
    const rn = Caml_option.valFromOption(r);
    const v$1 = {
      contents: rn.v
    };
    const r$1 = Belt__Belt_internalAVLset.removeMinAuxWithRef(rn, v$1);
    return Belt__Belt_internalAVLset.bal(l, v$1.contents, r$1);
  }
  if (c < 0) {
    const ll = remove(l, x, cmp);
    if (ll === l) {
      return t;
    } else {
      return Belt__Belt_internalAVLset.bal(ll, v, r);
    }
  }
  const rr = remove(r, x, cmp);
  if (rr === r) {
    return t;
  } else {
    return Belt__Belt_internalAVLset.bal(l, v, rr);
  }
}

function mergeMany(h, arr, cmp) {
  const len = arr.length;
  let v = h;
  for (let i = 0; i < len; ++i) {
    const key = arr[i];
    v = add(v, key, cmp);
  }
  return v;
}

function removeMany(h, arr, cmp) {
  const len = arr.length;
  let v = h;
  for (let i = 0; i < len; ++i) {
    const key = arr[i];
    v = remove(v, key, cmp);
  }
  return v;
}

function splitAuxNoPivot(cmp, n, x) {
  const v = n.v;
  const l = n.l;
  const r = n.r;
  const c = cmp(x, v);
  if (c === 0) {
    return [
      l,
      r
    ];
  }
  if (c < 0) {
    if (l === undefined) {
      return [
        undefined,
        n
      ];
    }
    const match = splitAuxNoPivot(cmp, Caml_option.valFromOption(l), x);
    return [
      match[0],
      Belt__Belt_internalAVLset.joinShared(match[1], v, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined
    ];
  }
  const match$1 = splitAuxNoPivot(cmp, Caml_option.valFromOption(r), x);
  return [
    Belt__Belt_internalAVLset.joinShared(l, v, match$1[0]),
    match$1[1]
  ];
}

function splitAuxPivot(cmp, n, x, pres) {
  const v = n.v;
  const l = n.l;
  const r = n.r;
  const c = cmp(x, v);
  if (c === 0) {
    pres.contents = true;
    return [
      l,
      r
    ];
  }
  if (c < 0) {
    if (l === undefined) {
      return [
        undefined,
        n
      ];
    }
    const match = splitAuxPivot(cmp, Caml_option.valFromOption(l), x, pres);
    return [
      match[0],
      Belt__Belt_internalAVLset.joinShared(match[1], v, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined
    ];
  }
  const match$1 = splitAuxPivot(cmp, Caml_option.valFromOption(r), x, pres);
  return [
    Belt__Belt_internalAVLset.joinShared(l, v, match$1[0]),
    match$1[1]
  ];
}

function split(t, x, cmp) {
  if (t === undefined) {
    return [
      [
        undefined,
        undefined
      ],
      false
    ];
  }
  const pres = {
    contents: false
  };
  const v = splitAuxPivot(cmp, Caml_option.valFromOption(t), x, pres);
  return [
    v,
    pres.contents
  ];
}

function union(s1, s2, cmp) {
  if (s1 === undefined) {
    return s2;
  }
  if (s2 === undefined) {
    return s1;
  }
  const n2 = Caml_option.valFromOption(s2);
  const n1 = Caml_option.valFromOption(s1);
  const h1 = n1.h;
  const h2 = n2.h;
  if (h1 >= h2) {
    if (h2 === 1) {
      return add(s1, n2.v, cmp);
    }
    const v1 = n1.v;
    const l1 = n1.l;
    const r1 = n1.r;
    const match = splitAuxNoPivot(cmp, n2, v1);
    return Belt__Belt_internalAVLset.joinShared(union(l1, match[0], cmp), v1, union(r1, match[1], cmp));
  }
  if (h1 === 1) {
    return add(s2, n1.v, cmp);
  }
  const v2 = n2.v;
  const l2 = n2.l;
  const r2 = n2.r;
  const match$1 = splitAuxNoPivot(cmp, n1, v2);
  return Belt__Belt_internalAVLset.joinShared(union(match$1[0], l2, cmp), v2, union(match$1[1], r2, cmp));
}

function intersect(s1, s2, cmp) {
  if (s1 === undefined) {
    return;
  }
  if (s2 === undefined) {
    return;
  }
  const n1 = Caml_option.valFromOption(s1);
  const v1 = n1.v;
  const l1 = n1.l;
  const r1 = n1.r;
  const pres = {
    contents: false
  };
  const match = splitAuxPivot(cmp, Caml_option.valFromOption(s2), v1, pres);
  const ll = intersect(l1, match[0], cmp);
  const rr = intersect(r1, match[1], cmp);
  if (pres.contents) {
    return Belt__Belt_internalAVLset.joinShared(ll, v1, rr);
  } else {
    return Belt__Belt_internalAVLset.concatShared(ll, rr);
  }
}

function diff(s1, s2, cmp) {
  if (s1 === undefined) {
    return s1;
  }
  if (s2 === undefined) {
    return s1;
  }
  const n1 = Caml_option.valFromOption(s1);
  const v1 = n1.v;
  const l1 = n1.l;
  const r1 = n1.r;
  const pres = {
    contents: false
  };
  const match = splitAuxPivot(cmp, Caml_option.valFromOption(s2), v1, pres);
  const ll = diff(l1, match[0], cmp);
  const rr = diff(r1, match[1], cmp);
  if (pres.contents) {
    return Belt__Belt_internalAVLset.concatShared(ll, rr);
  } else {
    return Belt__Belt_internalAVLset.joinShared(ll, v1, rr);
  }
}

let empty;

const fromArray = Belt__Belt_internalAVLset.fromArray;

const fromSortedArrayUnsafe = Belt__Belt_internalAVLset.fromSortedArrayUnsafe;

const isEmpty = Belt__Belt_internalAVLset.isEmpty;

const has = Belt__Belt_internalAVLset.has;

const subset = Belt__Belt_internalAVLset.subset;

const cmp = Belt__Belt_internalAVLset.cmp;

const eq = Belt__Belt_internalAVLset.eq;

const forEachU = Belt__Belt_internalAVLset.forEachU;

const forEach = Belt__Belt_internalAVLset.forEach;

const reduceU = Belt__Belt_internalAVLset.reduceU;

const reduce = Belt__Belt_internalAVLset.reduce;

const everyU = Belt__Belt_internalAVLset.everyU;

const every = Belt__Belt_internalAVLset.every;

const someU = Belt__Belt_internalAVLset.someU;

const some = Belt__Belt_internalAVLset.some;

const keepU = Belt__Belt_internalAVLset.keepSharedU;

const keep = Belt__Belt_internalAVLset.keepShared;

const partitionU = Belt__Belt_internalAVLset.partitionSharedU;

const partition = Belt__Belt_internalAVLset.partitionShared;

const size = Belt__Belt_internalAVLset.size;

const toList = Belt__Belt_internalAVLset.toList;

const toArray = Belt__Belt_internalAVLset.toArray;

const minimum = Belt__Belt_internalAVLset.minimum;

const minUndefined = Belt__Belt_internalAVLset.minUndefined;

const maximum = Belt__Belt_internalAVLset.maximum;

const maxUndefined = Belt__Belt_internalAVLset.maxUndefined;

const get = Belt__Belt_internalAVLset.get;

const getUndefined = Belt__Belt_internalAVLset.getUndefined;

const getExn = Belt__Belt_internalAVLset.getExn;

const checkInvariantInternal = Belt__Belt_internalAVLset.checkInvariantInternal;

module.exports = {
  empty,
  fromArray,
  fromSortedArrayUnsafe,
  isEmpty,
  has,
  add,
  mergeMany,
  remove,
  removeMany,
  union,
  intersect,
  diff,
  subset,
  cmp,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  keepU,
  keep,
  partitionU,
  partition,
  size,
  toList,
  toArray,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getExn,
  split,
  checkInvariantInternal,
}
/* No side effect */
