// melange.belt/belt_Set.js
// Generated by Melange
'use strict';

const Belt__Belt_SetDict = require("./belt_SetDict.js");
const Belt__Belt_internalAVLset = require("./belt_internalAVLset.js");
const Curry = require("melange.js/curry.js");

function fromArray(data, id) {
  const cmp = id.cmp;
  return {
    cmp,
    data: Belt__Belt_internalAVLset.fromArray(data, cmp)
  };
}

function remove(m, e) {
  const data = m.data;
  const cmp = m.cmp;
  const newData = Belt__Belt_SetDict.remove(data, e, cmp);
  if (newData === data) {
    return m;
  } else {
    return {
      cmp,
      data: newData
    };
  }
}

function add(m, e) {
  const data = m.data;
  const cmp = m.cmp;
  const newData = Belt__Belt_SetDict.add(data, e, cmp);
  if (newData === data) {
    return m;
  } else {
    return {
      cmp,
      data: newData
    };
  }
}

function mergeMany(m, e) {
  const cmp = m.cmp;
  return {
    cmp,
    data: Belt__Belt_SetDict.mergeMany(m.data, e, cmp)
  };
}

function removeMany(m, e) {
  const cmp = m.cmp;
  return {
    cmp,
    data: Belt__Belt_SetDict.removeMany(m.data, e, cmp)
  };
}

function union(m, n) {
  const cmp = m.cmp;
  return {
    cmp,
    data: Belt__Belt_SetDict.union(m.data, n.data, cmp)
  };
}

function intersect(m, n) {
  const cmp = m.cmp;
  return {
    cmp,
    data: Belt__Belt_SetDict.intersect(m.data, n.data, cmp)
  };
}

function diff(m, n) {
  const cmp = m.cmp;
  return {
    cmp,
    data: Belt__Belt_SetDict.diff(m.data, n.data, cmp)
  };
}

function subset(m, n) {
  const cmp = m.cmp;
  return Belt__Belt_internalAVLset.subset(m.data, n.data, cmp);
}

function split(m, e) {
  const cmp = m.cmp;
  const match = Belt__Belt_SetDict.split(m.data, e, cmp);
  const match$1 = match[0];
  return [
    [
      {
        cmp,
        data: match$1[0]
      },
      {
        cmp,
        data: match$1[1]
      }
    ],
    match[1]
  ];
}

function make(id) {
  return {
    cmp: id.cmp,
    data: undefined
  };
}

function isEmpty(m) {
  return Belt__Belt_internalAVLset.isEmpty(m.data);
}

function cmp(m, n) {
  const cmp$1 = m.cmp;
  return Belt__Belt_internalAVLset.cmp(m.data, n.data, cmp$1);
}

function eq(m, n) {
  return Belt__Belt_internalAVLset.eq(m.data, n.data, m.cmp);
}

function forEachU(m, f) {
  Belt__Belt_internalAVLset.forEachU(m.data, f);
}

function forEach(m, f) {
  Belt__Belt_internalAVLset.forEachU(m.data, Curry.__1(f));
}

function reduceU(m, acc, f) {
  return Belt__Belt_internalAVLset.reduceU(m.data, acc, f);
}

function reduce(m, acc, f) {
  return reduceU(m, acc, Curry.__2(f));
}

function everyU(m, f) {
  return Belt__Belt_internalAVLset.everyU(m.data, f);
}

function every(m, f) {
  return Belt__Belt_internalAVLset.everyU(m.data, Curry.__1(f));
}

function someU(m, f) {
  return Belt__Belt_internalAVLset.someU(m.data, f);
}

function some(m, f) {
  return Belt__Belt_internalAVLset.someU(m.data, Curry.__1(f));
}

function keepU(m, f) {
  return {
    cmp: m.cmp,
    data: Belt__Belt_internalAVLset.keepSharedU(m.data, f)
  };
}

function keep(m, f) {
  return keepU(m, Curry.__1(f));
}

function partitionU(m, f) {
  const match = Belt__Belt_internalAVLset.partitionSharedU(m.data, f);
  const cmp = m.cmp;
  return [
    {
      cmp,
      data: match[0]
    },
    {
      cmp,
      data: match[1]
    }
  ];
}

function partition(m, f) {
  return partitionU(m, Curry.__1(f));
}

function size(m) {
  return Belt__Belt_internalAVLset.size(m.data);
}

function toList(m) {
  return Belt__Belt_internalAVLset.toList(m.data);
}

function toArray(m) {
  return Belt__Belt_internalAVLset.toArray(m.data);
}

function minimum(m) {
  return Belt__Belt_internalAVLset.minimum(m.data);
}

function minUndefined(m) {
  return Belt__Belt_internalAVLset.minUndefined(m.data);
}

function maximum(m) {
  return Belt__Belt_internalAVLset.maximum(m.data);
}

function maxUndefined(m) {
  return Belt__Belt_internalAVLset.maxUndefined(m.data);
}

function get(m, e) {
  return Belt__Belt_internalAVLset.get(m.data, e, m.cmp);
}

function getUndefined(m, e) {
  return Belt__Belt_internalAVLset.getUndefined(m.data, e, m.cmp);
}

function getExn(m, e) {
  return Belt__Belt_internalAVLset.getExn(m.data, e, m.cmp);
}

function has(m, e) {
  return Belt__Belt_internalAVLset.has(m.data, e, m.cmp);
}

function fromSortedArrayUnsafe(xs, id) {
  return {
    cmp: id.cmp,
    data: Belt__Belt_internalAVLset.fromSortedArrayUnsafe(xs)
  };
}

function getData(m) {
  return m.data;
}

function getId(m) {
  const cmp = m.cmp;
  return {
    cmp
  };
}

function packIdData(id, data) {
  return {
    cmp: id.cmp,
    data
  };
}

function checkInvariantInternal(d) {
  Belt__Belt_internalAVLset.checkInvariantInternal(d.data);
}

module.exports = {
  make,
  fromArray,
  fromSortedArrayUnsafe,
  isEmpty,
  has,
  add,
  mergeMany,
  remove,
  removeMany,
  union,
  intersect,
  diff,
  subset,
  cmp,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  keepU,
  keep,
  partitionU,
  partition,
  size,
  toArray,
  toList,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getExn,
  split,
  checkInvariantInternal,
  getData,
  getId,
  packIdData,
}
/* Belt__Belt_internalAVLset Not a pure module */
