// melange.belt/belt_internalSetString.js
// Generated by Melange
'use strict';

const Belt__Belt_SortArrayString = require("./belt_SortArrayString.js");
const Belt__Belt_internalAVLset = require("./belt_internalAVLset.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Caml_option = require("melange.js/caml_option.js");
const Stdlib = require("melange/stdlib.js");

function has(_t, x) {
  while (true) {
    const t = _t;
    if (t === undefined) {
      return false;
    }
    const n = Caml_option.valFromOption(t);
    const v = n.v;
    if (x === v) {
      return true;
    }
    _t = x < v ? n.l : n.r;
    continue;
  };
}

function compareAux(_e1, _e2) {
  while (true) {
    const e2 = _e2;
    const e1 = _e1;
    if (!e1) {
      return 0;
    }
    if (!e2) {
      return 0;
    }
    const h2 = e2.hd;
    const h1 = e1.hd;
    const k1 = h1.v;
    const k2 = h2.v;
    if (k1 !== k2) {
      if (k1 < k2) {
        return -1;
      } else {
        return 1;
      }
    }
    _e2 = Belt__Belt_internalAVLset.stackAllLeft(h2.r, e2.tl);
    _e1 = Belt__Belt_internalAVLset.stackAllLeft(h1.r, e1.tl);
    continue;
  };
}

function cmp(s1, s2) {
  const len1 = Belt__Belt_internalAVLset.size(s1);
  const len2 = Belt__Belt_internalAVLset.size(s2);
  if (len1 === len2) {
    return compareAux(Belt__Belt_internalAVLset.stackAllLeft(s1, /* [] */ 0), Belt__Belt_internalAVLset.stackAllLeft(s2, /* [] */ 0));
  } else if (len1 < len2) {
    return -1;
  } else {
    return 1;
  }
}

function eq(s1, s2) {
  return cmp(s1, s2) === 0;
}

function subset(_s1, _s2) {
  while (true) {
    const s2 = _s2;
    const s1 = _s1;
    if (s1 === undefined) {
      return true;
    }
    if (s2 === undefined) {
      return false;
    }
    const t2 = Caml_option.valFromOption(s2);
    const t1 = Caml_option.valFromOption(s1);
    const v1 = t1.v;
    const l1 = t1.l;
    const r1 = t1.r;
    const v2 = t2.v;
    const l2 = t2.l;
    const r2 = t2.r;
    if (v1 === v2) {
      if (!subset(l1, l2)) {
        return false;
      }
      _s2 = r2;
      _s1 = r1;
      continue;
    }
    if (v1 < v2) {
      if (!subset(Belt__Belt_internalAVLset.create(l1, v1, undefined), l2)) {
        return false;
      }
      _s1 = r1;
      continue;
    }
    if (!subset(Belt__Belt_internalAVLset.create(undefined, v1, r1), r2)) {
      return false;
    }
    _s1 = l1;
    continue;
  };
}

function get(_n, x) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return;
    }
    const t = Caml_option.valFromOption(n);
    const v = t.v;
    if (x === v) {
      return v;
    }
    _n = x < v ? t.l : t.r;
    continue;
  };
}

function getUndefined(_n, x) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return;
    }
    const t = Caml_option.valFromOption(n);
    const v = t.v;
    if (x === v) {
      return v;
    }
    _n = x < v ? t.l : t.r;
    continue;
  };
}

function getExn(_n, x) {
  while (true) {
    const n = _n;
    if (n !== undefined) {
      const t = Caml_option.valFromOption(n);
      const v = t.v;
      if (x === v) {
        return v;
      }
      _n = x < v ? t.l : t.r;
      continue;
    }
    throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
        MEL_EXN_ID: Stdlib.Not_found
      });
  };
}

function addMutate(t, x) {
  if (t === undefined) {
    return Belt__Belt_internalAVLset.singleton(x);
  }
  const k = t.v;
  if (x === k) {
    return t;
  }
  const l = t.l;
  const r = t.r;
  if (x < k) {
    t.l = addMutate(l, x);
  } else {
    t.r = addMutate(r, x);
  }
  return Belt__Belt_internalAVLset.balMutate(t);
}

function fromArray(xs) {
  const len = xs.length;
  if (len === 0) {
    return;
  }
  let next = Belt__Belt_SortArrayString.strictlySortedLength(xs);
  let result;
  if (next >= 0) {
    result = Belt__Belt_internalAVLset.fromSortedArrayAux(xs, 0, next);
  } else {
    next = -next | 0;
    result = Belt__Belt_internalAVLset.fromSortedArrayRevAux(xs, next - 1 | 0, next);
  }
  for (let i = next; i < len; ++i) {
    result = addMutate(result, xs[i]);
  }
  return result;
}

module.exports = {
  has,
  compareAux,
  cmp,
  eq,
  subset,
  get,
  getUndefined,
  getExn,
  addMutate,
  fromArray,
}
/* No side effect */
