// melange.belt/belt_MutableSet.js
// Generated by Melange
'use strict';

const Belt__Belt_SortArray = require("./belt_SortArray.js");
const Belt__Belt_internalAVLset = require("./belt_internalAVLset.js");
const Caml_option = require("melange.js/caml_option.js");
const Curry = require("melange.js/curry.js");

function remove0(nt, x, cmp) {
  const k = nt.v;
  const c = cmp(x, k);
  if (c === 0) {
    const l = nt.l;
    const r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt__Belt_internalAVLset.removeMinAuxWithRootMutate(nt, Caml_option.valFromOption(r));
        return Belt__Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (c < 0) {
    const l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = remove0(Caml_option.valFromOption(l$1), x, cmp);
      return Belt__Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  const r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = remove0(Caml_option.valFromOption(r$1), x, cmp);
    return Belt__Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function remove(d, v) {
  const oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  const newRoot = remove0(Caml_option.valFromOption(oldRoot), v, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
    return;
  }
  
}

function removeMany0(_t, xs, _i, len, cmp) {
  while (true) {
    const i = _i;
    const t = _t;
    if (i >= len) {
      return t;
    }
    const ele = xs[i];
    const u = remove0(t, ele, cmp);
    if (u === undefined) {
      return;
    }
    _i = i + 1 | 0;
    _t = Caml_option.valFromOption(u);
    continue;
  };
}

function removeMany(d, xs) {
  const oldRoot = d.data;
  if (oldRoot === undefined) {
    return;
  }
  const len = xs.length;
  d.data = removeMany0(Caml_option.valFromOption(oldRoot), xs, 0, len, d.cmp);
}

function removeCheck0(nt, x, removed, cmp) {
  const k = nt.v;
  const c = cmp(x, k);
  if (c === 0) {
    removed.contents = true;
    const l = nt.l;
    const r = nt.r;
    if (l !== undefined) {
      if (r !== undefined) {
        nt.r = Belt__Belt_internalAVLset.removeMinAuxWithRootMutate(nt, Caml_option.valFromOption(r));
        return Belt__Belt_internalAVLset.balMutate(nt);
      } else {
        return l;
      }
    } else {
      return r;
    }
  }
  if (c < 0) {
    const l$1 = nt.l;
    if (l$1 !== undefined) {
      nt.l = removeCheck0(Caml_option.valFromOption(l$1), x, removed, cmp);
      return Belt__Belt_internalAVLset.balMutate(nt);
    } else {
      return nt;
    }
  }
  const r$1 = nt.r;
  if (r$1 !== undefined) {
    nt.r = removeCheck0(Caml_option.valFromOption(r$1), x, removed, cmp);
    return Belt__Belt_internalAVLset.balMutate(nt);
  } else {
    return nt;
  }
}

function removeCheck(d, v) {
  const oldRoot = d.data;
  if (oldRoot === undefined) {
    return false;
  }
  const removed = {
    contents: false
  };
  const newRoot = removeCheck0(Caml_option.valFromOption(oldRoot), v, removed, d.cmp);
  if (newRoot !== oldRoot) {
    d.data = newRoot;
  }
  return removed.contents;
}

function addCheck0(t, x, added, cmp) {
  if (t !== undefined) {
    const nt = Caml_option.valFromOption(t);
    const k = nt.v;
    const c = cmp(x, k);
    if (c === 0) {
      return t;
    }
    const l = nt.l;
    const r = nt.r;
    if (c < 0) {
      const ll = addCheck0(l, x, added, cmp);
      nt.l = ll;
    } else {
      nt.r = addCheck0(r, x, added, cmp);
    }
    return Belt__Belt_internalAVLset.balMutate(nt);
  }
  added.contents = true;
  return Belt__Belt_internalAVLset.singleton(x);
}

function addCheck(m, e) {
  const oldRoot = m.data;
  const added = {
    contents: false
  };
  const newRoot = addCheck0(oldRoot, e, added, m.cmp);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
  }
  return added.contents;
}

function add(m, e) {
  const oldRoot = m.data;
  const newRoot = Belt__Belt_internalAVLset.addMutate(m.cmp, oldRoot, e);
  if (newRoot !== oldRoot) {
    m.data = newRoot;
    return;
  }
  
}

function addArrayMutate(t, xs, cmp) {
  let v = t;
  for (let i = 0, i_finish = xs.length; i < i_finish; ++i) {
    v = Belt__Belt_internalAVLset.addMutate(cmp, v, xs[i]);
  }
  return v;
}

function mergeMany(d, xs) {
  d.data = addArrayMutate(d.data, xs, d.cmp);
}

function make(id) {
  return {
    cmp: id.cmp,
    data: undefined
  };
}

function isEmpty(d) {
  const n = d.data;
  return n === undefined;
}

function minimum(d) {
  return Belt__Belt_internalAVLset.minimum(d.data);
}

function minUndefined(d) {
  return Belt__Belt_internalAVLset.minUndefined(d.data);
}

function maximum(d) {
  return Belt__Belt_internalAVLset.maximum(d.data);
}

function maxUndefined(d) {
  return Belt__Belt_internalAVLset.maxUndefined(d.data);
}

function forEachU(d, f) {
  Belt__Belt_internalAVLset.forEachU(d.data, f);
}

function forEach(d, f) {
  Belt__Belt_internalAVLset.forEachU(d.data, Curry.__1(f));
}

function reduceU(d, acc, cb) {
  return Belt__Belt_internalAVLset.reduceU(d.data, acc, cb);
}

function reduce(d, acc, cb) {
  return reduceU(d, acc, Curry.__2(cb));
}

function everyU(d, p) {
  return Belt__Belt_internalAVLset.everyU(d.data, p);
}

function every(d, p) {
  return Belt__Belt_internalAVLset.everyU(d.data, Curry.__1(p));
}

function someU(d, p) {
  return Belt__Belt_internalAVLset.someU(d.data, p);
}

function some(d, p) {
  return Belt__Belt_internalAVLset.someU(d.data, Curry.__1(p));
}

function size(d) {
  return Belt__Belt_internalAVLset.size(d.data);
}

function toList(d) {
  return Belt__Belt_internalAVLset.toList(d.data);
}

function toArray(d) {
  return Belt__Belt_internalAVLset.toArray(d.data);
}

function fromSortedArrayUnsafe(xs, id) {
  return {
    cmp: id.cmp,
    data: Belt__Belt_internalAVLset.fromSortedArrayUnsafe(xs)
  };
}

function checkInvariantInternal(d) {
  Belt__Belt_internalAVLset.checkInvariantInternal(d.data);
}

function fromArray(data, id) {
  const cmp = id.cmp;
  return {
    cmp: cmp,
    data: Belt__Belt_internalAVLset.fromArray(data, cmp)
  };
}

function cmp(d0, d1) {
  return Belt__Belt_internalAVLset.cmp(d0.data, d1.data, d0.cmp);
}

function eq(d0, d1) {
  return Belt__Belt_internalAVLset.eq(d0.data, d1.data, d0.cmp);
}

function get(d, x) {
  return Belt__Belt_internalAVLset.get(d.data, x, d.cmp);
}

function getUndefined(d, x) {
  return Belt__Belt_internalAVLset.getUndefined(d.data, x, d.cmp);
}

function getExn(d, x) {
  return Belt__Belt_internalAVLset.getExn(d.data, x, d.cmp);
}

function split(d, key) {
  const arr = Belt__Belt_internalAVLset.toArray(d.data);
  const cmp = d.cmp;
  const i = Belt__Belt_SortArray.binarySearchByU(arr, key, cmp);
  const len = arr.length;
  if (i >= 0) {
    return [
      [
        {
          cmp: cmp,
          data: Belt__Belt_internalAVLset.fromSortedArrayAux(arr, 0, i)
        },
        {
          cmp: cmp,
          data: Belt__Belt_internalAVLset.fromSortedArrayAux(arr, i + 1 | 0, (len - i | 0) - 1 | 0)
        }
      ],
      true
    ];
  }
  const next = (-i | 0) - 1 | 0;
  return [
    [
      {
        cmp: cmp,
        data: Belt__Belt_internalAVLset.fromSortedArrayAux(arr, 0, next)
      },
      {
        cmp: cmp,
        data: Belt__Belt_internalAVLset.fromSortedArrayAux(arr, next, len - next | 0)
      }
    ],
    false
  ];
}

function keepU(d, p) {
  return {
    cmp: d.cmp,
    data: Belt__Belt_internalAVLset.keepCopyU(d.data, p)
  };
}

function keep(d, p) {
  return keepU(d, Curry.__1(p));
}

function partitionU(d, p) {
  const cmp = d.cmp;
  const match = Belt__Belt_internalAVLset.partitionCopyU(d.data, p);
  return [
    {
      cmp: cmp,
      data: match[0]
    },
    {
      cmp: cmp,
      data: match[1]
    }
  ];
}

function partition(d, p) {
  return partitionU(d, Curry.__1(p));
}

function subset(a, b) {
  return Belt__Belt_internalAVLset.subset(a.data, b.data, a.cmp);
}

function intersect(a, b) {
  const cmp = a.cmp;
  const match = a.data;
  const match$1 = b.data;
  if (match === undefined) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  if (match$1 === undefined) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  const datab0 = Caml_option.valFromOption(match$1);
  const dataa0 = Caml_option.valFromOption(match);
  const sizea = Belt__Belt_internalAVLset.lengthNode(dataa0);
  const sizeb = Belt__Belt_internalAVLset.lengthNode(datab0);
  const totalSize = sizea + sizeb | 0;
  const tmp = new Array(totalSize);
  Belt__Belt_internalAVLset.fillArray(dataa0, 0, tmp);
  Belt__Belt_internalAVLset.fillArray(datab0, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || cmp(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  const tmp2 = new Array(sizea < sizeb ? sizea : sizeb);
  const k = Belt__Belt_SortArray.intersectU(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
    cmp: cmp,
    data: Belt__Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function diff(a, b) {
  const cmp = a.cmp;
  const dataa = a.data;
  const match = b.data;
  if (dataa === undefined) {
    return {
      cmp: cmp,
      data: undefined
    };
  }
  if (match === undefined) {
    return {
      cmp: cmp,
      data: Belt__Belt_internalAVLset.copy(dataa)
    };
  }
  const datab0 = Caml_option.valFromOption(match);
  const dataa0 = Caml_option.valFromOption(dataa);
  const sizea = Belt__Belt_internalAVLset.lengthNode(dataa0);
  const sizeb = Belt__Belt_internalAVLset.lengthNode(datab0);
  const totalSize = sizea + sizeb | 0;
  const tmp = new Array(totalSize);
  Belt__Belt_internalAVLset.fillArray(dataa0, 0, tmp);
  Belt__Belt_internalAVLset.fillArray(datab0, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0 || cmp(tmp[totalSize - 1 | 0], tmp[0]) < 0) {
    return {
      cmp: cmp,
      data: Belt__Belt_internalAVLset.copy(dataa)
    };
  }
  const tmp2 = new Array(sizea);
  const k = Belt__Belt_SortArray.diffU(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
    cmp: cmp,
    data: Belt__Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function union(a, b) {
  const cmp = a.cmp;
  const dataa = a.data;
  const datab = b.data;
  if (dataa === undefined) {
    return {
      cmp: cmp,
      data: Belt__Belt_internalAVLset.copy(datab)
    };
  }
  if (datab === undefined) {
    return {
      cmp: cmp,
      data: Belt__Belt_internalAVLset.copy(dataa)
    };
  }
  const datab0 = Caml_option.valFromOption(datab);
  const dataa0 = Caml_option.valFromOption(dataa);
  const sizea = Belt__Belt_internalAVLset.lengthNode(dataa0);
  const sizeb = Belt__Belt_internalAVLset.lengthNode(datab0);
  const totalSize = sizea + sizeb | 0;
  const tmp = new Array(totalSize);
  Belt__Belt_internalAVLset.fillArray(dataa0, 0, tmp);
  Belt__Belt_internalAVLset.fillArray(datab0, sizea, tmp);
  if (cmp(tmp[sizea - 1 | 0], tmp[sizea]) < 0) {
    return {
      cmp: cmp,
      data: Belt__Belt_internalAVLset.fromSortedArrayAux(tmp, 0, totalSize)
    };
  }
  const tmp2 = new Array(totalSize);
  const k = Belt__Belt_SortArray.unionU(tmp, 0, sizea, tmp, sizea, sizeb, tmp2, 0, cmp);
  return {
    cmp: cmp,
    data: Belt__Belt_internalAVLset.fromSortedArrayAux(tmp2, 0, k)
  };
}

function has(d, x) {
  return Belt__Belt_internalAVLset.has(d.data, x, d.cmp);
}

function copy(d) {
  return {
    cmp: d.cmp,
    data: Belt__Belt_internalAVLset.copy(d.data)
  };
}

module.exports = {
  make,
  fromArray,
  fromSortedArrayUnsafe,
  copy,
  isEmpty,
  has,
  add,
  addCheck,
  mergeMany,
  remove,
  removeCheck,
  removeMany,
  union,
  intersect,
  diff,
  subset,
  cmp,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  keepU,
  keep,
  partitionU,
  partition,
  size,
  toList,
  toArray,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getExn,
  split,
  checkInvariantInternal,
}
/* No side effect */
