// melange.belt/belt_HashSetInt.js
// Generated by Melange
'use strict';

const Belt__Belt_internalBucketsType = require("./belt_internalBucketsType.js");
const Belt__Belt_internalSetBuckets = require("./belt_internalSetBuckets.js");
const Caml_hash_primitive = require("melange.js/caml_hash_primitive.js");

function copyBucket(h_buckets, ndata_tail, _old_bucket) {
  while (true) {
    const old_bucket = _old_bucket;
    if (old_bucket === undefined) {
      return;
    }
    const nidx = Caml_hash_primitive.caml_hash_final_mix(Caml_hash_primitive.caml_hash_mix_int(0, old_bucket.key)) & (h_buckets.length - 1 | 0);
    const tail = ndata_tail[nidx];
    if (tail !== undefined) {
      tail.next = old_bucket;
    } else {
      h_buckets[nidx] = old_bucket;
    }
    ndata_tail[nidx] = old_bucket;
    _old_bucket = old_bucket.next;
    continue;
  };
}

function remove(h, key) {
  const h_buckets = h.buckets;
  const i = Caml_hash_primitive.caml_hash_final_mix(Caml_hash_primitive.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  const l = h_buckets[i];
  if (l === undefined) {
    return;
  }
  const next_cell = l.next;
  if (l.key === key) {
    h.size = h.size - 1 | 0;
    h_buckets[i] = next_cell;
    return;
  } else if (next_cell !== undefined) {
    let _prec = l;
    let _cell = next_cell;
    while (true) {
      const cell = _cell;
      const prec = _prec;
      const cell_next = cell.next;
      if (cell.key === key) {
        prec.next = cell_next;
        h.size = h.size - 1 | 0;
        return;
      }
      if (cell_next === undefined) {
        return;
      }
      _cell = cell_next;
      _prec = cell;
      continue;
    };
  } else {
    return;
  }
}

function addBucket(h, key, _cell) {
  while (true) {
    const cell = _cell;
    if (cell.key === key) {
      return;
    }
    const n = cell.next;
    if (n !== undefined) {
      _cell = n;
      continue;
    }
    h.size = h.size + 1 | 0;
    cell.next = {
      key: key,
      next: undefined
    };
    return;
  };
}

function add(h, key) {
  const h_buckets = h.buckets;
  const buckets_len = h_buckets.length;
  const i = Caml_hash_primitive.caml_hash_final_mix(Caml_hash_primitive.caml_hash_mix_int(0, key)) & (buckets_len - 1 | 0);
  const l = h_buckets[i];
  if (l !== undefined) {
    addBucket(h, key, l);
  } else {
    h_buckets[i] = {
      key: key,
      next: undefined
    };
    h.size = h.size + 1 | 0;
  }
  if (h.size > (buckets_len << 1)) {
    const odata = h.buckets;
    const osize = odata.length;
    const nsize = (osize << 1);
    if (nsize < osize) {
      return;
    }
    const h_buckets$1 = new Array(nsize);
    const ndata_tail = new Array(nsize);
    h.buckets = h_buckets$1;
    for (let i$1 = 0; i$1 < osize; ++i$1) {
      copyBucket(h_buckets$1, ndata_tail, odata[i$1]);
    }
    for (let i$2 = 0; i$2 < nsize; ++i$2) {
      const tail = ndata_tail[i$2];
      if (tail !== undefined) {
        tail.next = undefined;
      }
      
    }
    return;
  }
  
}

function has(h, key) {
  const h_buckets = h.buckets;
  const nid = Caml_hash_primitive.caml_hash_final_mix(Caml_hash_primitive.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  const bucket = h_buckets[nid];
  if (bucket !== undefined) {
    let _cell = bucket;
    while (true) {
      const cell = _cell;
      if (cell.key === key) {
        return true;
      }
      const nextCell = cell.next;
      if (nextCell === undefined) {
        return false;
      }
      _cell = nextCell;
      continue;
    };
  } else {
    return false;
  }
}

function make(hintSize) {
  return Belt__Belt_internalBucketsType.make(undefined, undefined, hintSize);
}

function size(h) {
  return h.size;
}

function fromArray(arr) {
  const len = arr.length;
  const v = Belt__Belt_internalBucketsType.make(undefined, undefined, len);
  for (let i = 0; i < len; ++i) {
    add(v, arr[i]);
  }
  return v;
}

function mergeMany(h, arr) {
  const len = arr.length;
  for (let i = 0; i < len; ++i) {
    add(h, arr[i]);
  }
}

const clear = Belt__Belt_internalBucketsType.clear;

const isEmpty = Belt__Belt_internalBucketsType.isEmpty;

const copy = Belt__Belt_internalSetBuckets.copy;

const forEachU = Belt__Belt_internalSetBuckets.forEachU;

const forEach = Belt__Belt_internalSetBuckets.forEach;

const reduceU = Belt__Belt_internalSetBuckets.reduceU;

const reduce = Belt__Belt_internalSetBuckets.reduce;

const logStats = Belt__Belt_internalSetBuckets.logStats;

const toArray = Belt__Belt_internalSetBuckets.toArray;

const getBucketHistogram = Belt__Belt_internalSetBuckets.getBucketHistogram;

module.exports = {
  make,
  clear,
  isEmpty,
  add,
  copy,
  has,
  remove,
  forEachU,
  forEach,
  reduceU,
  reduce,
  size,
  logStats,
  toArray,
  fromArray,
  mergeMany,
  getBucketHistogram,
}
/* No side effect */
