// melange.belt/belt_internalAVLset.js
// Generated by Melange
'use strict';

const Belt__Belt_SortArray = require("./belt_SortArray.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Caml_option = require("melange.js/caml_option.js");
const Curry = require("melange.js/curry.js");
const Stdlib = require("melange/stdlib.js");

function copy(n) {
  if (n === undefined) {
    return n;
  }
  const n$1 = Caml_option.valFromOption(n);
  return {
    v: n$1.v,
    h: n$1.h,
    l: copy(n$1.l),
    r: copy(n$1.r)
  };
}

function create(l, v, r) {
  const hl = l !== undefined ? Caml_option.valFromOption(l).h : 0;
  const hr = r !== undefined ? Caml_option.valFromOption(r).h : 0;
  return {
    v: v,
    h: (
      hl >= hr ? hl : hr
    ) + 1 | 0,
    l: l,
    r: r
  };
}

function singleton(x) {
  return {
    v: x,
    h: 1,
    l: undefined,
    r: undefined
  };
}

function heightGe(l, r) {
  if (r !== undefined) {
    if (l !== undefined) {
      return l.h >= r.h;
    } else {
      return false;
    }
  } else {
    return true;
  }
}

function bal(l, v, r) {
  const hl = l !== undefined ? Caml_option.valFromOption(l).h : 0;
  const hr = r !== undefined ? Caml_option.valFromOption(r).h : 0;
  if (hl > (hr + 2 | 0)) {
    const l$1 = Caml_option.valFromOption(l);
    const ll = l$1.l;
    const lr = l$1.r;
    if (heightGe(ll, lr)) {
      return create(ll, l$1.v, create(lr, v, r));
    }
    const lr$1 = Caml_option.valFromOption(lr);
    return create(create(ll, l$1.v, lr$1.l), lr$1.v, create(lr$1.r, v, r));
  }
  if (hr <= (hl + 2 | 0)) {
    return {
      v: v,
      h: (
        hl >= hr ? hl : hr
      ) + 1 | 0,
      l: l,
      r: r
    };
  }
  const r$1 = Caml_option.valFromOption(r);
  const rl = r$1.l;
  const rr = r$1.r;
  if (heightGe(rr, rl)) {
    return create(create(l, v, rl), r$1.v, rr);
  }
  const rl$1 = Caml_option.valFromOption(rl);
  return create(create(l, v, rl$1.l), rl$1.v, create(rl$1.r, r$1.v, rr));
}

function min0Aux(_n) {
  while (true) {
    const n = _n;
    const n$1 = n.l;
    if (n$1 === undefined) {
      return n.v;
    }
    _n = Caml_option.valFromOption(n$1);
    continue;
  };
}

function minimum(n) {
  if (n !== undefined) {
    return Caml_option.some(min0Aux(n));
  }
  
}

function minUndefined(n) {
  if (n !== undefined) {
    return min0Aux(n);
  }
  
}

function max0Aux(_n) {
  while (true) {
    const n = _n;
    const n$1 = n.r;
    if (n$1 === undefined) {
      return n.v;
    }
    _n = Caml_option.valFromOption(n$1);
    continue;
  };
}

function maximum(n) {
  if (n !== undefined) {
    return Caml_option.some(max0Aux(n));
  }
  
}

function maxUndefined(n) {
  if (n !== undefined) {
    return max0Aux(n);
  }
  
}

function removeMinAuxWithRef(n, v) {
  const ln = n.l;
  if (ln !== undefined) {
    return bal(removeMinAuxWithRef(Caml_option.valFromOption(ln), v), n.v, n.r);
  } else {
    v.contents = n.v;
    return n.r;
  }
}

function isEmpty(n) {
  return n === undefined;
}

function stackAllLeft(_v, _s) {
  while (true) {
    const s = _s;
    const v = _v;
    if (v === undefined) {
      return s;
    }
    const x = Caml_option.valFromOption(v);
    _s = {
      hd: x,
      tl: s
    };
    _v = x.l;
    continue;
  };
}

function forEachU(_n, f) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return;
    }
    const n$1 = Caml_option.valFromOption(n);
    forEachU(n$1.l, f);
    f(n$1.v);
    _n = n$1.r;
    continue;
  };
}

function forEach(n, f) {
  forEachU(n, Curry.__1(f));
}

function reduceU(_s, _accu, f) {
  while (true) {
    const accu = _accu;
    const s = _s;
    if (s === undefined) {
      return accu;
    }
    const n = Caml_option.valFromOption(s);
    _accu = f(reduceU(n.l, accu, f), n.v);
    _s = n.r;
    continue;
  };
}

function reduce(s, accu, f) {
  return reduceU(s, accu, Curry.__2(f));
}

function everyU(_n, p) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return true;
    }
    const n$1 = Caml_option.valFromOption(n);
    if (!p(n$1.v)) {
      return false;
    }
    if (!everyU(n$1.l, p)) {
      return false;
    }
    _n = n$1.r;
    continue;
  };
}

function every(n, p) {
  return everyU(n, Curry.__1(p));
}

function someU(_n, p) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return false;
    }
    const n$1 = Caml_option.valFromOption(n);
    if (p(n$1.v)) {
      return true;
    }
    if (someU(n$1.l, p)) {
      return true;
    }
    _n = n$1.r;
    continue;
  };
}

function some(n, p) {
  return someU(n, Curry.__1(p));
}

function addMinElement(n, v) {
  if (n === undefined) {
    return singleton(v);
  }
  const n$1 = Caml_option.valFromOption(n);
  return bal(addMinElement(n$1.l, v), n$1.v, n$1.r);
}

function addMaxElement(n, v) {
  if (n === undefined) {
    return singleton(v);
  }
  const n$1 = Caml_option.valFromOption(n);
  return bal(n$1.l, n$1.v, addMaxElement(n$1.r, v));
}

function joinShared(ln, v, rn) {
  if (ln === undefined) {
    return addMinElement(rn, v);
  }
  if (rn === undefined) {
    return addMaxElement(ln, v);
  }
  const r = Caml_option.valFromOption(rn);
  const l = Caml_option.valFromOption(ln);
  const lh = l.h;
  const rh = r.h;
  if (lh > (rh + 2 | 0)) {
    return bal(l.l, l.v, joinShared(l.r, v, rn));
  } else if (rh > (lh + 2 | 0)) {
    return bal(joinShared(ln, v, r.l), r.v, r.r);
  } else {
    return create(ln, v, rn);
  }
}

function concatShared(t1, t2) {
  if (t1 === undefined) {
    return t2;
  }
  if (t2 === undefined) {
    return t1;
  }
  const t2n = Caml_option.valFromOption(t2);
  const v = {
    contents: t2n.v
  };
  const t2r = removeMinAuxWithRef(t2n, v);
  return joinShared(t1, v.contents, t2r);
}

function partitionSharedU(n, p) {
  if (n === undefined) {
    return [
      undefined,
      undefined
    ];
  }
  const n$1 = Caml_option.valFromOption(n);
  const value = n$1.v;
  const match = partitionSharedU(n$1.l, p);
  const lf = match[1];
  const lt = match[0];
  const pv = p(value);
  const match$1 = partitionSharedU(n$1.r, p);
  const rf = match$1[1];
  const rt = match$1[0];
  if (pv) {
    return [
      joinShared(lt, value, rt),
      concatShared(lf, rf)
    ];
  } else {
    return [
      concatShared(lt, rt),
      joinShared(lf, value, rf)
    ];
  }
}

function partitionShared(n, p) {
  return partitionSharedU(n, Curry.__1(p));
}

function lengthNode(n) {
  const l = n.l;
  const r = n.r;
  const sizeL = l !== undefined ? lengthNode(Caml_option.valFromOption(l)) : 0;
  const sizeR = r !== undefined ? lengthNode(Caml_option.valFromOption(r)) : 0;
  return (1 + sizeL | 0) + sizeR | 0;
}

function size(n) {
  if (n !== undefined) {
    return lengthNode(n);
  } else {
    return 0;
  }
}

function toListAux(_n, _accu) {
  while (true) {
    const accu = _accu;
    const n = _n;
    if (n === undefined) {
      return accu;
    }
    const n$1 = Caml_option.valFromOption(n);
    _accu = {
      hd: n$1.v,
      tl: toListAux(n$1.r, accu)
    };
    _n = n$1.l;
    continue;
  };
}

function toList(s) {
  return toListAux(s, /* [] */ 0);
}

function checkInvariantInternal(_v) {
  while (true) {
    const v = _v;
    if (v === undefined) {
      return;
    }
    const n = Caml_option.valFromOption(v);
    const l = n.l;
    const r = n.r;
    const diff = (
      l !== undefined ? Caml_option.valFromOption(l).h : 0
    ) - (
      r !== undefined ? Caml_option.valFromOption(r).h : 0
    ) | 0;
    if (!(diff <= 2 && diff >= -2)) {
      throw new Caml_js_exceptions.MelangeError("Assert_failure", {
          MEL_EXN_ID: "Assert_failure",
          _1: [
            "belt/belt_internalAVLset.ml",
            244,
            6
          ]
        });
    }
    checkInvariantInternal(l);
    _v = r;
    continue;
  };
}

function fillArray(_n, _i, arr) {
  while (true) {
    const i = _i;
    const n = _n;
    const v = n.v;
    const l = n.l;
    const r = n.r;
    const next = l !== undefined ? fillArray(Caml_option.valFromOption(l), i, arr) : i;
    arr[next] = v;
    const rnext = next + 1 | 0;
    if (r === undefined) {
      return rnext;
    }
    _i = rnext;
    _n = Caml_option.valFromOption(r);
    continue;
  };
}

function fillArrayWithPartition(_n, cursor, arr, p) {
  while (true) {
    const n = _n;
    const v = n.v;
    const l = n.l;
    const r = n.r;
    if (l !== undefined) {
      fillArrayWithPartition(Caml_option.valFromOption(l), cursor, arr, p);
    }
    if (p(v)) {
      const c = cursor.forward;
      arr[c] = v;
      cursor.forward = c + 1 | 0;
    } else {
      const c$1 = cursor.backward;
      arr[c$1] = v;
      cursor.backward = c$1 - 1 | 0;
    }
    if (r === undefined) {
      return;
    }
    _n = Caml_option.valFromOption(r);
    continue;
  };
}

function fillArrayWithFilter(_n, _i, arr, p) {
  while (true) {
    const i = _i;
    const n = _n;
    const v = n.v;
    const l = n.l;
    const r = n.r;
    const next = l !== undefined ? fillArrayWithFilter(Caml_option.valFromOption(l), i, arr, p) : i;
    const rnext = p(v) ? (arr[next] = v, next + 1 | 0) : next;
    if (r === undefined) {
      return rnext;
    }
    _i = rnext;
    _n = Caml_option.valFromOption(r);
    continue;
  };
}

function toArray(n) {
  if (n === undefined) {
    return [];
  }
  const size = lengthNode(n);
  const v = new Array(size);
  fillArray(n, 0, v);
  return v;
}

function fromSortedArrayRevAux(arr, off, len) {
  switch (len) {
    case 0 : {
      return;
      }
    case 1 : {
      return singleton(arr[off]);
      }
    case 2 : {
      const x0 = arr[off];
      const x1 = arr[off - 1 | 0];
      return {
        v: x1,
        h: 2,
        l: singleton(x0),
        r: undefined
      };
      }
    case 3 : {
      const x0$1 = arr[off];
      const x1$1 = arr[off - 1 | 0];
      const x2 = arr[off - 2 | 0];
      return {
        v: x1$1,
        h: 2,
        l: singleton(x0$1),
        r: singleton(x2)
      };
      }
    default:
      const nl = len / 2 | 0;
      const left = fromSortedArrayRevAux(arr, off, nl);
      const mid = arr[off - nl | 0];
      const right = fromSortedArrayRevAux(arr, (off - nl | 0) - 1 | 0, (len - nl | 0) - 1 | 0);
      return create(left, mid, right);
  }
}

function fromSortedArrayAux(arr, off, len) {
  switch (len) {
    case 0 : {
      return;
      }
    case 1 : {
      return singleton(arr[off]);
      }
    case 2 : {
      const x0 = arr[off];
      const x1 = arr[off + 1 | 0];
      return {
        v: x1,
        h: 2,
        l: singleton(x0),
        r: undefined
      };
      }
    case 3 : {
      const x0$1 = arr[off];
      const x1$1 = arr[off + 1 | 0];
      const x2 = arr[off + 2 | 0];
      return {
        v: x1$1,
        h: 2,
        l: singleton(x0$1),
        r: singleton(x2)
      };
      }
    default:
      const nl = len / 2 | 0;
      const left = fromSortedArrayAux(arr, off, nl);
      const mid = arr[off + nl | 0];
      const right = fromSortedArrayAux(arr, (off + nl | 0) + 1 | 0, (len - nl | 0) - 1 | 0);
      return create(left, mid, right);
  }
}

function fromSortedArrayUnsafe(arr) {
  return fromSortedArrayAux(arr, 0, arr.length);
}

function keepSharedU(n, p) {
  if (n === undefined) {
    return;
  }
  const n$1 = Caml_option.valFromOption(n);
  const v = n$1.v;
  const l = n$1.l;
  const r = n$1.r;
  const newL = keepSharedU(l, p);
  const pv = p(v);
  const newR = keepSharedU(r, p);
  if (pv) {
    if (l === newL && r === newR) {
      return n$1;
    } else {
      return joinShared(newL, v, newR);
    }
  } else {
    return concatShared(newL, newR);
  }
}

function keepShared(n, p) {
  return keepSharedU(n, Curry.__1(p));
}

function keepCopyU(n, p) {
  if (n === undefined) {
    return;
  }
  const size = lengthNode(n);
  const v = new Array(size);
  const last = fillArrayWithFilter(n, 0, v, p);
  return fromSortedArrayAux(v, 0, last);
}

function keepCopy(n, p) {
  return keepCopyU(n, Curry.__1(p));
}

function partitionCopyU(n, p) {
  if (n === undefined) {
    return [
      undefined,
      undefined
    ];
  }
  const size = lengthNode(n);
  const v = new Array(size);
  const backward = size - 1 | 0;
  const cursor = {
    forward: 0,
    backward: backward
  };
  fillArrayWithPartition(n, cursor, v, p);
  const forwardLen = cursor.forward;
  return [
    fromSortedArrayAux(v, 0, forwardLen),
    fromSortedArrayRevAux(v, backward, size - forwardLen | 0)
  ];
}

function partitionCopy(n, p) {
  return partitionCopyU(n, Curry.__1(p));
}

function has(_t, x, cmp) {
  while (true) {
    const t = _t;
    if (t === undefined) {
      return false;
    }
    const n = Caml_option.valFromOption(t);
    const v = n.v;
    const c = cmp(x, v);
    if (c === 0) {
      return true;
    }
    _t = c < 0 ? n.l : n.r;
    continue;
  };
}

function cmp(s1, s2, cmp$1) {
  const len1 = size(s1);
  const len2 = size(s2);
  if (len1 === len2) {
    let _e1 = stackAllLeft(s1, /* [] */ 0);
    let _e2 = stackAllLeft(s2, /* [] */ 0);
    while (true) {
      const e2 = _e2;
      const e1 = _e1;
      if (!e1) {
        return 0;
      }
      if (!e2) {
        return 0;
      }
      const h2 = e2.hd;
      const h1 = e1.hd;
      const c = cmp$1(h1.v, h2.v);
      if (c !== 0) {
        return c;
      }
      _e2 = stackAllLeft(h2.r, e2.tl);
      _e1 = stackAllLeft(h1.r, e1.tl);
      continue;
    };
  } else if (len1 < len2) {
    return -1;
  } else {
    return 1;
  }
}

function eq(s1, s2, c) {
  return cmp(s1, s2, c) === 0;
}

function subset(_s1, _s2, cmp) {
  while (true) {
    const s2 = _s2;
    const s1 = _s1;
    if (s1 === undefined) {
      return true;
    }
    if (s2 === undefined) {
      return false;
    }
    const t2 = Caml_option.valFromOption(s2);
    const t1 = Caml_option.valFromOption(s1);
    const v1 = t1.v;
    const l1 = t1.l;
    const r1 = t1.r;
    const v2 = t2.v;
    const l2 = t2.l;
    const r2 = t2.r;
    const c = cmp(v1, v2);
    if (c === 0) {
      if (!subset(l1, l2, cmp)) {
        return false;
      }
      _s2 = r2;
      _s1 = r1;
      continue;
    }
    if (c < 0) {
      if (!subset(create(l1, v1, undefined), l2, cmp)) {
        return false;
      }
      _s1 = r1;
      continue;
    }
    if (!subset(create(undefined, v1, r1), r2, cmp)) {
      return false;
    }
    _s1 = l1;
    continue;
  };
}

function get(_n, x, cmp) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return;
    }
    const t = Caml_option.valFromOption(n);
    const v = t.v;
    const c = cmp(x, v);
    if (c === 0) {
      return Caml_option.some(v);
    }
    _n = c < 0 ? t.l : t.r;
    continue;
  };
}

function getUndefined(_n, x, cmp) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return;
    }
    const t = Caml_option.valFromOption(n);
    const v = t.v;
    const c = cmp(x, v);
    if (c === 0) {
      return v;
    }
    _n = c < 0 ? t.l : t.r;
    continue;
  };
}

function getExn(_n, x, cmp) {
  while (true) {
    const n = _n;
    if (n !== undefined) {
      const t = Caml_option.valFromOption(n);
      const v = t.v;
      const c = cmp(x, v);
      if (c === 0) {
        return v;
      }
      _n = c < 0 ? t.l : t.r;
      continue;
    }
    throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
        MEL_EXN_ID: Stdlib.Not_found
      });
  };
}

function rotateWithLeftChild(k2) {
  const k1 = k2.l;
  const k1$1 = Caml_option.valFromOption(k1);
  k2.l = k1$1.r;
  k1$1.r = k2;
  const n = k2.l;
  const hlk2 = n !== undefined ? Caml_option.valFromOption(n).h : 0;
  const n$1 = k2.r;
  const hrk2 = n$1 !== undefined ? Caml_option.valFromOption(n$1).h : 0;
  k2.h = (
    hlk2 > hrk2 ? hlk2 : hrk2
  ) + 1 | 0;
  const n$2 = k1$1.l;
  const hlk1 = n$2 !== undefined ? Caml_option.valFromOption(n$2).h : 0;
  const hk2 = k2.h;
  k1$1.h = (
    hlk1 > hk2 ? hlk1 : hk2
  ) + 1 | 0;
  return k1$1;
}

function rotateWithRightChild(k1) {
  const k2 = k1.r;
  const k2$1 = Caml_option.valFromOption(k2);
  k1.r = k2$1.l;
  k2$1.l = k1;
  const n = k1.l;
  const hlk1 = n !== undefined ? Caml_option.valFromOption(n).h : 0;
  const n$1 = k1.r;
  const hrk1 = n$1 !== undefined ? Caml_option.valFromOption(n$1).h : 0;
  k1.h = (
    hlk1 > hrk1 ? hlk1 : hrk1
  ) + 1 | 0;
  const n$2 = k2$1.r;
  const hrk2 = n$2 !== undefined ? Caml_option.valFromOption(n$2).h : 0;
  const hk1 = k1.h;
  k2$1.h = (
    hrk2 > hk1 ? hrk2 : hk1
  ) + 1 | 0;
  return k2$1;
}

function doubleWithLeftChild(k3) {
  const k3l = k3.l;
  const v = rotateWithRightChild(Caml_option.valFromOption(k3l));
  k3.l = v;
  return rotateWithLeftChild(k3);
}

function doubleWithRightChild(k2) {
  const k2r = k2.r;
  const v = rotateWithLeftChild(Caml_option.valFromOption(k2r));
  k2.r = v;
  return rotateWithRightChild(k2);
}

function heightUpdateMutate(t) {
  const n = t.l;
  const hlt = n !== undefined ? Caml_option.valFromOption(n).h : 0;
  const n$1 = t.r;
  const hrt = n$1 !== undefined ? Caml_option.valFromOption(n$1).h : 0;
  t.h = (
    hlt > hrt ? hlt : hrt
  ) + 1 | 0;
  return t;
}

function balMutate(nt) {
  const l = nt.l;
  const r = nt.r;
  const hl = l !== undefined ? Caml_option.valFromOption(l).h : 0;
  const hr = r !== undefined ? Caml_option.valFromOption(r).h : 0;
  if (hl > (2 + hr | 0)) {
    const match = Caml_option.valFromOption(l);
    const ll = match.l;
    const lr = match.r;
    if (heightGe(ll, lr)) {
      return heightUpdateMutate(rotateWithLeftChild(nt));
    } else {
      return heightUpdateMutate(doubleWithLeftChild(nt));
    }
  }
  if (hr > (2 + hl | 0)) {
    const match$1 = Caml_option.valFromOption(r);
    const rl = match$1.l;
    const rr = match$1.r;
    if (heightGe(rr, rl)) {
      return heightUpdateMutate(rotateWithRightChild(nt));
    } else {
      return heightUpdateMutate(doubleWithRightChild(nt));
    }
  }
  nt.h = (
    hl > hr ? hl : hr
  ) + 1 | 0;
  return nt;
}

function addMutate(cmp, t, x) {
  if (t === undefined) {
    return singleton(x);
  }
  const nt = Caml_option.valFromOption(t);
  const k = nt.v;
  const c = cmp(x, k);
  if (c === 0) {
    return t;
  }
  const l = nt.l;
  const r = nt.r;
  if (c < 0) {
    const ll = addMutate(cmp, l, x);
    nt.l = ll;
  } else {
    nt.r = addMutate(cmp, r, x);
  }
  return balMutate(nt);
}

function fromArray(xs, cmp) {
  const len = xs.length;
  if (len === 0) {
    return;
  }
  let next = Belt__Belt_SortArray.strictlySortedLengthU(xs, (function (x, y) {
    return cmp(x, y) < 0;
  }));
  let result;
  if (next >= 0) {
    result = fromSortedArrayAux(xs, 0, next);
  } else {
    next = -next | 0;
    result = fromSortedArrayRevAux(xs, next - 1 | 0, next);
  }
  for (let i = next; i < len; ++i) {
    result = addMutate(cmp, result, xs[i]);
  }
  return result;
}

function removeMinAuxWithRootMutate(nt, n) {
  const ln = n.l;
  const rn = n.r;
  if (ln !== undefined) {
    n.l = removeMinAuxWithRootMutate(nt, Caml_option.valFromOption(ln));
    return balMutate(n);
  } else {
    nt.v = n.v;
    return rn;
  }
}

module.exports = {
  copy,
  create,
  bal,
  singleton,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  removeMinAuxWithRef,
  isEmpty,
  stackAllLeft,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  joinShared,
  concatShared,
  keepSharedU,
  keepShared,
  keepCopyU,
  keepCopy,
  partitionSharedU,
  partitionShared,
  partitionCopyU,
  partitionCopy,
  lengthNode,
  size,
  toList,
  checkInvariantInternal,
  fillArray,
  toArray,
  fromSortedArrayAux,
  fromSortedArrayRevAux,
  fromSortedArrayUnsafe,
  has,
  cmp,
  eq,
  subset,
  get,
  getUndefined,
  getExn,
  fromArray,
  addMutate,
  balMutate,
  removeMinAuxWithRootMutate,
}
/* No side effect */
