// melange.belt/belt_Id.js
// Generated by Melange
'use strict';

const Curry = require("melange.js/curry.js");

function MakeComparableU(M) {
  return M;
}

function MakeComparable(M) {
  const cmp = M.cmp;
  const cmp$1 = Curry.__2(cmp);
  return {
    cmp: cmp$1
  };
}

function comparableU(cmp) {
  return {
    cmp: cmp
  };
}

function comparable(cmp) {
  const cmp$1 = Curry.__2(cmp);
  return {
    cmp: cmp$1
  };
}

function MakeHashableU(M) {
  return M;
}

function MakeHashable(M) {
  const hash = M.hash;
  const hash$1 = Curry.__1(hash);
  const eq = M.eq;
  const eq$1 = Curry.__2(eq);
  return {
    hash: hash$1,
    eq: eq$1
  };
}

function hashableU(hash, eq) {
  return {
    hash: hash,
    eq: eq
  };
}

function hashable(hash, eq) {
  const hash$1 = Curry.__1(hash);
  const eq$1 = Curry.__2(eq);
  return {
    hash: hash$1,
    eq: eq$1
  };
}

module.exports = {
  MakeComparableU,
  MakeComparable,
  comparableU,
  comparable,
  MakeHashableU,
  MakeHashable,
  hashableU,
  hashable,
}
/* No side effect */
