// melange.belt/belt_internalBuckets.js
// Generated by Melange
'use strict';

const Belt__Belt_Array = require("./belt_Array.js");
const Caml_option = require("melange.js/caml_option.js");
const Curry = require("melange.js/curry.js");

function copyAuxCont(_c, _prec) {
  while (true) {
    const prec = _prec;
    const c = _c;
    if (c === undefined) {
      return;
    }
    const ncopy = {
      key: c.key,
      value: c.value,
      next: undefined
    };
    prec.next = ncopy;
    _prec = ncopy;
    _c = c.next;
    continue;
  };
}

function copyBucket(c) {
  if (c === undefined) {
    return c;
  }
  const head = {
    key: c.key,
    value: c.value,
    next: undefined
  };
  copyAuxCont(c.next, head);
  return head;
}

function copyBuckets(buckets) {
  const len = buckets.length;
  const newBuckets = new Array(len);
  for (let i = 0; i < len; ++i) {
    newBuckets[i] = copyBucket(buckets[i]);
  }
  return newBuckets;
}

function copy(x) {
  return {
    size: x.size,
    buckets: copyBuckets(x.buckets),
    hash: x.hash,
    eq: x.eq
  };
}

function bucketLength(_accu, _buckets) {
  while (true) {
    const buckets = _buckets;
    const accu = _accu;
    if (buckets === undefined) {
      return accu;
    }
    _buckets = buckets.next;
    _accu = accu + 1 | 0;
    continue;
  };
}

function do_bucket_iter(f, _buckets) {
  while (true) {
    const buckets = _buckets;
    if (buckets === undefined) {
      return;
    }
    f(buckets.key, buckets.value);
    _buckets = buckets.next;
    continue;
  };
}

function forEachU(h, f) {
  const d = h.buckets;
  for (let i = 0, i_finish = d.length; i < i_finish; ++i) {
    do_bucket_iter(f, d[i]);
  }
}

function forEach(h, f) {
  forEachU(h, Curry.__2(f));
}

function do_bucket_fold(f, _b, _accu) {
  while (true) {
    const accu = _accu;
    const b = _b;
    if (b === undefined) {
      return accu;
    }
    _accu = f(accu, b.key, b.value);
    _b = b.next;
    continue;
  };
}

function reduceU(h, init, f) {
  const d = h.buckets;
  let accu = init;
  for (let i = 0, i_finish = d.length; i < i_finish; ++i) {
    accu = do_bucket_fold(f, d[i], accu);
  }
  return accu;
}

function reduce(h, init, f) {
  return reduceU(h, init, Curry.__3(f));
}

function getMaxBucketLength(h) {
  return Belt__Belt_Array.reduceU(h.buckets, 0, (function (m, b) {
    const len = bucketLength(0, b);
    if (m > len) {
      return m;
    } else {
      return len;
    }
  }));
}

function getBucketHistogram(h) {
  const mbl = getMaxBucketLength(h);
  const histo = Belt__Belt_Array.makeByU(mbl + 1 | 0, (function (param) {
    return 0;
  }));
  Belt__Belt_Array.forEachU(h.buckets, (function (b) {
    const l = bucketLength(0, b);
    histo[l] = histo[l] + 1 | 0;
  }));
  return histo;
}

function logStats(h) {
  const histogram = getBucketHistogram(h);
  console.log({
    bindings: h.size,
    buckets: h.buckets.length,
    histogram: histogram
  });
}

function filterMapInplaceBucket(f, h, i, _prec, _cell) {
  while (true) {
    const cell = _cell;
    const prec = _prec;
    const n = cell.next;
    const data = f(cell.key, cell.value);
    if (data !== undefined) {
      if (prec !== undefined) {
        cell.next = cell;
      } else {
        h.buckets[i] = cell;
      }
      cell.value = Caml_option.valFromOption(data);
      if (n === undefined) {
        cell.next = n;
        return;
      }
      _cell = n;
      _prec = cell;
      continue;
    }
    h.size = h.size - 1 | 0;
    if (n === undefined) {
      if (prec !== undefined) {
        prec.next = n;
      } else {
        h.buckets[i] = prec;
      }
      return;
    }
    _cell = n;
    continue;
  };
}

function keepMapInPlaceU(h, f) {
  const h_buckets = h.buckets;
  for (let i = 0, i_finish = h_buckets.length; i < i_finish; ++i) {
    const v = h_buckets[i];
    if (v !== undefined) {
      filterMapInplaceBucket(f, h, i, undefined, v);
    }
    
  }
}

function keepMapInPlace(h, f) {
  keepMapInPlaceU(h, Curry.__2(f));
}

function fillArray(_i, arr, _cell) {
  while (true) {
    const cell = _cell;
    const i = _i;
    arr[i] = [
      cell.key,
      cell.value
    ];
    const v = cell.next;
    if (v === undefined) {
      return i + 1 | 0;
    }
    _cell = v;
    _i = i + 1 | 0;
    continue;
  };
}

function fillArrayMap(_i, arr, _cell, f) {
  while (true) {
    const cell = _cell;
    const i = _i;
    arr[i] = f(cell);
    const v = cell.next;
    if (v === undefined) {
      return i + 1 | 0;
    }
    _cell = v;
    _i = i + 1 | 0;
    continue;
  };
}

function linear(h, f) {
  const d = h.buckets;
  let current = 0;
  const arr = new Array(h.size);
  for (let i = 0, i_finish = d.length; i < i_finish; ++i) {
    const cell = d[i];
    if (cell !== undefined) {
      current = fillArrayMap(current, arr, cell, f);
    }
    
  }
  return arr;
}

function keysToArray(h) {
  return linear(h, (function (x) {
    return x.key;
  }));
}

function valuesToArray(h) {
  return linear(h, (function (x) {
    return x.value;
  }));
}

function toArray(h) {
  return linear(h, (function (x) {
    return [
      x.key,
      x.value
    ];
  }));
}

module.exports = {
  copy,
  forEachU,
  forEach,
  reduceU,
  reduce,
  logStats,
  keepMapInPlaceU,
  keepMapInPlace,
  fillArray,
  keysToArray,
  valuesToArray,
  toArray,
  getBucketHistogram,
}
/* No side effect */
