// melange.belt/belt_internalMapString.js
// Generated by Melange
'use strict';

const Belt__Belt_SortArray = require("./belt_SortArray.js");
const Belt__Belt_internalAVLtree = require("./belt_internalAVLtree.js");
const Caml = require("melange.js/caml.js");
const Caml_js_exceptions = require("melange.js/caml_js_exceptions.js");
const Caml_option = require("melange.js/caml_option.js");
const Curry = require("melange.js/curry.js");
const Stdlib = require("melange/stdlib.js");

function add(t, x, data) {
  if (t === undefined) {
    return Belt__Belt_internalAVLtree.singleton(x, data);
  }
  const n = Caml_option.valFromOption(t);
  const k = n.k;
  if (x === k) {
    return Belt__Belt_internalAVLtree.updateValue(n, data);
  }
  const v = n.v;
  if (x < k) {
    return Belt__Belt_internalAVLtree.bal(add(n.l, x, data), k, v, n.r);
  } else {
    return Belt__Belt_internalAVLtree.bal(n.l, k, v, add(n.r, x, data));
  }
}

function get(_n, x) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return;
    }
    const n$1 = Caml_option.valFromOption(n);
    const v = n$1.k;
    if (x === v) {
      return Caml_option.some(n$1.v);
    }
    _n = x < v ? n$1.l : n$1.r;
    continue;
  };
}

function getUndefined(_n, x) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return;
    }
    const n$1 = Caml_option.valFromOption(n);
    const v = n$1.k;
    if (x === v) {
      return n$1.v;
    }
    _n = x < v ? n$1.l : n$1.r;
    continue;
  };
}

function getExn(_n, x) {
  while (true) {
    const n = _n;
    if (n !== undefined) {
      const n$1 = Caml_option.valFromOption(n);
      const v = n$1.k;
      if (x === v) {
        return n$1.v;
      }
      _n = x < v ? n$1.l : n$1.r;
      continue;
    }
    throw new Caml_js_exceptions.MelangeError(Stdlib.Not_found, {
        MEL_EXN_ID: Stdlib.Not_found
      });
  };
}

function getWithDefault(_n, x, def) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return def;
    }
    const n$1 = Caml_option.valFromOption(n);
    const v = n$1.k;
    if (x === v) {
      return n$1.v;
    }
    _n = x < v ? n$1.l : n$1.r;
    continue;
  };
}

function has(_n, x) {
  while (true) {
    const n = _n;
    if (n === undefined) {
      return false;
    }
    const n$1 = Caml_option.valFromOption(n);
    const v = n$1.k;
    if (x === v) {
      return true;
    }
    _n = x < v ? n$1.l : n$1.r;
    continue;
  };
}

function remove(n, x) {
  if (n === undefined) {
    return n;
  }
  const n$1 = Caml_option.valFromOption(n);
  const v = n$1.k;
  const l = n$1.l;
  const r = n$1.r;
  if (x !== v) {
    if (x < v) {
      return Belt__Belt_internalAVLtree.bal(remove(l, x), v, n$1.v, r);
    } else {
      return Belt__Belt_internalAVLtree.bal(l, v, n$1.v, remove(r, x));
    }
  }
  if (l === undefined) {
    return r;
  }
  if (r === undefined) {
    return l;
  }
  const rn = Caml_option.valFromOption(r);
  const kr = {
    contents: rn.k
  };
  const vr = {
    contents: rn.v
  };
  const r$1 = Belt__Belt_internalAVLtree.removeMinAuxWithRef(rn, kr, vr);
  return Belt__Belt_internalAVLtree.bal(l, kr.contents, vr.contents, r$1);
}

function splitAux(x, n) {
  const v = n.k;
  const d = n.v;
  const l = n.l;
  const r = n.r;
  if (x === v) {
    return [
      l,
      Caml_option.some(d),
      r
    ];
  }
  if (x < v) {
    if (l === undefined) {
      return [
        undefined,
        undefined,
        n
      ];
    }
    const match = splitAux(x, Caml_option.valFromOption(l));
    return [
      match[0],
      match[1],
      Belt__Belt_internalAVLtree.join(match[2], v, d, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined,
      undefined
    ];
  }
  const match$1 = splitAux(x, Caml_option.valFromOption(r));
  return [
    Belt__Belt_internalAVLtree.join(l, v, d, match$1[0]),
    match$1[1],
    match$1[2]
  ];
}

function split(x, n) {
  if (n !== undefined) {
    return splitAux(x, n);
  } else {
    return [
      undefined,
      undefined,
      undefined
    ];
  }
}

function mergeU(s1, s2, f) {
  if (s1 !== undefined) {
    const n = Caml_option.valFromOption(s1);
    if (n.h >= (
        s2 !== undefined ? s2.h : 0
      )) {
      const v1 = n.k;
      const d1 = n.v;
      const l1 = n.l;
      const r1 = n.r;
      const match = split(v1, s2);
      return Belt__Belt_internalAVLtree.concatOrJoin(mergeU(l1, match[0], f), v1, f(v1, Caml_option.some(d1), match[1]), mergeU(r1, match[2], f));
    }
    
  } else if (s2 === undefined) {
    return;
  }
  const n$1 = Caml_option.valFromOption(s2);
  const v2 = n$1.k;
  const d2 = n$1.v;
  const l2 = n$1.l;
  const r2 = n$1.r;
  const match$1 = split(v2, s1);
  return Belt__Belt_internalAVLtree.concatOrJoin(mergeU(match$1[0], l2, f), v2, f(v2, match$1[1], Caml_option.some(d2)), mergeU(match$1[2], r2, f));
}

function merge(s1, s2, f) {
  return mergeU(s1, s2, Curry.__3(f));
}

function compareAux(_e1, _e2, vcmp) {
  while (true) {
    const e2 = _e2;
    const e1 = _e1;
    if (!e1) {
      return 0;
    }
    if (!e2) {
      return 0;
    }
    const h2 = e2.hd;
    const h1 = e1.hd;
    const c = Caml.caml_string_compare(h1.k, h2.k);
    if (c !== 0) {
      return c;
    }
    const cx = vcmp(h1.v, h2.v);
    if (cx !== 0) {
      return cx;
    }
    _e2 = Belt__Belt_internalAVLtree.stackAllLeft(h2.r, e2.tl);
    _e1 = Belt__Belt_internalAVLtree.stackAllLeft(h1.r, e1.tl);
    continue;
  };
}

function cmpU(s1, s2, cmp) {
  const len1 = Belt__Belt_internalAVLtree.size(s1);
  const len2 = Belt__Belt_internalAVLtree.size(s2);
  if (len1 === len2) {
    return compareAux(Belt__Belt_internalAVLtree.stackAllLeft(s1, /* [] */ 0), Belt__Belt_internalAVLtree.stackAllLeft(s2, /* [] */ 0), cmp);
  } else if (len1 < len2) {
    return -1;
  } else {
    return 1;
  }
}

function cmp(s1, s2, f) {
  return cmpU(s1, s2, Curry.__2(f));
}

function eqAux(_e1, _e2, eq) {
  while (true) {
    const e2 = _e2;
    const e1 = _e1;
    if (!e1) {
      return true;
    }
    if (!e2) {
      return true;
    }
    const h2 = e2.hd;
    const h1 = e1.hd;
    if (!(h1.k === h2.k && eq(h1.v, h2.v))) {
      return false;
    }
    _e2 = Belt__Belt_internalAVLtree.stackAllLeft(h2.r, e2.tl);
    _e1 = Belt__Belt_internalAVLtree.stackAllLeft(h1.r, e1.tl);
    continue;
  };
}

function eqU(s1, s2, eq) {
  const len1 = Belt__Belt_internalAVLtree.size(s1);
  const len2 = Belt__Belt_internalAVLtree.size(s2);
  if (len1 === len2) {
    return eqAux(Belt__Belt_internalAVLtree.stackAllLeft(s1, /* [] */ 0), Belt__Belt_internalAVLtree.stackAllLeft(s2, /* [] */ 0), eq);
  } else {
    return false;
  }
}

function eq(s1, s2, f) {
  return eqU(s1, s2, Curry.__2(f));
}

function addMutate(t, x, data) {
  if (t === undefined) {
    return Belt__Belt_internalAVLtree.singleton(x, data);
  }
  const nt = Caml_option.valFromOption(t);
  const k = nt.k;
  if (x === k) {
    nt.k = x;
    nt.v = data;
    return nt;
  }
  const l = nt.l;
  const r = nt.r;
  if (x < k) {
    const ll = addMutate(l, x, data);
    nt.l = ll;
  } else {
    nt.r = addMutate(r, x, data);
  }
  return Belt__Belt_internalAVLtree.balMutate(nt);
}

function fromArray(xs) {
  const len = xs.length;
  if (len === 0) {
    return;
  }
  let next = Belt__Belt_SortArray.strictlySortedLengthU(xs, (function (param, param$1) {
    return param[0] < param$1[0];
  }));
  let result;
  if (next >= 0) {
    result = Belt__Belt_internalAVLtree.fromSortedArrayAux(xs, 0, next);
  } else {
    next = -next | 0;
    result = Belt__Belt_internalAVLtree.fromSortedArrayRevAux(xs, next - 1 | 0, next);
  }
  for (let i = next; i < len; ++i) {
    const match = xs[i];
    result = addMutate(result, match[0], match[1]);
  }
  return result;
}

module.exports = {
  add,
  get,
  getUndefined,
  getExn,
  getWithDefault,
  has,
  remove,
  splitAux,
  split,
  mergeU,
  merge,
  compareAux,
  cmpU,
  cmp,
  eqAux,
  eqU,
  eq,
  addMutate,
  fromArray,
}
/* No side effect */
